<!doctype html>
<html lang="en">
<head>
  <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
  <meta name="viewport" content="width=1034, initial-scale=1.0">
  <title>Primer3 Release 2.5.0 Manual</title>
  <style>
  body {
  background-color:white;
  color:black;
  font-size:100.01%;
  margin:0;
  min-width:41em;
  padding:5px;
  text-align:left;
  }
  td {
  font-size:0.78em;
  }
  h2 {
  padding-top:1.2em;
  }
  h3 {
  padding-top:1.2em;
  }
  div#page {
  background:#FFFFFF;
  margin:25px;
  padding:0;
  text-align:left;
  width:850px;
  }
  .p3_prop {
    font-family: monospace;
    white-space: pre;
    margin: 1em 0;
  }
  </style>
</head>
<body>
<div id="page">
<h1>PRIMER3 RELEASE 2.5.0 MANUAL</h1>
<h2>CONTENTS</h2><p>
<a href="#copyrightLicense">1. COPYRIGHT AND LICENSE</a><br>
<a href="#introduction">2. INTRODUCTION</a><br>
<a href="#citationRequest">3. CITING PRIMER3</a><br>
<a href="#licenseExplain">4. FAIR USE OF PRIMER3</a><br>
<a href="#differencePrimer3vsPrimer3Plus">5. DIFFERENCE BETWEEN PRIMER3 AND PRIMER3PLUS</a><br>
<a href="#changesFrom2.2.3">6. CHANGES FROM VERSION 2.2.3</a><br>
<a href="#installLinux">7. INSTALLATION INSTRUCTIONS - UNIX/LINUX</a><br>
<a href="#installMac">8. INSTALLATION INSTRUCTIONS - Mac OSX</a><br>
<a href="#installWindows">9. INSTALLATION INSTRUCTIONS - WINDOWS</a><br>
<a href="#systemRequirements">10. SYSTEM REQUIREMENTS</a><br>
<a href="#invokingPrimer3">11. INVOKING primer3_core</a><br>
<a href="#commandLineTags">12. COMMAND LINE ARGUMENTS</a><br>
<a href="#inputOutputConventions">13. INPUT AND OUTPUT CONVENTIONS</a><br>
<a href="#example">14. AN EXAMPLE</a><br>
<a href="#migrateTags">15. HOW TO MIGRATE TAGS TO IO VERSION 4</a><br>
<a href="#sequenceTags">16. "SEQUENCE" INPUT TAGS</a><br>
<a href="#globalTags">17. "GLOBAL" INPUT TAGS</a><br>
<a href="#programTags">18. "PROGRAM" INPUT TAGS</a><br>
<a href="#calculatePenalties">19. HOW PRIMER3 CALCULATES THE PENALTY VALUE</a><br>
<a href="#fileFormat">20. PRIMER3 SETTINGS FILE FORMAT</a><br>
<a href="#outputTags">21. OUTPUT TAGS</a><br>
<a href="#exampleOutput">22. EXAMPLE OUTPUT</a><br>
<a href="#pickAdvice">23. ADVICE FOR PICKING PRIMERS</a><br>
<a href="#cautions">24. CAUTIONS</a><br>
<a href="#findNoPrimers">25. WHAT TO DO IF PRIMER3 CANNOT FIND ANY PRIMERS?</a><br>
<a href="#earlierVersions">26. DIFFERENCES FROM EARLIER VERSIONS</a><br>
<a href="#exitStatusCodes">27. EXIT STATUS CODES</a><br>
<a href="#webInterface">28. PRIMER3 WWW INTERFACES</a><br>
<a href="#acknowledgments">29. ACKNOWLEDGMENTS</a><br>
</p>
<h2><a id="copyrightLicense">1. COPYRIGHT AND LICENSE</a></h2>
<pre>Copyright (c) 1996,1997,1998,1999,2000,2001,2004,2006,2007,2008,2009,2010
              2011,2012,2013,2016,2017
Whitehead Institute for Biomedical Research, Steve Rozen
(<a href="http://purl.com/STEVEROZEN/">http://purl.com/STEVEROZEN/</a>), Helen Skaletsky, Triinu Koressaar,
Maido Remm and Andreas Untergasser. All rights reserved.
    This file is part of the Primer3 suite and libraries.
    The Primer3 suite and libraries are free software;
    you can redistribute them and/or modify them under the terms
    of the GNU General Public License as published by the Free
    Software Foundation; either version 2 of the License, or (at
    your option) any later version.
    This software is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    You should have received a copy of the GNU General Public License
    along with this software (file gpl-2.0.txt in the source
    distribution); if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
</pre>
<h2><a id="introduction">2. INTRODUCTION</a></h2>
<p>Primer3 picks primers for PCR reactions, considering as criteria:
<br>
o oligonucleotide melting temperature, size, GC content,
  and primer-dimer possibilities,<br>
<br>
o PCR product size,<br>
<br>
o positional constraints within the source (template) sequence, and<br>
<br>
o possibilities for ectopic priming (amplifying the wrong sequence)<br>
<br>
o many other constraints.<br>
<br>
All of these criteria are user-specifiable as constraints, and
some are specifiable as terms in an objective function that
characterizes an optimal primer pair.<br>
<h2><a id="citationRequest">3. CITING PRIMER3</a></h2>
<p>We request but do not require that use of this software be cited in
publications as<br>
<br>
* Untergasser A, Cutcutache I, Koressaar T, Ye J, Faircloth BC, Remm M and Rozen SG.<br>
Primer3--new capabilities and interfaces.<br>
Nucleic Acids Res. 2012 Aug 1;40(15):e115.<br>
The paper is available at
<a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3424584/">http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3424584/</a><br>
<br>
and <br>
<br>
* Koressaar T and Remm M.<br>
Enhancements and modifications of primer design program Primer3.<br>
Bioinformatics 2007;23(10):1289-1291.<br>
The paper is available at
<a href="https://www.ncbi.nlm.nih.gov/pubmed/17379693">https://www.ncbi.nlm.nih.gov/pubmed/17379693</a><br>
<br>
If you use masker function, please cite:<br>
* Koressaar T, Lepamets M, Kaplinski L, Raime K, Andreson R and Remm M.<br>
Primer3_masker: integrating masking of template sequence with primer design software.<br>
Bioinformatics 2018;34(11):1937-1938.<br>
The paper is available at
<a href="https://www.ncbi.nlm.nih.gov/pubmed/29360956">https://www.ncbi.nlm.nih.gov/pubmed/29360956</a><br>
<br>
Source code available at <a href="https://github.com/primer3-org/primer3">https://github.com/primer3-org/primer3</a>.</p>
<h2><a id="licenseExplain">4. FAIR USE OF PRIMER3</a></h2>
<p>
The development of Primer3 is promoted by a small group of
enthusiastic scientists mainly in their free time.They do not gain
any financial profit with Primer3.<br>
<br>
There are two groups of Primer3 users: end users, who run
Primer3 to pick their primers and programmers, who use Primer3
in their scripts or software packages. We encourage both to use
Primer3.
<br>
<br>
If you are an end user, we request but do not
require that use of this software be cited in publications
as listed above under CITING PRIMER3.
<br>
<br>
If you are a programmer, you  will see that Primer3 is now
distributed under the GNU  General Public License, version 2 or
(at  your option) any later version of the License (GPL2).
As we understand it, if you include parts of the Primer3 source
code in your source code or link to Primer3 binary libraries in
your executable, you have to release your software also under
GPL2. If you only call Primer3 from your software and interpret
its output, you can use any license you want for your software.
If you modify Primer3 and then release your modified software,
you have to release your modifications in source code under
GPL2 as well.
<br>
<br>
We chose GPL2 because we wanted Primer3 to evolve and for the
improvements to find their way back into the main distribution.
If you are programming a new web interface which runs Primer3,
please include in the about page of the tool the sentence
"&lt;your software name&gt; uses Primer3 version ...".
Please consider releasing your software under GPL2 as well,
especially if you do not want to maintain it in the future.
<br>
<br>
There is no need to ask us for permission to include Primer3
in your tools.
<h2><a id="differencePrimer3vsPrimer3Plus">5. DIFFERENCE BETWEEN PRIMER3 AND PRIMER3PLUS</a></h2>
<p>Primer3Plus is a web interface to Primer3, so if you pick
primers with Primer3Plus, it will collect and reformat your input, run
the command line tool Primer3, collet and reformats it's output and
display it to you.<br>
In principle, both tools would give you the same output. In practice,
the default settings ob both tools differ. While Primer3 default
settings are usually kept for backward compatibility, the Primer3Plus
default settings are adapted for regular wetlab use.</p>
<h4>Primer3 Values:</h4>
<pre>
  PRIMER_PRODUCT_SIZE_RANGE=100-300
  PRIMER_SECONDARY_STRUCTURE_ALIGNMENT=0
  PRIMER_NUM_RETURN=5
  PRIMER_MAX_HAIRPIN_TH=47.0
  PRIMER_INTERNAL_MAX_HAIRPIN_TH=47.0
  PRIMER_MAX_END_STABILITY=100.0
  PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE=-1
  PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE=-1
  PRIMER_EXPLAIN_FLAG=0
  PRIMER_LIBERAL_BASE=0
  PRIMER_FIRST_BASE_INDEX=0
  PRIMER_MAX_TEMPLATE_MISPRIMING=-1.00
  PRIMER_MAX_TEMPLATE_MISPRIMING_TH=-1.00
  PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING=-1.00
  PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING_TH=-1.00
</pre>
<h4>Primer3Plus Values:</h4>
<pre>
  PRIMER_PRODUCT_SIZE_RANGE=501-600 601-700 401-500 701-850 851-1000 1001-1500
                            1501-3000 3001-5000 401-500 301-400 201-300
                            101-200 5001-7000 7001-10000 10001-20000
  PRIMER_SECONDARY_STRUCTURE_ALIGNMENT=1
  PRIMER_NUM_RETURN=10
  PRIMER_MAX_HAIRPIN_TH=47.00
  PRIMER_INTERNAL_MAX_HAIRPIN_TH=47.00
  PRIMER_MAX_END_STABILITY=9.0
  PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE=3
  PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE=3
  PRIMER_EXPLAIN_FLAG=1
  PRIMER_LIBERAL_BASE=1
  PRIMER_FIRST_BASE_INDEX=1
  PRIMER_MAX_TEMPLATE_MISPRIMING=12.00
  PRIMER_MAX_TEMPLATE_MISPRIMING_TH=47.00
  PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING=24.00
  PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING_TH=47.00
</pre>
<h2><a id="changesFrom2.2.3">6. CHANGES FROM VERSION 2.2.3</a></h2>
<p>
1. The tag PRIMER_THERMODYNAMIC_ALIGNMENT was replaced by two new tags:
<br><br>
   - <a href="#PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT">PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT</a> which governs hairpin
and oligo-oligo interactions. Therefore, whenever it is set to 1
(the default) thermodynamic alignments will be used for oligo-oligo
interactions and hairpins.
<br><br>
   - <a href="#PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT">PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT</a> which governs the
oligo-template interactions and, when set to 1 (default is 0), will
cause Primer3 to use the thermodynamic alignment against templates.
<br><br>
The reason for this change is to provide the option of using the
old alignment for oligo-template interactions when the thermodynamic
alignment is too slow or the template is too long (there is a hard
limit THAL_MAX_SEQ=10k on the length of sequences involved in
thermodynamic alignments).
<br><br>
2. We changed the following default values:<br>
<br>
2.1. Changing default oligo temperature calculations
<br><br>
<a href="#PRIMER_TM_FORMULA">PRIMER_TM_FORMULA</a>=1  (was 0)<br>
<a href="#PRIMER_SALT_CORRECTIONS">PRIMER_SALT_CORRECTIONS</a>=1 (was 0)<br>
<br>
2.2. Making thermodynamic secondary structure calculations for oligos the default:<br>
<br>
<a href="#PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT">PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT</a>=1 (was 0)<br>
<br>
2.3. The following need reasonable values to support the changes in 2.1
and 2.2 above:<br>
<br>
<a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a>=1.5 (was 0.0)<br>
<a href="#PRIMER_DNTP_CONC">PRIMER_DNTP_CONC</a>=0.6 (was 0.0)<br>
<br>
2.4. To make old defaults easily accessible, we added a command line
argument, --default_version=1, which directs primer3_core to use the
old defaults. --default_version=2 directs primer3_core to use the new
defaults.  The default is --default_version=2.<br>
<br>
2.5. IMPORTANT: because <a href="#PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT">PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT</a>=1,
<a href="#PRIMER_THERMODYNAMIC_PARAMETERS_PATH">PRIMER_THERMODYNAMIC_PARAMETERS_PATH</a> must point to the right location.
This tag specifies the path to the directory that contains all the
parameter files used by the thermodynamic approach. In Linux, there
are two *default* locations that are tested if this tag is not
defined: ./primer3_config/ and /opt/primer3_config/. For Windows,
there is only one default location: .\primer3_config\.  If the the
parameter files are not in one these locations, be sure to set
<a href="#PRIMER_THERMODYNAMIC_PARAMETERS_PATH">PRIMER_THERMODYNAMIC_PARAMETERS_PATH</a>.<br>
<br>
2.6. Changed default for <br>
<br>
<a href="#PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS</a>=0 (was 1)<br>
<br>
(0 is almost always the the behavior one wants.)<br>
<br>
2.7. To get the behavior of -default_version=1 when -default_version=2
set the following:<br>
<br>
<a href="#PRIMER_TM_FORMULA">PRIMER_TM_FORMULA</a>=0<br>
<a href="#PRIMER_SALT_CORRECTIONS">PRIMER_SALT_CORRECTIONS</a>=0<br>
<a href="#PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT">PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT</a>=0<br>
<a href="#PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT">PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT</a>=0<br>
<a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a>=0.0<br>
<a href="#PRIMER_DNTP_CONC">PRIMER_DNTP_CONC</a>=0.0<br>
<a href="#PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS</a>=1<br>
<br>
2.8. To get the behavior of -default_version=2 when -default_version=1
set the following:<br>
<br>
<a href="#PRIMER_TM_FORMULA">PRIMER_TM_FORMULA</a>=1<br>
<a href="#PRIMER_SALT_CORRECTIONS">PRIMER_SALT_CORRECTIONS</a>=1<br>
<a href="#PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT">PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT</a>=1<br>
<a href="#PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT">PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT</a>=0<br>
<a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a>=1.5<br>
<a href="#PRIMER_DNTP_CONC">PRIMER_DNTP_CONC</a>=0.6<br>
<a href="#PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS</a>=0<br>
<br>
3. We changed the NULL value for SEQUENCE_FORCE_{LEFT,RIGHT}_START_END to
-1000000, and made it an error to select <a href="#PRIMER_FIRST_BASE_INDEX">PRIMER_FIRST_BASE_INDEX</a> &lt;=
this value.  This is to correct an error when
SEQUENCE_FORCE_{LEFT,RIGHT}_START_END was -1 (value previously used to
indicate a NULL) but <a href="#PRIMER_FIRST_BASE_INDEX">PRIMER_FIRST_BASE_INDEX</a> was &lt; 0, which caused the
intended NULL value (-1) to be treated as a constraint on primer
location (a constraint that was not possible to satisfy).<br>
<br>
4. We changed the <a href="#PRIMER_TASK">PRIMER_TASK</a> called 'pick_detection_primers' to
'generic' while retaining 'pick_detection_primers' as an alias for
backward compatibility.<br>
<br>
5. The code now uses 'end' alignments when assessing template mispriming
using thermodynamic alignments. This is consistent with the approach
taken with the previous alignment algorithm and with checking for
mispriming against repeat libraries.<br>
<br>
6. We removed PRIMER_PAIR_MAX_HAIRPIN_TH (which was ignored previously).<br>
<br>
7. Primer3 now requires the user to set <a href="#SEQUENCE_TARGET">SEQUENCE_TARGET</a>, not
<a href="#SEQUENCE_INCLUDED_REGION">SEQUENCE_INCLUDED_REGION</a> when <a href="#PRIMER_TASK">PRIMER_TASK</a>=pick_discriminative_primers<br>
<br>
8. When <a href="#PRIMER_TASK">PRIMER_TASK</a>=pick_discriminative_primers or
<a href="#PRIMER_TASK">PRIMER_TASK</a>=pick_cloning_primers the value of <a href="#SEQUENCE_INCLUDED_REGION">SEQUENCE_INCLUDED_REGION</a>
is no longer changed to the entire input sequence.<br>
<br>
9. The handling of divalent cations when <a href="#PRIMER_SALT_CORRECTIONS">PRIMER_SALT_CORRECTIONS</a>=2
(not the default and not the recommended value) has been updated. The
rationale is that, when divalent cations are present, the formula by
Owczarzy et al., 2004 (used previously by Primer3) can be improved
upon as described in (Ahsen et al., 2010; Owczarzy et al.,
2008). Therefore we have updated the melting temperature calculation
to follow the scheme in Figure 9 of (Owczarzy et al., 2008). Please
find references to these papers below in the manual.<br>
<br>
10. We modified the error handling of non-memory related errors that
can occur during the thermodynamic alignment. They are no longer
fatal errors.<br>
<br>
11. We removed io_version 3, which was only kept for backward compatibility
with very old versions of Primer3.<br>
<br>
12. In addition, several error corrections:<br>
<br>
12.1. Corrected a short, fixed-size buffer for the file/path names specified
by the command line arguments -output, -error and for the settings
file.<br>
<br>
12.2. Two other corrected errors were in p3_set_gs_primer_self_end and
p3_set_gs_primer_internal_oligo_self_end, which erroneously multiplied
their 'val' arguments by 100.<br>
<br>
12.3. Primer3 now detects and handles the situation in which user-supplied
primers (<a href="#SEQUENCE_PRIMER">SEQUENCE_PRIMER</a> and <a href="#SEQUENCE_PRIMER_REVCOMP">SEQUENCE_PRIMER_REVCOMP</a>) have the left
primer to the right of the right primer.  Also issue a warning if
user-supplied primers occur in more than one location in the template.<br></p>
<h2><a id="installLinux">7. INSTALLATION INSTRUCTIONS - UNIX/LINUX</a></h2>
<p>Unzip and untar the distribution.
DO NOT do this on a PC if you are going to
compile and/or test on a different operation system.
Primer3_core will not compile
and various tests will fail if pc
newlines get inserted into the code and test files.  Instead, move the
distribution (primer3-&lt;release&gt;.tar.gz) to Unix/Linux, and then<br>
<br>
$ unzip primer3-&lt;release&gt;.tar.gz<br>
$ tar xvf primer3-&lt;release&gt;.tar<br>
$ cd primer3-&lt;release&gt;/src<br>
<br>
If you do not use gcc, then you will probably have
to modify the makefile to
use your (ANSI) C compiler with the compile
and link flags it understands.<br>
<br>
$ make all<br>
<br>
# You should have created executables primer3_core, ntdpal,<br>
# olgotm, and long_seq_tm_test<br>
<br>
$ make test<br>
<br>
# You should not see 'FAILED' during the tests.<br>
<br>
If your perl command is not called perl (for example, if it is
called perl5) you will have to modify the
Makefile in the test/ directory.<br>
<br>
ntdpal (NucleoTide Dynamic Programming ALignment) is a
stand-alone program that provides Primer3's alignment
functionality (local, a.k.a. Smith-Waterman, global,
a.k.a. Needleman-Wunsch, plus "half global").  It is provided
strictly as is; for further documentation please see the code.</p>
<h2><a id="installMac">8. INSTALLATION INSTRUCTIONS - Mac OSX</a></h2>
<p>How to install this software<br>
============================<br>
<br>
1.  Double click on the .tar.gz file to extract the archive.<br>
<br>
2.  The binary files are located in the 'bin' [for 'binary']
    folder<br>
<br>
3.  (Optional) To run the tests, cd to the new directory and
    then the test folder<br>
<br>
4.  (Optional) Within this folder run:
	a. 'perl -w p3test.pl'<br>
<br>
5.  (Optional) You should not see 'FAILED' during the tests.<br>
<br>
6.  (Optional) *NOTE*:  If your perl command is not called
    perl (for example, if it is called perl5) you will have
    to modify the internals of the test scripts).<br>
<br>
7.	Copy the following files to a location of your choice:<br>
	a.  bin/long_seq_tm_test<br>
	b.  bin/ntdpal<br>
	c.  bin/oligotm<br>
	d.  bin/primer3_core<br>
<br>
8.	(Optional) Make sure this location is within your $PATH
    (see below)<br>
<br>
<br>
Where to put the binary files<br>
=============================<br>
<br>
A good place to put these is within ~/bin/ (this means in
your home folder, within a folder named `bin` [for 'binary']).<br>
<br>
You can also just drag the 'bin' folder to a location within
your home directory.<br>
<br>
You can certainly also copy the files within 'bin' to
/usr/local/bin (if you are an administrator) or another
similar location.<br>
<br>
You may need to adjust the permissions on the binaries if
you get fancy.<br>
<br>
<br>
Add the location to your $PATH<br>
==============================<br>
<br>
This is an optional step, but it will allow you to run
Primer3 in any directory on your machine as your user just
by typing its name (primer3_core).<br>
<br>
*** You should be very careful when altering your $PATH as
things can go very wrong.  See below for an alternate
method. *** <br>
<br>
If you added the binaries to /usr/local/bin, then you do
not need to do this.<br>
<br>
If you added the binaries to a local directory (let's say
~/bin/), do the following:<br>
<br>
	1.  Edit your ~/.bash_profile.  You can edit this on
        the command line (Terminal) with:<br>
<br>
		nano ~/.bash_profile<br>
<br>
	2. Add the following line if it is not present
       (replacing '~/bin' if you used another directory):<br>
<br>
		PATH=$PATH:~/bin/<br>
<br>
	3. If a PATH line *is* present, ensure you add a colon
       to the end of what is there and then the directory,
       so if you have something like:<br>
<br>
		a) PATH=$PATH:/usr/local/genome/bin:/sw/bin<br>
<br>
       make it look like:<br>
<br>
		b) PATH=$PATH:/usr/local/genome/bin:/sw/bin:~/bin<br>
<br>
	4. Quit and restart terminal for the changes to take
       effect.<br>
<br>
If you don't add the location to your $PATH<br>
===========================================<br>
<br>
Assuming you don't want to modify your $PATH, you can still
run the binaries.  Let's assume you put the files in
'~/bin/.  You may run primer3_core by doing either of the
following:<br>
<br>
	1. ~/bin/primer3_core &lt; yourInputFile<br>
	2. /Users/&lt;your username&gt;/bin/primer3_core &lt; yourInputFile<br>
<br>
The first option is just a shortcut to the second.</p>
<h2><a id="installWindows">9. INSTALLATION INSTRUCTIONS - WINDOWS</a></h2>
<p>Functionality warning<br>
=====================<br>
<br>
The windows version does not support the masker functionality.
</p>
<p>How to install this software<br>
============================<br>
<br>
Get your copy of the source code<br>
--------------------------------<br>
1. Install Git for Windows from<br>
   https://git-scm.com/download/win<br>
2. Open Git for Windows and run<br>
   git clone https://github.com/primer3-org/primer3.git<br>
   cd primer3<br>
3. Use git commands to obtain tagged version<br>
   git tag<br>
   git checkout tags/v2.3.7<br>
or
1. Download and unpack latest from<br>
   https://github.com/primer3-org/primer3/releases<br>
Compile Primer3<br>
---------------<br>
<br>
1. Download and install TDM-GCC MinGW Compiler from<br>
   https://sourceforge.net/projects/tdm-gcc/
2. Open the windows command line and change to the src folder in
   Primer3<br>
3. Run: mingw32-make TESTOPTS=--windows<br>
<br>
Running the tests<br>
=================<br>
You must install a perl distribution to run the
windows tests.<br>
<br>
We *strongly* recommend you install ActiveState perl
(http://www.activestate.com/products/activeperl/)
as this was used to test our Primer3 builds, and it is
known to work.<br>
<br>
1. Click on 'Start > Run...'<br>
2. Type 'cmd' into the space provided<br>
3. Hit enter (or select 'OK')<br>
4. Navigate to the location of the tests:<br>
<br>
    A. if you put it in <br>
    C:/Documents and Settings/YourName/primer3-&lt;release&gt;/test/,<br>
    you would type <br>
    'cd c:/Documents and Settings/YourName/primer3-&lt;release&gt;/test/'<br>
    <br>
    B. you can also type 'cd ' (don't forget the space
    after cd) and drag the primer3-&lt;release&gt; folder onto the
    command-line window from windows explorer, this will
    fill in the location for you<br>
<br>
5. On the command line, run 'mingw32-make TESTOPTS=--windows' in this
   directory. You can also run the tests individually,
   for example 'perl p3test.pl --windows'.<br>
6. You should see [OK] for the majority of the tests. The masker
   functionality is not supported on Windows and coresponding tests
   will fail.<br>
<br>
Running the software<br>
====================<br>
<br>
To run the program, you must do so from the MS-DOS
command-line.  The intricacies of the DOS command line are
beyond the scope of this document.  Google for more
information, if needed.  Here is a quick summary:<br>
<br>
1. Click on 'Start > Run...'<br>
2. Type 'cmd' into the space provided<br>
3. Hit enter (or select 'OK')<br>
4. Navigate to the location of the binary:<br>
<br>
    A. if you put it in<br>
    C:/Documents and Settings/YourName/Temp,<br>
    you would type <br>
    'cd c:/Documents and Settings/YourName/Temp'<br>
<br>
    B. you can also type 'cd ' (don't forget the space
    after cd) and drag the Primer3 folder onto the
    command-line window from windows explorer, this will
    fill in the location for you<br>
<br>
5. Run the example file by typing:<br>
<br>
    primer3_core.exe &lt; example<br>
<br>
Other files may be run in a similar fashion.  If your input
filename is 'MyData.txt' you can run Primer3 using this
file (in the correct format; see README) with:<br>
<br>
    primer3_core.exe &lt; MyData.txt<br>
<br>
If your file is not in the folder containing
primer3_core.exe, you could run the program from the
primer3_core folder using:<br>
<br>
    primer3_core.exe &lt; c:/someOtherFolder/someOtherFolder/MyData.txt<br>
<br>
Finally, if you want to run the program without going to
its folder, assuming primer3_core.exe is in c:/Temp, you
could run:<br>
<br>
    c:/Temp/primer3_core.exe &lt; c:/someOtherFolder/someOtherFolder/MyData.txt</p>
<h2><a id="systemRequirements">10. SYSTEM REQUIREMENTS</a></h2>
<p>Please see <a href="https://github.com/primer3-org/primer3">https://github.com/primer3-org/primer3</a> for up-to-date
information.  Primer3 should compile on any system using with an
ANSI C compiler like Linux/Unix, MacOS 10 or MS Windows.  The
Makefile will probably need to be modified for compilation with
C compilers other than gcc.
</p>
<h2><a id="invokingPrimer3">11. INVOKING primer3_core</a></h2>
<p>By default, the executable program produced by the Makefile is
called primer3_core.  This is the C program that does the heavy
lifting of primer picking.   There are also two, more user-friendly,
web interfaces (distributed separately).<br>
<br>
The command line for Primer3 is:<br>
<br>
primer3_core [ --format_output ] [ --default_version=1|--default_version=2 ] [ --io_version=4 ] [ --p3_settings_file=&lt;file_path&gt; ] [ --echo_settings_file ] [ --strict_tags ] [ --output=&lt;file_path&gt; ] [ --error=&lt;file_path&gt; ] [ input_file ] <br>
<br>
A complete list of valid command line arguments can be found in
COMMAND LINE ARGUMENTS below.<br>
<br>
If no input file is specified, primer3_core will read its input
from stdin.</p>
<h2><a id="commandLineTags">12. COMMAND LINE ARGUMENTS</a></h2>
<p>This section presents the list of arguments that may given in
the command line. Any unique abbreviation of the arguments is allowed
(e.g. -ab instead of -about) and each argument can be preceded by
either one or two dashes. In the case of arguments that receive
values, the '=' can be replaced
by a space.</p>
<h3>--about</h3>
<p>With this argument Primer3 generates one line of output
indicating the release number and then exits. This allows scripts to
query Primer3 for its version.</p>
<h3>--default_version=n</h3>
<p><i>n</i>=2 is the default, and directs primer3_core
to use the latest default values (the ones documented here).
<i>n</i>=1 directs primer3_core to use defaults from version
2.2.3 and before.
</p>
<h3>--format_output</h3>
<p>This argument indicates that primer3_core should generate
user-oriented (rather than program-oriented) output.</p>
<h3>--strict_tags</h3>
<p>This argument indicates that primer3_core should generate
a fatal error if there is any tag in the input that
it does not recognize.  This tag also applies to the settings file (see
documentation for the --p3_settings_file argument),
if any, but with limitations:  lines that do not begin with
PRIMER_ or P3_FILE_ID are always ignored in the settings file.</p>
<h3>--p3_settings_file=file_path</h3>
<p>This argument specifies a settings file,
<i>file_path</i>. Primer3 reads the global
("PRIMER_...") parameters from this file first.
Tags appearing in the settings file override
default Primer3 settings. The values set in
the settings files can be also overridden
by tags in the input file. See
Primer3 file documentation for details on the file
format.  <strong>WARNING: the --strict_tags flag applies only to
tags beginning with the string "PRIMER_"; lines that do
not begin with the string "PRIMER_" are ignored.</strong></p>
<h3>--echo_settings_file</h3>
<p>This argument indicates that primer3_core
should print the input tags specified
in the given settings file. If no settings file is
provided or if the --format_output option is given,
this argument will have no effect.</p>
<h3>--io_version=n</h3>
<p>This argument is provided for backward compatibility.
--io_version=4 <strong>is the only legal value and the default
</strong>.</p>
<h3>--output=file_path</h3>
<p>This argument specifies the file where the output
should be written. If omitted, stdout is used. </p>
<h3>--error=file_path</h3>
<p>This argument specifies the file where the error
messages should be written. If omitted, stderr is used.</p>
<h2><a id="inputOutputConventions">13. INPUT AND OUTPUT CONVENTIONS</a></h2>
<p>By default, Primer3 accepts input in Boulder-IO format, a
pre-XML, pre-RDF, text-based input/output format for
program-to-program data interchange.  By default, Primer3 also
produces output in the same format.<br>
<br>
When run with the --format_output command-line flag, Primer3
prints a more user-oriented report for each sequence.<br>
<br>
Primer3 exits with 0 status if it operates correctly.  See EXIT
STATUS CODES below for additional information.<br>
<br>
The syntax of the version of Boulder-IO recognized by Primer3 is
as follows:</p>
<pre>
  o Input consists of a sequence of RECORDs.
  o A RECORD consists of a sequence of (TAG,VALUE) pairs, each terminated
    by a newline character (\n). A RECORD is terminated by  '='
    appearing by itself on a line.
  o A (TAG,VALUE) pair has the following requirements:
    o the TAG must be immediately (without spaces)
          followed by '='.
</pre>
<p>An example of a legal (TAG,VALUE) pair is<br>
<br>
<a href="#SEQUENCE_ID">SEQUENCE_ID</a>=my_marker<br>
<br>
and an example of a Boulder-IO record is<br>
<br>
<a href="#SEQUENCE_ID">SEQUENCE_ID</a>=test1<br>
<a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>=GACTGATCGATGCTAGCTACGATCGATCGATGCATGCTAGCTAGCTAGCTGCTAGC<br>
=<br>
<br>
Many records can be sent, one after another. Below is an example
of three different records which might be passed through a
Boulder-IO stream:<br>
<br>
<a href="#SEQUENCE_ID">SEQUENCE_ID</a>=test1<br>
<a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>=GACTGATCGATGCTAGCTACGATCGATCGATGCATGCTAGCTAGCTAGCTGCTAGC<br>
=<br>
<a href="#SEQUENCE_ID">SEQUENCE_ID</a>=test2<br>
<a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>=CATCATCATCATCGATGCTAGCATCNNACGTACGANCANATGCATCGATCGT<br>
=<br>
<a href="#SEQUENCE_ID">SEQUENCE_ID</a>=test3<br>
<a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>=NACGTAGCTAGCATGCACNACTCGACNACGATGCACNACAGCTGCATCGATGC<br>
=<br>
<br>
Primer3 reads Boulder-IO on stdin and echos its input and returns
results in Boulder-IO format on stdout.  Primer3 indicates many
user-correctable errors by a value in the output tag <a href="#PRIMER_ERROR">PRIMER_ERROR</a>
(see below).  Primer3 indicates other errors, including system configuration
errors, resource errors (such out-of-memory errors), and detected
programming errors by a message on stderr and a non-zero exit
status.<br>
<br>
This document lists the input tags that Primer3 recognizes.
Primer3 echos and ignores any tags it does not recognize, unless
the --strict_tags flag is set on the command line, in which case
Primer3 prints an error in the <a href="#PRIMER_ERROR">PRIMER_ERROR</a> output tag (see
below), and prints additional information on stdout; this option
can be useful for debugging systems that incorporate Primer3.<br>
<br>
Except for tags with the type "interval list" each tag should only
appear ONCE in any given input record.  This restriction is not
systematically checked, and if a tag appears more than once, the
new value silently over-writes the previous value.
</p>
<h2><a id="example">14. AN EXAMPLE</a></h2>
<p>One might be interested in performing PCR on an STS with a CA
repeat in the middle of it. Primers need to be chosen based on
the criteria of the experiment.<br>
<br>
We need to create a Boulder-IO record to send to Primer3 via
stdin. There are lots of ways to accomplish this. We could save
the record into a text file called 'example', and then type the
Unix command 'primer3_core &lt; example'.<br>
<br>
Let's look at the input record itself:
<br>
<a href="#SEQUENCE_ID">SEQUENCE_ID</a>=example<br>
<a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>=GTAGTCAGTAGACNATGACNACTGACGATGCAGACNACACACACACACACAGCACACAGGTATTAGTGGGCCATTCGATCCCGACCCAAATCGATAGCTACGATGACG<br>
<a href="#SEQUENCE_TARGET">SEQUENCE_TARGET</a>=37,21<br>
<a href="#PRIMER_TASK">PRIMER_TASK</a>=generic<br>
<a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>=1<br>
<a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>=1<br>
<a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>=1<br>
<a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a>=18<br>
<a href="#PRIMER_MIN_SIZE">PRIMER_MIN_SIZE</a>=15<br>
<a href="#PRIMER_MAX_SIZE">PRIMER_MAX_SIZE</a>=21<br>
<a href="#PRIMER_MAX_NS_ACCEPTED">PRIMER_MAX_NS_ACCEPTED</a>=1<br>
<a href="#PRIMER_PRODUCT_SIZE_RANGE">PRIMER_PRODUCT_SIZE_RANGE</a>=75-100<br>
<a href="#P3_FILE_FLAG">P3_FILE_FLAG</a>=1<br>
<a href="#SEQUENCE_INTERNAL_EXCLUDED_REGION">SEQUENCE_INTERNAL_EXCLUDED_REGION</a>=37,21<br>
<a href="#PRIMER_EXPLAIN_FLAG">PRIMER_EXPLAIN_FLAG</a>=1<br>
=<br>
<br>
A breakdown of the reasoning behind each of the TAG=VALUE pairs
is below:<br>
<br>
<a href="#SEQUENCE_ID">SEQUENCE_ID</a>=example<br>
<br>
The main intent of this tag is to provide an identifier for the
sequence that is meaningful to the user, for example when Primer3
processes multiple records, and by default this tag is optional.
However, this tag is _required_ when <a href="#P3_FILE_FLAG">P3_FILE_FLAG</a> is non-0
Because it provides the names of the files that contain lists
of oligos that Primer3 considered.<br>
<br>
<a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>=GTAGTCAGTAGACNATGACNACTGACGATGCAGACNACACACACACACACAGCACACAGGTATTAGTGGGCCATTCGATCCCGACCCAAATCGATAGCTACGATGACG<br>
<br>
The <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a> provides the sequence
from which Primer3 will design primers.
Note that there is no newline until the sequence terminates completely.<br>
<br>
<a href="#SEQUENCE_TARGET">SEQUENCE_TARGET</a>=37,21<br>
<br>
There is a simple sequence repeat in this sequence.
The repeat starts
at base 37, and has a length of 21 bases. We instruct Primer3 to
choose primers that flank the repeat site (because
we want to use the PCR product for
determining the length of the repeat, which is likely to be polymorphic).
<br>
<br>
<a href="#PRIMER_TASK">PRIMER_TASK</a>=generic<br>
<br>
The <a href="#PRIMER_TASK">PRIMER_TASK</a> tells Primer3 which
type of primers to pick. You can select typical primers for PCR
detection, primers for cloning or for sequencing.<br>
<br>
<a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>=1<br>
<a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>=1<br>
<a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>=1<br>
<br>
We would like to pick a left primer, an internal oligo and a
right primer, so we set these flags to 1 (true). In combination
with the <a href="#PRIMER_TASK">PRIMER_TASK</a> this tags control which primers are
picked.<br>
<br>
<a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a>=18<br>
<br>
Since our sequence length is rather small (only 108 bases long),
we lower the <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a> from 20 to 18. It is
more likely that Primer3 will succeed if it aims for smaller
primers.<br>
<br>
<a href="#PRIMER_MIN_SIZE">PRIMER_MIN_SIZE</a>=15<br>
<a href="#PRIMER_MAX_SIZE">PRIMER_MAX_SIZE</a>=21<br>
<br>
With the lowering of optimal primer size, it's good to lower
the minimum and maximum sizes as well.<br>
<br>
<a href="#PRIMER_MAX_NS_ACCEPTED">PRIMER_MAX_NS_ACCEPTED</a>=1<br>
<br>
Since the sequence is short with a
non-negligible amount of unknown bases (N's) in it, we
make Primer3's job easier by allowing it to pick primers that have
at most 1 unknown base.<br>
<br>
<a href="#PRIMER_PRODUCT_SIZE_RANGE">PRIMER_PRODUCT_SIZE_RANGE</a>=75-100<br>
<br>
We reduce the product size range from the default of 100-300
because our source sequence is only 108 base pairs long.  If we
insisted on a product size of 100 base pairs Primer3 would have
few possibilities to choose from.<br>
<br>
<a href="#P3_FILE_FLAG">P3_FILE_FLAG</a>=1<br>
<br>
Because we have a short sequence with many Ns and
with a simple sequence repeat that we must amplify,
Primer3 might fail to
pick and primers. Therefore, we want to get the list of primers it
considered, so that we could manually pick primers
ourselves if Primer3 fails to do so. Setting this flag to 1
asks Primer3 to output the primers it considered to a
forwar primers and reverse primers to output files.<br>
<br>
<a href="#SEQUENCE_INTERNAL_EXCLUDED_REGION">SEQUENCE_INTERNAL_EXCLUDED_REGION</a>=37,21<br>
<br>
Normally CA-repeats make poor hybridization probes (because they
not specific enough).  Therefore, we exclude the CA repeat (which
is the TARGET) from consideration for the internal oligo.<br>
<br>
<a href="#PRIMER_EXPLAIN_FLAG">PRIMER_EXPLAIN_FLAG</a>=1<br>
<br>
We want to see statistics about the oligos and oligo triples
(left primer, internal oligo, right primer) that Primer3
examined.<br>
<br>
=<br>
<br>
The '=' character terminates the record.<br>
<br>
There were many Boulder-IO input tags that were not
specified,
which is legal.  For the tags with default
values, those defaults will be used in the analysis. For the
tags with NO default values (for example <a href="#SEQUENCE_TARGET">SEQUENCE_TARGET</a>), the
functionality requested by the those tags will be absent.
Also, please note
that it is not the case that we need to surround a simple sequence
repeat every time we want to pick primers!</p>
<h2><a id="migrateTags">15. HOW TO MIGRATE TAGS TO IO VERSION 4</a></h2>
<p>With Primer3 release 2.0, many Boulder-IO tags were
modified and new tags were introduced. The new Primer3 tags are
designed with the idea in mind that computer scripts and other
programs use primer3_core. The modifications make it easier for
programs to read and write Primer3 input and output.<br>
<br>
Furthermore the Primer3 default values and the use of
<a href="#PRIMER_WT_TEMPLATE_MISPRIMING">PRIMER_WT_TEMPLATE_MISPRIMING</a> and <a href="#PRIMER_PAIR_WT_TEMPLATE_MISPRIMING">PRIMER_PAIR_WT_TEMPLATE_MISPRIMING</a>
have changed in version 2.0.
<br>
<br>
There are three classes of input: "Sequence" input tags describe
a particular input sequence to Primer3, and are reset after
every Boulder record (now starting with SEQUENCE_).  "Global"
input tags describe the general parameters that Primer3 should
use in its searches, and the values of these tags persist
between input Boulder records until or unless they are
explicitly reset (now starting with PRIMER_). "Program"
parameters that deal with the behavior of the Primer3 program
itself (now starting with P3_). See below for a list of the
modified tags.<br>
<br>
The handling of PRIMER_TASK changed completely. In the past we
used it to tell Primer3 what task to perform. Now the task is
complemented with <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>,
<a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a> and <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>, which
specify which primers are to be picked.<br>
<br>
These Tags are modified:<br>
<br>
The "per sequence" tags:</p>
<pre>
NEW VERSION                       - OLD VERSION
----------------------------------------------------------------------------------------
<a href="#SEQUENCE_ID">SEQUENCE_ID</a>                       - PRIMER_SEQUENCE_ID
<a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>                 - SEQUENCE
<a href="#SEQUENCE_QUALITY">SEQUENCE_QUALITY</a>                  - PRIMER_SEQUENCE_QUALITY
<a href="#SEQUENCE_INCLUDED_REGION">SEQUENCE_INCLUDED_REGION</a>          - INCLUDED_REGION
<a href="#SEQUENCE_TARGET">SEQUENCE_TARGET</a>                   - TARGET
<a href="#SEQUENCE_EXCLUDED_REGION">SEQUENCE_EXCLUDED_REGION</a>          - EXCLUDED_REGION
<a href="#SEQUENCE_START_CODON_POSITION">SEQUENCE_START_CODON_POSITION</a>     - PRIMER_START_CODON_POSITION
<a href="#SEQUENCE_PRIMER">SEQUENCE_PRIMER</a>                   - PRIMER_LEFT_INPUT
<a href="#SEQUENCE_PRIMER_REVCOMP">SEQUENCE_PRIMER_REVCOMP</a>           - PRIMER_RIGHT_INPUT
<a href="#SEQUENCE_INTERNAL_OLIGO">SEQUENCE_INTERNAL_OLIGO</a>           - PRIMER_INTERNAL_OLIGO_INPUT
<a href="#SEQUENCE_INTERNAL_EXCLUDED_REGION">SEQUENCE_INTERNAL_EXCLUDED_REGION</a> - PRIMER_INTERNAL_OLIGO_EXCLUDED_REGION
--------------------------------------------------------------------------------
The "global" tags:
NEW VERSION                       - OLD VERSION
<a href="#PRIMER_TASK">PRIMER_TASK</a>                       - PRIMER_TASK (modified use)
<a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>          - --- did not exist
<a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>        - PRIMER_PICK_INTERNAL_OLIGO (modified use)
<a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>           - --- did not exist
<a href="#PRIMER_PAIR_WT_TEMPLATE_MISPRIMING">PRIMER_PAIR_WT_TEMPLATE_MISPRIMING</a>- PRIMER_PAIR_WT_TEMPLATE_MISPRIMING (modified use)
<a href="#PRIMER_WT_TEMPLATE_MISPRIMING">PRIMER_WT_TEMPLATE_MISPRIMING</a>     - PRIMER_WT_TEMPLATE_MISPRIMING (modified use)
<a href="#PRIMER_MAX_LIBRARY_MISPRIMING">PRIMER_MAX_LIBRARY_MISPRIMING</a>     - PRIMER_MAX_MISPRIMING
<a href="#PRIMER_INTERNAL_MAX_LIBRARY_MISHYB">PRIMER_INTERNAL_MAX_LIBRARY_MISHYB</a>- PRIMER_INTERNAL_OLIGO_MAX_MISHYB
<a href="#PRIMER_PAIR_MAX_LIBRARY_MISPRIMING">PRIMER_PAIR_MAX_LIBRARY_MISPRIMING</a>- PRIMER_PAIR_MAX_MISPRIMING
<a href="#PRIMER_WT_LIBRARY_MISPRIMING">PRIMER_WT_LIBRARY_MISPRIMING</a>      - PRIMER_WT_REP_SIM
<a href="#PRIMER_INTERNAL_WT_LIBRARY_MISHYB">PRIMER_INTERNAL_WT_LIBRARY_MISHYB</a> - PRIMER_INTERNAL_WT_REP_SIM
<a href="#PRIMER_PAIR_WT_LIBRARY_MISPRIMING">PRIMER_PAIR_WT_LIBRARY_MISPRIMING</a> - PRIMER_PAIR_WT_REP_SIM
<a href="#PRIMER_MAX_NS_ACCEPTED">PRIMER_MAX_NS_ACCEPTED</a>            - PRIMER_NUM_NS_ACCEPTED
<a href="#PRIMER_PAIR_MAX_DIFF_TM">PRIMER_PAIR_MAX_DIFF_TM</a>           - PRIMER_MAX_DIFF_TM
<a href="#PRIMER_SALT_MONOVALENT">PRIMER_SALT_MONOVALENT</a>            - PRIMER_SALT_CONC
<a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a>              - PRIMER_DIVALENT_CONC
<a href="#PRIMER_TM_FORMULA">PRIMER_TM_FORMULA</a>                 - PRIMER_TM_SANTALUCIA
<a href="#PRIMER_MAX_SELF_ANY">PRIMER_MAX_SELF_ANY</a>               - PRIMER_SELF_ANY
<a href="#PRIMER_MAX_SELF_END">PRIMER_MAX_SELF_END</a>               - PRIMER_SELF_END
<a href="#PRIMER_WT_SELF_ANY">PRIMER_WT_SELF_ANY</a>                - PRIMER_WT_COMPL_ANY
<a href="#PRIMER_WT_SELF_END">PRIMER_WT_SELF_END</a>                - PRIMER_WT_COMPL_END
<a href="#PRIMER_PAIR_MAX_COMPL_ANY">PRIMER_PAIR_MAX_COMPL_ANY</a>         - PRIMER_PAIR_ANY
<a href="#PRIMER_PAIR_MAX_COMPL_END">PRIMER_PAIR_MAX_COMPL_END</a>         - PRIMER_PAIR_END
<a href="#P3_FILE_FLAG">P3_FILE_FLAG</a>                      - PRIMER_FILE_FLAG
<a href="#P3_COMMENT">P3_COMMENT</a>                        - PRIMER_COMMENT
<a href="#PRIMER_INTERNAL_SALT_MONOVALENT">PRIMER_INTERNAL_SALT_MONOVALENT</a>   - PRIMER_INTERNAL_OLIGO_SALT_CONC
<a href="#PRIMER_INTERNAL_SALT_DIVALENT">PRIMER_INTERNAL_SALT_DIVALENT</a>     - PRIMER_INTERNAL_OLIGO_DIVALENT_CONC
<a href="#PRIMER_INTERNAL_WT_SELF_ANY">PRIMER_INTERNAL_WT_SELF_ANY</a>       - PRIMER_IO_WT_COMPL_ANY
<a href="#PRIMER_INTERNAL_WT_SELF_END">PRIMER_INTERNAL_WT_SELF_END</a>       - PRIMER_IO_WT_COMPL_END
<a href="#PRIMER_INTERNAL_MAX_NS_ACCEPTED">PRIMER_INTERNAL_MAX_NS_ACCEPTED</a>   - PRIMER_INTERNAL_OLIGO_NUM_NS
<a href="#PRIMER_INTERNAL_MAX_SELF_ANY">PRIMER_INTERNAL_MAX_SELF_ANY</a>      - PRIMER_INTERNAL_OLIGO_SELF_ANY
<a href="#PRIMER_INTERNAL_MAX_SELF_END">PRIMER_INTERNAL_MAX_SELF_END</a>      - PRIMER_INTERNAL_OLIGO_SELF_END
The following tags INTERNAL_OLIGO is replaced by INTERNAL:
<a href="#PRIMER_INTERNAL_OPT_SIZE">PRIMER_INTERNAL_OPT_SIZE</a>          - PRIMER_INTERNAL_OLIGO_OPT_SIZE
<a href="#PRIMER_INTERNAL_MIN_SIZE">PRIMER_INTERNAL_MIN_SIZE</a>          - PRIMER_INTERNAL_OLIGO_MIN_SIZE
<a href="#PRIMER_INTERNAL_MAX_SIZE">PRIMER_INTERNAL_MAX_SIZE</a>          - PRIMER_INTERNAL_OLIGO_MAX_SIZE
<a href="#PRIMER_INTERNAL_OPT_TM">PRIMER_INTERNAL_OPT_TM</a>            - PRIMER_INTERNAL_OLIGO_OPT_TM
<a href="#PRIMER_INTERNAL_MIN_TM">PRIMER_INTERNAL_MIN_TM</a>            - PRIMER_INTERNAL_OLIGO_MIN_TM
<a href="#PRIMER_INTERNAL_MAX_TM">PRIMER_INTERNAL_MAX_TM</a>            - PRIMER_INTERNAL_OLIGO_MAX_TM
<a href="#PRIMER_INTERNAL_MIN_GC">PRIMER_INTERNAL_MIN_GC</a>            - PRIMER_INTERNAL_OLIGO_MIN_GC
<a href="#PRIMER_INTERNAL_OPT_GC_PERCENT">PRIMER_INTERNAL_OPT_GC_PERCENT</a>    - PRIMER_INTERNAL_OLIGO_OPT_GC_PERCENT
<a href="#PRIMER_INTERNAL_MAX_GC">PRIMER_INTERNAL_MAX_GC</a>            - PRIMER_INTERNAL_OLIGO_MAX_GC
<a href="#PRIMER_INTERNAL_DNTP_CONC">PRIMER_INTERNAL_DNTP_CONC</a>         - PRIMER_INTERNAL_OLIGO_DNTP_CONC
<a href="#PRIMER_INTERNAL_DNA_CONC">PRIMER_INTERNAL_DNA_CONC</a>          - PRIMER_INTERNAL_OLIGO_DNA_CONC
<a href="#PRIMER_INTERNAL_MAX_POLY_X">PRIMER_INTERNAL_MAX_POLY_X</a>        - PRIMER_INTERNAL_OLIGO_MAX_POLY_X
<a href="#PRIMER_INTERNAL_MISHYB_LIBRARY">PRIMER_INTERNAL_MISHYB_LIBRARY</a>    - PRIMER_INTERNAL_OLIGO_MISHYB_LIBRARY
<a href="#PRIMER_INTERNAL_MIN_QUALITY">PRIMER_INTERNAL_MIN_QUALITY</a>       - PRIMER_INTERNAL_OLIGO_MIN_QUALITY
The following tags IO is replaced by INTERNAL:
<a href="#PRIMER_INTERNAL_WT_TM_GT">PRIMER_INTERNAL_WT_TM_GT</a>          - PRIMER_IO_WT_TM_GT
<a href="#PRIMER_INTERNAL_WT_TM_LT">PRIMER_INTERNAL_WT_TM_LT</a>          - PRIMER_IO_WT_TM_LT
<a href="#PRIMER_INTERNAL_WT_SIZE_LT">PRIMER_INTERNAL_WT_SIZE_LT</a>        - PRIMER_IO_WT_SIZE_LT
<a href="#PRIMER_INTERNAL_WT_SIZE_GT">PRIMER_INTERNAL_WT_SIZE_GT</a>        - PRIMER_IO_WT_SIZE_GT
<a href="#PRIMER_INTERNAL_WT_GC_PERCENT_LT">PRIMER_INTERNAL_WT_GC_PERCENT_LT</a>  - PRIMER_IO_WT_GC_PERCENT_LT
<a href="#PRIMER_INTERNAL_WT_GC_PERCENT_GT">PRIMER_INTERNAL_WT_GC_PERCENT_GT</a>  - PRIMER_IO_WT_GC_PERCENT_GT
<a href="#PRIMER_INTERNAL_WT_NUM_NS">PRIMER_INTERNAL_WT_NUM_NS</a>         - PRIMER_IO_WT_NUM_NS
<a href="#PRIMER_INTERNAL_WT_SEQ_QUAL">PRIMER_INTERNAL_WT_SEQ_QUAL</a>       - PRIMER_IO_WT_SEQ_QUAL
</pre>
<p>--------------------------------------------------------------------------------<br>
OUTPUT TAGS:<br>
<br>
There are three big changes on the output:<br>
- INTERNAL_OLIGO is now replaced by INTERNAL.<br>
- The first version is numbered 0.<br>
- The "PRODUCT" tags are renamed<br>
- The errors are modified<br>
- Errors caused by a specific primer are given as <a href="#PRIMER_LEFT_4_PROBLEMS">PRIMER_LEFT_4_PROBLEMS</a><br>
<br>
Now all primer related output follows the rule:
PRIMER_{LEFT,RIGHT,INTERNAL,PAIR}_&lt;j&gt;_&lt;tag_name&gt;.
where &lt;j&gt; is an integer from 0 to n, where n is at most the
value of <a href="#PRIMER_NUM_RETURN">PRIMER_NUM_RETURN</a> - 1.<br>
<br>
This allows easy scripting by using the underscores _ to split
the name. The first part is PRIMER, the second the type of oligo
or pair parameters, the third is always a number, starting at
0 and the rest is used by the tags.<br>
<br>
That affects also (shown for output number 4):</p>
<pre>
NEW VERSION                             - OLD VERSION
------------------------------------------------------------------------------------------------------
<a href="#PRIMER_PAIR_4_PENALTY">PRIMER_PAIR_4_PENALTY</a>                   - PRIMER_PAIR_PENALTY_4 (number moved behind PAIR)
<a href="#PRIMER_PAIR_4_PRODUCT_SIZE">PRIMER_PAIR_4_PRODUCT_SIZE</a>              - PRIMER_PRODUCT_SIZE_4 (grouped with PAIR)
<a href="#PRIMER_PAIR_4_PRODUCT_TM">PRIMER_PAIR_4_PRODUCT_TM</a>                - PRIMER_PRODUCT_TM_4 (grouped with PAIR)
<a href="#PRIMER_PAIR_4_PRODUCT_TM_OLIGO_TM_DIFF">PRIMER_PAIR_4_PRODUCT_TM_OLIGO_TM_DIFF</a>  - PRIMER_PRODUCT_TM_OLIGO_TM_DIFF_4 (grouped with PAIR)
<a href="#PRIMER_INTERNAL_EXPLAIN">PRIMER_INTERNAL_EXPLAIN</a>                 - PRIMER_INTERNAL_OLIGO_EXPLAIN
PRIMER_LEFT_4_LIBRARY_MISPRIMING                   - PRIMER_LEFT_4_MISPRIMING_SCORE
PRIMER_INTERNAL_4_LIBRARY_MISHYB                   - PRIMER_INTERNAL_OLIGO_4_MISHYB_SCORE
PRIMER_RIGHT_4_LIBRARY_MISPRIMING                  - PRIMER_RIGHT_4_MISPRIMING_SCORE
PRIMER_PAIR_4_LIBRARY_MISPRIMING                   - PRIMER_PAIR_4_MISPRIMING_SCORE
</pre>
<h2><a id="sequenceTags">16. "SEQUENCE" INPUT TAGS</a></h2>
<p>"Sequence" input tags start with SEQUENCE_... and describe a
particular input sequence to Primer3. They are reset after every
Boulder record. Errors in "Sequence" input tags invalidate the
current record, but Primer3 will continue to process additional
records.</p>
<table class="p3p_tab_help_table" style="text-align: left; width: 800px; border: 1px;">
	  <tr>
	    <td><a href="#SEQUENCE_EXCLUDED_REGION">SEQUENCE_EXCLUDED_REGION</a></td>
	    <td><a href="#SEQUENCE_INCLUDED_REGION">SEQUENCE_INCLUDED_REGION</a></td>
	    <td><a href="#SEQUENCE_PRIMER_REVCOMP">SEQUENCE_PRIMER_REVCOMP</a></td>
	  </tr>
	  <tr>
	    <td><a href="#SEQUENCE_FORCE_LEFT_END">SEQUENCE_FORCE_LEFT_END</a></td>
	    <td><a href="#SEQUENCE_INTERNAL_EXCLUDED_REGION">SEQUENCE_INTERNAL_EXCLUDED_REGION</a></td>
	    <td><a href="#SEQUENCE_QUALITY">SEQUENCE_QUALITY</a></td>
	  </tr>
	  <tr>
	    <td><a href="#SEQUENCE_FORCE_LEFT_START">SEQUENCE_FORCE_LEFT_START</a></td>
	    <td><a href="#SEQUENCE_INTERNAL_OLIGO">SEQUENCE_INTERNAL_OLIGO</a></td>
	    <td><a href="#SEQUENCE_START_CODON_POSITION">SEQUENCE_START_CODON_POSITION</a></td>
	  </tr>
	  <tr>
	    <td><a href="#SEQUENCE_FORCE_RIGHT_END">SEQUENCE_FORCE_RIGHT_END</a></td>
	    <td><a href="#SEQUENCE_OVERLAP_JUNCTION_LIST">SEQUENCE_OVERLAP_JUNCTION_LIST</a></td>
	    <td><a href="#SEQUENCE_TARGET">SEQUENCE_TARGET</a></td>
	  </tr>
	  <tr>
	    <td><a href="#SEQUENCE_FORCE_RIGHT_START">SEQUENCE_FORCE_RIGHT_START</a></td>
	    <td><a href="#SEQUENCE_PRIMER">SEQUENCE_PRIMER</a></td>
	    <td><a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a></td>
	  </tr>
	  <tr>
	    <td><a href="#SEQUENCE_ID">SEQUENCE_ID</a></td>
	    <td><a href="#SEQUENCE_PRIMER_PAIR_OK_REGION_LIST">SEQUENCE_PRIMER_PAIR_OK_REGION_LIST</a></td>
	    <td>&nbsp;</td>
	  </tr>
</table>
<h3><a id="SEQUENCE_ID">SEQUENCE_ID (string; default empty)</a></h3>
<p>Short description of the sequence. It is used as an identifier
that is reproduced in the output to enable users to identify the
source of the chosen primers.<br>
<br>
This tag must be present if <a href="#P3_FILE_FLAG">P3_FILE_FLAG</a> is non-zero.</p>
<h3><a id="SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE (nucleotide sequence; default empty)</a></h3>
<p>The sequence from which to choose primers.  The sequence
must be presented 5' -> 3' (i.e, in the normal way).
In general, the bases may be upper or lower case, but
lower case letters are treated
specially if <a href="#PRIMER_LOWERCASE_MASKING">PRIMER_LOWERCASE_MASKING</a> is set.
The entire sequence MUST be all on a single line.
(In other words, the sequence cannot span several lines.)
</p>
<h3><a id="SEQUENCE_INCLUDED_REGION">SEQUENCE_INCLUDED_REGION (interval list; default empty)</a></h3>
<p>A sub-region of the given sequence in which to pick primers.  For
example, often the first dozen or so bases of a sequence are
vector, and should be excluded from consideration. The value for
this parameter has the form</p>
<pre>
<i>&lt;start&gt;</i>,<i>&lt;length&gt;</i>
</pre>
<p>where <a class="p3_prop"><i>&lt;start&gt;</i></a> is the index of the first base to consider,
and <a class="p3_prop"><i>&lt;length&gt;</i></a> is the number of subsequent bases in the
primer-picking region.</p>
<h3><a id="SEQUENCE_TARGET">SEQUENCE_TARGET (interval list; default empty)</a></h3>
<p>If one or more targets is specified then a legal primer pair must
flank at least one of them.  A target might be a simple sequence
repeat site (for example a CA repeat) or a single-base-pair
polymorphism, or an exon for resequencing.  The value should be a space-separated list of</p>
<pre>
<i>&lt;start&gt;</i>,<i>&lt;length&gt;</i>
</pre>
<p>pairs where <a class="p3_prop"><i>&lt;start&gt;</i></a> is the index of the first base of a
target, and <a class="p3_prop"><i>&lt;length&gt;</i></a> is its length.</p>
See also PRIMER_INSIDE_PENALTY, PRIMER_OUTSIDE_PENALTY.
Has a different meaning when PRIMER_TASK=pick_sequencing_primers.
See PRIMER_TASK for more information.
<h3><a id="SEQUENCE_EXCLUDED_REGION">SEQUENCE_EXCLUDED_REGION (interval list; default empty)</a></h3>
<p>Primers and oligos may not overlap any region specified in this tag.
The associated value must be a space-separated list of</p>
<pre>
<i>&lt;start&gt;</i>,<i>&lt;length&gt;</i>
</pre>
<p>pairs where <a class="p3_prop"><i>&lt;start&gt;</i></a> is the index of the first base of
the excluded region, and <a class="p3_prop"><i>&lt;length&gt;</i></a> is its length.  This tag is
useful for tasks such as excluding regions of low sequence
quality or for excluding regions containing repetitive elements
such as ALUs or LINEs.</p>
<h3><a id="SEQUENCE_PRIMER_PAIR_OK_REGION_LIST">SEQUENCE_PRIMER_PAIR_OK_REGION_LIST (semicolon separated list of integer quadruples; default empty)</a></h3>
<p>
This tag allows detailed specification of possible
locations of left and right primers in primer pairs.
<br><br>
The associated value must be a semicolon-separated list of</p>
<pre>
<i>&lt;left_start&gt;</i>,<i>&lt;left_length&gt;</i>,<i>&lt;right_start&gt;</i>,<i>&lt;right_length&gt;</i>
</pre>
<p>quadruples.
The left primer must be in the region specified by
<a class="p3_prop"><i>&lt;left_start&gt;</i></a>,<a class="p3_prop"><i>&lt;left_length&gt;</i></a> and
the right primer must be in the region specified by
<a class="p3_prop"><i>&lt;right_start&gt;</i></a>,<a class="p3_prop"><i>&lt;right_length&gt;</i></a>.
<a class="p3_prop"><i>&lt;left_start&gt;</i></a> and <a class="p3_prop"><i>&lt;left_length&gt;</i></a> specify the location of the left primer in terms of the index of
the first base in the region and the length of the region.
<a class="p3_prop"><i>&lt;right_start&gt;</i></a> and <a class="p3_prop"><i>&lt;right_length&gt;</i></a> specify the location of the right primer in analogous fashion.
As seen in the example below, if no integers are specified for a region then the location of the
corresponding primer is not constrained.
<br><br>
Example:</p>
<pre>
SEQUENCE_PRIMER_PAIR_OK_REGION_LIST=100,50,300,50 ; 900,60,, ; ,,930,100
</pre>
<p>
<br>
Specifies that there are three choices:
<br><br>
Left primer in the 50 bp region starting at position 100 AND right primer
in the 50 bp region starting at position 300
<br><br>
OR
<br><br>
Left primer in the 60 bp region starting at position 900 (and right primer anywhere)
<br><br>
OR
<br><br>
Right primer in the 100 bp region starting at position 930 (and left
primer anywhere)
</p>
<h3><a id="SEQUENCE_OVERLAP_JUNCTION_LIST">SEQUENCE_OVERLAP_JUNCTION_LIST (space separated integers; default empty)</a></h3>
<p>If this list is not empty, then
the forward OR the reverse primer must overlap one of
these junction points by at least <a href="#PRIMER_MIN_3_PRIME_OVERLAP_OF_JUNCTION">PRIMER_MIN_3_PRIME_OVERLAP_OF_JUNCTION</a>
nucleotides at the 3' end and at least <a href="#PRIMER_MIN_5_PRIME_OVERLAP_OF_JUNCTION">PRIMER_MIN_5_PRIME_OVERLAP_OF_JUNCTION</a>
nucleotides at the 5' end.
<br><br>
In more detail: The junction associated with a given position is the space immediately to the right
of that position in the template sequence on the strand given as input.
<br><br>
For example:</p>
<pre>
<code>
SEQUENCE_OVERLAP_JUNCTION_LIST=20
# 1-based indexes
PRIMER_MIN_3_PRIME_OVERLAP_OF_JUNCTION=4
template: atcataggccatgcctgagc^gctacgact
          ok           ...gagc^gcta-3'  (left primer)
          not ok       ...gagc^gct-3'   (left primer)
          ok           3'-ctcg^cgat...  (right pimer)
          not ok        3'-tcg^cgat...  (right primer)
PRIMER_MIN_5_PRIME_OVERLAP_OF_JUNCTION=5
         ok           5'-tgagc^gccg...  (left primer)
         not ok        5'-gagc^gccg...  (left primer)
         ok           ...tgagc^gctac-5' (right primer)
         not ok       ...tgagc^gcta-5'  (right primer)
</code>
</pre>
<h3><a id="SEQUENCE_INTERNAL_EXCLUDED_REGION">SEQUENCE_INTERNAL_EXCLUDED_REGION (interval list; default empty)</a></h3>
<p>Middle oligos may not overlap any region specified by this tag.
The associated value must be a space-separated list of</p>
<pre>
<i>&lt;start&gt;</i>,<i>&lt;length&gt;</i>
</pre>
<p>pairs, where &lt;start&gt; is the index of the first base of
an excluded region, and &lt;length&gt; is its length.  Often one would
make Target regions excluded regions for internal oligos.</p>
<h3><a id="SEQUENCE_PRIMER">SEQUENCE_PRIMER (nucleotide sequence; default empty)</a></h3>
<p>The sequence of a left primer to check and around which to design
right primers and optional internal oligos.  Must be a substring
of <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>.</p>
<h3><a id="SEQUENCE_INTERNAL_OLIGO">SEQUENCE_INTERNAL_OLIGO (nucleotide sequence; default empty)</a></h3>
<p>The sequence of an internal oligo to check and around which to
design left and right primers.  Must be a substring of
<a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>.</p>
<h3><a id="SEQUENCE_PRIMER_REVCOMP">SEQUENCE_PRIMER_REVCOMP (nucleotide sequence; default empty)</a></h3>
<p>The sequence of a right primer to check and around which to
design left primers and optional internal oligos.  Must be a
substring of the reverse strand of <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>.</p>
<h3><a id="SEQUENCE_START_CODON_POSITION">SEQUENCE_START_CODON_POSITION (int; default -2000000)</a></h3>
<p>This parameter should be considered EXPERIMENTAL at this point.
Please check the output carefully; some erroneous inputs might
cause an error in Primer3.<br>
<br>
Index of the first base of a start codon.  This parameter allows
Primer3 to select primer pairs to create in-frame amplicons
e.g. to create a template for a fusion protein.  Primer3 will
attempt to select an in-frame left primer, ideally starting at or
to the left of the start codon, or to the right if necessary.
Negative values of this parameter are legal if the actual start
codon is to the left of available sequence. If this parameter is
non-negative Primer3 signals an error if the codon at the
position specified by this parameter is not an ATG.  A value less
than or equal to -10^6 indicates that Primer3 should ignore this
parameter.<br>
<br>
Primer3 selects the position of the right primer by scanning
right from the left primer for a stop codon.  Ideally the right
primer will end at or after the stop codon.</p>
<h3><a id="SEQUENCE_QUALITY">SEQUENCE_QUALITY (space separated integers; default empty)</a></h3>
<p>A list of space separated integers. There must be exactly
one integer for each base in <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a> if this argument is
non-empty.  For example, for the sequence ANNTTCA...
<a href="#SEQUENCE_QUALITY">SEQUENCE_QUALITY</a> might be 45 10 0 50 30 34 50 67 ....
High numbers indicate high confidence in the base called at
that position and low numbers indicate low confidence in the
base call at that position.  This parameter is only relevant
if you are using a base calling program that provides
quality information (for example phred).</p>
<h3><a id="SEQUENCE_FORCE_LEFT_START">SEQUENCE_FORCE_LEFT_START (int; default -1000000)</a></h3>
<p>Forces the 5' end of the left primer to be at the indicated
position. Primers are also picked if they violate certain
constraints. The default value indicates that the start of the left primer
can be anywhere.</p>
<h3><a id="SEQUENCE_FORCE_LEFT_END">SEQUENCE_FORCE_LEFT_END (int; default -1000000)</a></h3>
<p>Forces the 3' end of the left primer to be at the indicated
position. Primers are also picked if they violate certain
constraints. The default value indicates that the end of the left primer
can be anywhere.</p>
<h3><a id="SEQUENCE_FORCE_RIGHT_START">SEQUENCE_FORCE_RIGHT_START (int; default -1000000)</a></h3>
<p>Forces the 5' end of the right primer to be at the indicated
position. Primers are also picked if they violate certain
constraints. The default value indicates that the start of the right primer
can be anywhere.</p>
<h3><a id="SEQUENCE_FORCE_RIGHT_END">SEQUENCE_FORCE_RIGHT_END (int; default -1000000)</a></h3>
<p>Forces the 3' end of the right primer to be at the indicated
position. Primers are also picked if they violate certain
constraints. The default value indicates that the end of the right primer
can be anywhere.</p>
<h2><a id="globalTags">17. "GLOBAL" INPUT TAGS</a></h2>
<p>"Global" input tags start with PRIMER_... and describe the
general parameters that Primer3 should use in its searches.
The values of these tags persist between input Boulder
records until or unless they are explicitly reset. Errors
in "Global" input tags are fatal because they invalidate
the basic conditions under which primers are being picked.<br>
<br>
Because the laboratory detection step using internal oligos is
independent of the PCR amplification procedure, internal oligo
tags have defaults that are independent of the parameters that
govern the selection of PCR primers. For example, the melting
temperature of an oligo used for hybridization might be
considerably lower than that used as a PCR primer.
<br></p>
<table class="p3p_tab_help_table" style="text-align: left; width: 800px; border: 1px;">
	  <tr>
	    <td><a href="#PRIMER_DNA_CONC">PRIMER_DNA_CONC</a></td>
	    <td><a href="#PRIMER_MASK_KMERLIST_PATH">PRIMER_MASK_KMERLIST_PATH</a></td>
	    <td><a href="#PRIMER_PAIR_WT_PRODUCT_SIZE_LT">PRIMER_PAIR_WT_PRODUCT_SIZE_LT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_DNTP_CONC">PRIMER_DNTP_CONC</a></td>
	    <td><a href="#PRIMER_MASK_KMERLIST_PREFIX">PRIMER_MASK_KMERLIST_PREFIX</a></td>
	    <td><a href="#PRIMER_PAIR_WT_PRODUCT_TM_GT">PRIMER_PAIR_WT_PRODUCT_TM_GT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_EXPLAIN_FLAG">PRIMER_EXPLAIN_FLAG</a></td>
	    <td><a href="#PRIMER_MASK_TEMPLATE">PRIMER_MASK_TEMPLATE</a></td>
	    <td><a href="#PRIMER_PAIR_WT_PRODUCT_TM_LT">PRIMER_PAIR_WT_PRODUCT_TM_LT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_FIRST_BASE_INDEX">PRIMER_FIRST_BASE_INDEX</a></td>
	    <td><a href="#PRIMER_MAX_END_GC">PRIMER_MAX_END_GC</a></td>
	    <td><a href="#PRIMER_PAIR_WT_PR_PENALTY">PRIMER_PAIR_WT_PR_PENALTY</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_GC_CLAMP">PRIMER_GC_CLAMP</a></td>
	    <td><a href="#PRIMER_MAX_END_STABILITY">PRIMER_MAX_END_STABILITY</a></td>
	    <td><a href="#PRIMER_PAIR_WT_TEMPLATE_MISPRIMING">PRIMER_PAIR_WT_TEMPLATE_MISPRIMING</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INSIDE_PENALTY">PRIMER_INSIDE_PENALTY</a></td>
	    <td><a href="#PRIMER_MAX_GC">PRIMER_MAX_GC</a></td>
	    <td><a href="#PRIMER_PAIR_WT_TEMPLATE_MISPRIMING_TH">PRIMER_PAIR_WT_TEMPLATE_MISPRIMING_TH</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_DNA_CONC">PRIMER_INTERNAL_DNA_CONC</a></td>
	    <td><a href="#PRIMER_MAX_HAIRPIN_TH">PRIMER_MAX_HAIRPIN_TH</a></td>
	    <td><a href="#PRIMER_PICK_ANYWAY">PRIMER_PICK_ANYWAY</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_DNTP_CONC">PRIMER_INTERNAL_DNTP_CONC</a></td>
	    <td><a href="#PRIMER_MAX_LIBRARY_MISPRIMING">PRIMER_MAX_LIBRARY_MISPRIMING</a></td>
	    <td><a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MAX_GC">PRIMER_INTERNAL_MAX_GC</a></td>
	    <td><a href="#PRIMER_MAX_NS_ACCEPTED">PRIMER_MAX_NS_ACCEPTED</a></td>
	    <td><a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MAX_HAIRPIN_TH">PRIMER_INTERNAL_MAX_HAIRPIN_TH</a></td>
	    <td><a href="#PRIMER_MAX_POLY_X">PRIMER_MAX_POLY_X</a></td>
	    <td><a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MAX_LIBRARY_MISHYB">PRIMER_INTERNAL_MAX_LIBRARY_MISHYB</a></td>
	    <td><a href="#PRIMER_MAX_SELF_ANY">PRIMER_MAX_SELF_ANY</a></td>
	    <td><a href="#PRIMER_PRODUCT_MAX_TM">PRIMER_PRODUCT_MAX_TM</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MAX_NS_ACCEPTED">PRIMER_INTERNAL_MAX_NS_ACCEPTED</a></td>
	    <td><a href="#PRIMER_MAX_SELF_ANY_TH">PRIMER_MAX_SELF_ANY_TH</a></td>
	    <td><a href="#PRIMER_PRODUCT_MIN_TM">PRIMER_PRODUCT_MIN_TM</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MAX_POLY_X">PRIMER_INTERNAL_MAX_POLY_X</a></td>
	    <td><a href="#PRIMER_MAX_SELF_END">PRIMER_MAX_SELF_END</a></td>
	    <td><a href="#PRIMER_PRODUCT_OPT_SIZE">PRIMER_PRODUCT_OPT_SIZE</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MAX_SELF_ANY">PRIMER_INTERNAL_MAX_SELF_ANY</a></td>
	    <td><a href="#PRIMER_MAX_SELF_END_TH">PRIMER_MAX_SELF_END_TH</a></td>
	    <td><a href="#PRIMER_PRODUCT_OPT_TM">PRIMER_PRODUCT_OPT_TM</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MAX_SELF_ANY_TH">PRIMER_INTERNAL_MAX_SELF_ANY_TH</a></td>
	    <td><a href="#PRIMER_MAX_SIZE">PRIMER_MAX_SIZE</a></td>
	    <td><a href="#PRIMER_PRODUCT_SIZE_RANGE">PRIMER_PRODUCT_SIZE_RANGE</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MAX_SELF_END">PRIMER_INTERNAL_MAX_SELF_END</a></td>
	    <td><a href="#PRIMER_MAX_TEMPLATE_MISPRIMING">PRIMER_MAX_TEMPLATE_MISPRIMING</a></td>
	    <td><a href="#PRIMER_QUALITY_RANGE_MAX">PRIMER_QUALITY_RANGE_MAX</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MAX_SELF_END_TH">PRIMER_INTERNAL_MAX_SELF_END_TH</a></td>
	    <td><a href="#PRIMER_MAX_TEMPLATE_MISPRIMING_TH">PRIMER_MAX_TEMPLATE_MISPRIMING_TH</a></td>
	    <td><a href="#PRIMER_QUALITY_RANGE_MIN">PRIMER_QUALITY_RANGE_MIN</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MAX_SIZE">PRIMER_INTERNAL_MAX_SIZE</a></td>
	    <td><a href="#PRIMER_MAX_TM">PRIMER_MAX_TM</a></td>
	    <td><a href="#PRIMER_SALT_CORRECTIONS">PRIMER_SALT_CORRECTIONS</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MAX_TM">PRIMER_INTERNAL_MAX_TM</a></td>
	    <td><a href="#PRIMER_MIN_3_PRIME_OVERLAP_OF_JUNCTION">PRIMER_MIN_3_PRIME_OVERLAP_OF_JUNCTION</a></td>
	    <td><a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MIN_GC">PRIMER_INTERNAL_MIN_GC</a></td>
	    <td><a href="#PRIMER_MIN_5_PRIME_OVERLAP_OF_JUNCTION">PRIMER_MIN_5_PRIME_OVERLAP_OF_JUNCTION</a></td>
	    <td><a href="#PRIMER_SALT_MONOVALENT">PRIMER_SALT_MONOVALENT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MIN_QUALITY">PRIMER_INTERNAL_MIN_QUALITY</a></td>
	    <td><a href="#PRIMER_MIN_END_QUALITY">PRIMER_MIN_END_QUALITY</a></td>
	    <td><a href="#PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MIN_SIZE">PRIMER_INTERNAL_MIN_SIZE</a></td>
	    <td><a href="#PRIMER_MIN_GC">PRIMER_MIN_GC</a></td>
	    <td><a href="#PRIMER_SEQUENCING_ACCURACY">PRIMER_SEQUENCING_ACCURACY</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MIN_TM">PRIMER_INTERNAL_MIN_TM</a></td>
	    <td><a href="#PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE">PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE</a></td>
	    <td><a href="#PRIMER_SEQUENCING_INTERVAL">PRIMER_SEQUENCING_INTERVAL</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MISHYB_LIBRARY">PRIMER_INTERNAL_MISHYB_LIBRARY</a></td>
	    <td><a href="#PRIMER_MIN_QUALITY">PRIMER_MIN_QUALITY</a></td>
	    <td><a href="#PRIMER_SEQUENCING_LEAD">PRIMER_SEQUENCING_LEAD</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MUST_MATCH_FIVE_PRIME">PRIMER_INTERNAL_MUST_MATCH_FIVE_PRIME</a></td>
	    <td><a href="#PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE">PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE</a></td>
	    <td><a href="#PRIMER_SEQUENCING_SPACING">PRIMER_SEQUENCING_SPACING</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_MUST_MATCH_THREE_PRIME">PRIMER_INTERNAL_MUST_MATCH_THREE_PRIME</a></td>
	    <td><a href="#PRIMER_MIN_SIZE">PRIMER_MIN_SIZE</a></td>
	    <td><a href="#PRIMER_TASK">PRIMER_TASK</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_OPT_GC_PERCENT">PRIMER_INTERNAL_OPT_GC_PERCENT</a></td>
	    <td><a href="#PRIMER_MIN_THREE_PRIME_DISTANCE">PRIMER_MIN_THREE_PRIME_DISTANCE</a></td>
	    <td><a href="#PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT">PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_OPT_SIZE">PRIMER_INTERNAL_OPT_SIZE</a></td>
	    <td><a href="#PRIMER_MIN_TM">PRIMER_MIN_TM</a></td>
	    <td><a href="#PRIMER_THERMODYNAMIC_PARAMETERS_PATH">PRIMER_THERMODYNAMIC_PARAMETERS_PATH</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_OPT_TM">PRIMER_INTERNAL_OPT_TM</a></td>
	    <td><a href="#PRIMER_MISPRIMING_LIBRARY">PRIMER_MISPRIMING_LIBRARY</a></td>
	    <td><a href="#PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT">PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_SALT_DIVALENT">PRIMER_INTERNAL_SALT_DIVALENT</a></td>
	    <td><a href="#PRIMER_MUST_MATCH_FIVE_PRIME">PRIMER_MUST_MATCH_FIVE_PRIME</a></td>
	    <td><a href="#PRIMER_TM_FORMULA">PRIMER_TM_FORMULA</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_SALT_MONOVALENT">PRIMER_INTERNAL_SALT_MONOVALENT</a></td>
	    <td><a href="#PRIMER_MUST_MATCH_THREE_PRIME">PRIMER_MUST_MATCH_THREE_PRIME</a></td>
	    <td><a href="#PRIMER_WT_END_QUAL">PRIMER_WT_END_QUAL</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_END_QUAL">PRIMER_INTERNAL_WT_END_QUAL</a></td>
	    <td><a href="#PRIMER_NUM_RETURN">PRIMER_NUM_RETURN</a></td>
	    <td><a href="#PRIMER_WT_END_STABILITY">PRIMER_WT_END_STABILITY</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_GC_PERCENT_GT">PRIMER_INTERNAL_WT_GC_PERCENT_GT</a></td>
	    <td><a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a></td>
	    <td><a href="#PRIMER_WT_GC_PERCENT_GT">PRIMER_WT_GC_PERCENT_GT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_GC_PERCENT_LT">PRIMER_INTERNAL_WT_GC_PERCENT_LT</a></td>
	    <td><a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a></td>
	    <td><a href="#PRIMER_WT_GC_PERCENT_LT">PRIMER_WT_GC_PERCENT_LT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_HAIRPIN_TH">PRIMER_INTERNAL_WT_HAIRPIN_TH</a></td>
	    <td><a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a></td>
	    <td><a href="#PRIMER_WT_HAIRPIN_TH">PRIMER_WT_HAIRPIN_TH</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_LIBRARY_MISHYB">PRIMER_INTERNAL_WT_LIBRARY_MISHYB</a></td>
	    <td><a href="#PRIMER_OUTSIDE_PENALTY">PRIMER_OUTSIDE_PENALTY</a></td>
	    <td><a href="#PRIMER_WT_LIBRARY_MISPRIMING">PRIMER_WT_LIBRARY_MISPRIMING</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_NUM_NS">PRIMER_INTERNAL_WT_NUM_NS</a></td>
	    <td><a href="#PRIMER_PAIR_MAX_COMPL_ANY">PRIMER_PAIR_MAX_COMPL_ANY</a></td>
	    <td><a href="#PRIMER_WT_MASK_FAILURE_RATE">PRIMER_WT_MASK_FAILURE_RATE</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_SELF_ANY">PRIMER_INTERNAL_WT_SELF_ANY</a></td>
	    <td><a href="#PRIMER_PAIR_MAX_COMPL_ANY_TH">PRIMER_PAIR_MAX_COMPL_ANY_TH</a></td>
	    <td><a href="#PRIMER_WT_NUM_NS">PRIMER_WT_NUM_NS</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_SELF_ANY_TH">PRIMER_INTERNAL_WT_SELF_ANY_TH</a></td>
	    <td><a href="#PRIMER_PAIR_MAX_COMPL_END">PRIMER_PAIR_MAX_COMPL_END</a></td>
	    <td><a href="#PRIMER_WT_POS_PENALTY">PRIMER_WT_POS_PENALTY</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_SELF_END">PRIMER_INTERNAL_WT_SELF_END</a></td>
	    <td><a href="#PRIMER_PAIR_MAX_COMPL_END_TH">PRIMER_PAIR_MAX_COMPL_END_TH</a></td>
	    <td><a href="#PRIMER_WT_SELF_ANY">PRIMER_WT_SELF_ANY</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_SELF_END_TH">PRIMER_INTERNAL_WT_SELF_END_TH</a></td>
	    <td><a href="#PRIMER_PAIR_MAX_DIFF_TM">PRIMER_PAIR_MAX_DIFF_TM</a></td>
	    <td><a href="#PRIMER_WT_SELF_ANY_TH">PRIMER_WT_SELF_ANY_TH</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_SEQ_QUAL">PRIMER_INTERNAL_WT_SEQ_QUAL</a></td>
	    <td><a href="#PRIMER_PAIR_MAX_LIBRARY_MISPRIMING">PRIMER_PAIR_MAX_LIBRARY_MISPRIMING</a></td>
	    <td><a href="#PRIMER_WT_SELF_END">PRIMER_WT_SELF_END</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_SIZE_GT">PRIMER_INTERNAL_WT_SIZE_GT</a></td>
	    <td><a href="#PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING">PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING</a></td>
	    <td><a href="#PRIMER_WT_SELF_END_TH">PRIMER_WT_SELF_END_TH</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_SIZE_LT">PRIMER_INTERNAL_WT_SIZE_LT</a></td>
	    <td><a href="#PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING_TH">PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING_TH</a></td>
	    <td><a href="#PRIMER_WT_SEQ_QUAL">PRIMER_WT_SEQ_QUAL</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_TM_GT">PRIMER_INTERNAL_WT_TM_GT</a></td>
	    <td><a href="#PRIMER_PAIR_WT_COMPL_ANY">PRIMER_PAIR_WT_COMPL_ANY</a></td>
	    <td><a href="#PRIMER_WT_SIZE_GT">PRIMER_WT_SIZE_GT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_INTERNAL_WT_TM_LT">PRIMER_INTERNAL_WT_TM_LT</a></td>
	    <td><a href="#PRIMER_PAIR_WT_COMPL_ANY_TH">PRIMER_PAIR_WT_COMPL_ANY_TH</a></td>
	    <td><a href="#PRIMER_WT_SIZE_LT">PRIMER_WT_SIZE_LT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_LIBERAL_BASE">PRIMER_LIBERAL_BASE</a></td>
	    <td><a href="#PRIMER_PAIR_WT_COMPL_END">PRIMER_PAIR_WT_COMPL_END</a></td>
	    <td><a href="#PRIMER_WT_TEMPLATE_MISPRIMING">PRIMER_WT_TEMPLATE_MISPRIMING</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS</a></td>
	    <td><a href="#PRIMER_PAIR_WT_COMPL_END_TH">PRIMER_PAIR_WT_COMPL_END_TH</a></td>
	    <td><a href="#PRIMER_WT_TEMPLATE_MISPRIMING_TH">PRIMER_WT_TEMPLATE_MISPRIMING_TH</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_LOWERCASE_MASKING">PRIMER_LOWERCASE_MASKING</a></td>
	    <td><a href="#PRIMER_PAIR_WT_DIFF_TM">PRIMER_PAIR_WT_DIFF_TM</a></td>
	    <td><a href="#PRIMER_WT_TM_GT">PRIMER_WT_TM_GT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_MASK_3P_DIRECTION">PRIMER_MASK_3P_DIRECTION</a></td>
	    <td><a href="#PRIMER_PAIR_WT_IO_PENALTY">PRIMER_PAIR_WT_IO_PENALTY</a></td>
	    <td><a href="#PRIMER_WT_TM_LT">PRIMER_WT_TM_LT</a></td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_MASK_5P_DIRECTION">PRIMER_MASK_5P_DIRECTION</a></td>
	    <td><a href="#PRIMER_PAIR_WT_LIBRARY_MISPRIMING">PRIMER_PAIR_WT_LIBRARY_MISPRIMING</a></td>
	    <td>&nbsp;</td>
	  </tr>
	  <tr>
	    <td><a href="#PRIMER_MASK_FAILURE_RATE">PRIMER_MASK_FAILURE_RATE</a></td>
	    <td><a href="#PRIMER_PAIR_WT_PRODUCT_SIZE_GT">PRIMER_PAIR_WT_PRODUCT_SIZE_GT</a></td>
	    <td>&nbsp;</td>
	  </tr>
</table>
<h3><a id="PRIMER_TASK">PRIMER_TASK (string; default generic)</a></h3>
<p>This tag tells Primer3 what task to perform.
Legal values are:<br>
<br>
   <a class="p3_prop"><i>generic</i></a><br>
<br>
   Design a primer pair (the classic Primer3 task) if
   the <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>=1, and
   <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>=1. In addition, pick
   an internal hybridization oligo if
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>=1.
   <br><br>
   NOTE: If <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>=1,
   <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>=0
   and
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>=1,
   then behaves similarly to
   <a href="#PRIMER_TASK">PRIMER_TASK</a>=pick_primer_list.
<br>
<br>
   <a class="p3_prop"><i>pick_detection_primers</i></a><br>
<br>
   Deprecated alias for <a href="#PRIMER_TASK">PRIMER_TASK</a>=generic<br>
<br>
   <a class="p3_prop"><i>check_primers</i></a><br>
<br>
   Primer3 only checks the primers
   provided in <a href="#SEQUENCE_PRIMER">SEQUENCE_PRIMER</a>, <a href="#SEQUENCE_INTERNAL_OLIGO">SEQUENCE_INTERNAL_OLIGO</a> and
   <a href="#SEQUENCE_PRIMER_REVCOMP">SEQUENCE_PRIMER_REVCOMP</a>. It is the only task that does not
   require a sequence. However, if <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a> is provided, it is used.
   <br>
<br>
   <a class="p3_prop"><i>pick_primer_list</i></a><br>
<br>
   Pick all primers in <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a> (possibly
   limited by <a href="#SEQUENCE_INCLUDED_REGION">SEQUENCE_INCLUDED_REGION</a>,
   <a href="#SEQUENCE_EXCLUDED_REGION">SEQUENCE_EXCLUDED_REGION</a>,
   <a href="#SEQUENCE_PRIMER_PAIR_OK_REGION_LIST">SEQUENCE_PRIMER_PAIR_OK_REGION_LIST</a>, etc.).
   Returns the primers sorted by quality starting with the
   best primers. If <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a> and
   <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a> is selected Primer3 does not to pick
   primer pairs but generates independent lists of left primers, right primers,
   and, if requested, internal oligos.<br>
<br>
   <a class="p3_prop"><i>pick_sequencing_primers</i></a><br>
<br>
   Pick primers suited to sequence a region. <a href="#SEQUENCE_TARGET">SEQUENCE_TARGET</a> can be
   used to indicate several targets. The position of each primer is
   calculated for optimal sequencing results.<br>
<br>
   <a class="p3_prop"><i>pick_cloning_primers</i></a><br>
<br>
   Pick primers suited to clone a gene were the start nucleotide and
   the end nucleotide of the PCR fragment must be fixed, for example
   to clone an ORF. <a href="#SEQUENCE_INCLUDED_REGION">SEQUENCE_INCLUDED_REGION</a> must be used to
   indicate the first and the last nucleotide. Due to these
   limitations Primer3 can only vary the length of the primers. Set
   <a href="#PRIMER_PICK_ANYWAY">PRIMER_PICK_ANYWAY</a>=1 to obtain primers even if they violate
   specific constraints.<br>
<br>
   <a class="p3_prop"><i>pick_discriminative_primers</i></a><br>
<br>
   Pick primers suited to select primers which bind with their end at
   a specific position. This can be used to force the end of a primer
   to a polymorphic site, with the goal of discriminating between
   sequence variants. <a href="#SEQUENCE_TARGET">SEQUENCE_TARGET</a> must be used to
   provide a single target indicating the last nucleotide of the left
   (nucleotide before the first nucleotide of the target) and the right
   primer (nucleotide following the target). The primers border the
   <a href="#SEQUENCE_TARGET">SEQUENCE_TARGET</a>. Due to these limitations
   Primer3 can only vary the length of the primers. Set
   <a href="#PRIMER_PICK_ANYWAY">PRIMER_PICK_ANYWAY</a>=1 to obtain
   primers even if they violate specific constraints.<br>
<br>
   <a class="p3_prop"><i>pick_pcr_primers</i></a><br>
<br>
   Deprecated shortcut for the following settings:<br>
   <a href="#PRIMER_TASK">PRIMER_TASK</a>=generic<br>
   <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>=1<br>
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>=0<br>
   <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>=1<br>
   <br>
   WARNING: this task changes the values of <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>,
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>, and <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>
   in a way that is not obvious by looking at the input.<br>
<br>
   <a class="p3_prop"><i>pick_pcr_primers_and_hyb_probe</i></a><br>
<br>
   Deprecated shortcut for the following settings:<br>
   <a href="#PRIMER_TASK">PRIMER_TASK</a>=generic<br>
   <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>=1<br>
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>=1<br>
   <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>=1<br>
   <br>
   WARNING: this task changes the values of <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>,
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>, and <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>
   in a way that is not obvious by looking at the input.<br>
<br>
   <a class="p3_prop"><i>pick_left_only</i></a><br>
<br>
   Deprecated shortcut for the following settings:<br>
   <a href="#PRIMER_TASK">PRIMER_TASK</a>=generic<br>
   <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>=1<br>
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>=0<br>
   <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>=0<br>
   <br>
   WARNING: this task changes the values of <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>,
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>, and <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>
   in a way that is not obvious by looking at the input.<br>
<br>
   <a class="p3_prop"><i>pick_right_only</i></a><br>
<br>
   Deprecated shortcut for the following settings:<br>
   <a href="#PRIMER_TASK">PRIMER_TASK</a>=generic<br>
   <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>=0<br>
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>=0<br>
   <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>=1<br>
   <br>
   WARNING: this task changes the values of <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>,
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>, and <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>
   in a way that is not obvious by looking at the input.<br>
<br>
   <a class="p3_prop"><i>pick_hyb_probe_only</i></a><br>
<br>
   Deprecated shortcut for the following settings:<br>
   <a href="#PRIMER_TASK">PRIMER_TASK</a>=generic<br>
   <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>=0<br>
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>=1<br>
   <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>=0<br>
   <br>
   WARNING: this task changes the values of <a href="#PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER</a>,
   <a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a>, and <a href="#PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER</a>
   in a way that is not obvious by looking at the input.<br>
</p>
<h3><a id="PRIMER_PICK_LEFT_PRIMER">PRIMER_PICK_LEFT_PRIMER (boolean; default 1)</a></h3>
<p>If the associated value = 1 (non-0), then Primer3 will attempt to
pick left primers.</p>
<h3><a id="PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO (boolean; default 0)</a></h3>
<p>If the associated value = 1 (non-0), then Primer3 will attempt to
pick an internal oligo (hybridization probe to detect the PCR
product).</p>
<h3><a id="PRIMER_PICK_RIGHT_PRIMER">PRIMER_PICK_RIGHT_PRIMER (boolean; default 1)</a></h3>
<p>If the associated value = 1 (non-0), then Primer3 will attempt to
pick a right primer.</p>
<h3><a id="PRIMER_NUM_RETURN">PRIMER_NUM_RETURN (int; default 5)</a></h3>
<p>The maximum number of primer (pairs) to return.  Primer pairs
returned are sorted by their &quot;quality&quot;, in other words by the
value of the objective function (where a lower number indicates a
better primer pair).  Caution: setting this parameter to a large
value will increase running time.</p>
<h3><a id="PRIMER_MIN_3_PRIME_OVERLAP_OF_JUNCTION">PRIMER_MIN_3_PRIME_OVERLAP_OF_JUNCTION (int; default 4)</a></h3>
<p>The 3' end of the left OR the right primer must
overlap one of the junctions in <a href="#SEQUENCE_OVERLAP_JUNCTION_LIST">SEQUENCE_OVERLAP_JUNCTION_LIST</a> by this amount.
See details in <a href="#SEQUENCE_OVERLAP_JUNCTION_LIST">SEQUENCE_OVERLAP_JUNCTION_LIST</a>.
</p>
<h3><a id="PRIMER_MIN_5_PRIME_OVERLAP_OF_JUNCTION">PRIMER_MIN_5_PRIME_OVERLAP_OF_JUNCTION (int; default 7)</a></h3>
<p>The 5' end of the left OR the right primer must
overlap one of the junctions in <a href="#SEQUENCE_OVERLAP_JUNCTION_LIST">SEQUENCE_OVERLAP_JUNCTION_LIST</a> by this amount.
See details in <a href="#SEQUENCE_OVERLAP_JUNCTION_LIST">SEQUENCE_OVERLAP_JUNCTION_LIST</a>.
</p>
<h3><a id="PRIMER_MUST_MATCH_FIVE_PRIME">PRIMER_MUST_MATCH_FIVE_PRIME (ambiguous nucleotide sequence; default empty)</a></h3>
<p>Discards all primers which do not match this match sequence at the 5' end.
(New in v. 2.3.6, added by A. Untergasser.)<br>
<br>
The match sequence must be 5 nucletides long and can contain the following letters:</p>
<pre>
    N   Any nucleotide
    A   Adenine
    G   Guanine
    C   Cytosine
    T   Thymine
    R   Purine (A or G)
    Y   Pyrimidine (C or T)
    W   Weak (A or T)
    S   Strong (G or C)
    M   Amino (A or C)
    K   Keto (G or T)
    B   Not A (G or C or T)
    H   Not G (A or C or T)
    D   Not C (A or G or T)
    V   Not T (A or G or C)
</pre>
<p>Any primer which will not match the entire match sequence at th 5' end will
be discarded and not evaluated. Setting strict requirements here will result
in low quality primers due to the high numbers of primers discarded at this
step.<br>
<br>
<strong>Example 1:</strong><br>
PRIMER_MUST_MATCH_FIVE_PRIME=tgnnn<br><br>
Could result in the following matching:</p>
<pre>
    tgcatgattggatacgtttga
    |||||
    tgnnn
 -> This primer would be used.
    attcgattctccccggtatc
      |||
    tgnnn
 -> This primer would be discarded.
</pre><br>
<p><strong>Example 2:</strong><br>
PRIMER_MUST_MATCH_FIVE_PRIME=hnnnn<br><br>
Could result in the following matching:</p>
<pre>
    tgcatgattggatacgtttga
    |||||
    hnnnn
 -> This primer would be used.
    ggctgatgaaggaaagcaag
     ||||
    hnnnn
 -> This primer would be discarded.
</pre><br>
<p>This parameter would force all primers selected by Primer3 to not have guanosine
at the 5' end of any primer which could be useful to avoid quenching of
flourochromes.
</p>
<h3><a id="PRIMER_INTERNAL_MUST_MATCH_FIVE_PRIME">PRIMER_INTERNAL_MUST_MATCH_FIVE_PRIME (ambiguous nucleotide sequence; default empty)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MUST_MATCH_FIVE_PRIME">PRIMER_MUST_MATCH_FIVE_PRIME</a> for the internal oligo.</p>
<h3><a id="PRIMER_MUST_MATCH_THREE_PRIME">PRIMER_MUST_MATCH_THREE_PRIME (ambiguous nucleotide sequence; default empty)</a></h3>
<p>Discards all primers which do not match this match sequence at the 3' end. Similar
parameter to <a href="#PRIMER_MUST_MATCH_FIVE_PRIME">PRIMER_MUST_MATCH_FIVE_PRIME</a>, but limits the 3' end.
(New in v. 2.3.6, added by A. Untergasser.)<br>
<br>
The match sequence must be 5 nucletides long and can contain the following letters:</p>
<pre>
    N   Any nucleotide
    A   Adenine
    G   Guanine
    C   Cytosine
    T   Thymine
    R   Purine (A or G)
    Y   Pyrimidine (C or T)
    W   Weak (A or T)
    S   Strong (G or C)
    M   Amino (A or C)
    K   Keto (G or T)
    B   Not A (G or C or T)
    H   Not G (A or C or T)
    D   Not C (A or G or T)
    V   Not T (A or G or C)
</pre>
<p>Any primer which will not match the entire match sequence at th 3' end will
be discarded and not evaluated. Setting strict requirements here will result
in low quality primers due to the high numbers of primers discarded at this
step.<br>
<br>
<strong>Example 1:</strong><br>
PRIMER_MUST_MATCH_FIVE_PRIME=nnnga<br><br>
Could result in the following matching:</p>
<pre>
    tgcatgattggatacgtttga
                    |||||
                    nnnga
 -> This primer would be used.
    attcgattctccccggtatc
                   |||
                   nnnga
 -> This primer would be discarded.
</pre>
<h3><a id="PRIMER_INTERNAL_MUST_MATCH_THREE_PRIME">PRIMER_INTERNAL_MUST_MATCH_THREE_PRIME (ambiguous nucleotide sequence; default empty)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MUST_MATCH_THREE_PRIME">PRIMER_MUST_MATCH_THREE_PRIME</a> for the internal oligo.</p>
<h3><a id="PRIMER_PRODUCT_SIZE_RANGE">PRIMER_PRODUCT_SIZE_RANGE (size range list; default 100-300)</a></h3>
<p>The associated values specify the lengths of the product that the
user wants the primers to create, and is a space separated list
of elements of the form</p>
<pre>
<i>&lt;x&gt;-&lt;y&gt;</i>
</pre>
<p>where an <a class="p3_prop"><i>&lt;x&gt;-&lt;y&gt;</i></a> pair is a legal range of lengths for the
product.  For example, if one wants PCR products to be between
100 to 150 bases (inclusive) then one would set this parameter to
<a class="p3_prop"><i>100-150</i></a>.  If one desires PCR products in either the range from
100 to 150 bases or in the range from 200 to 250 bases then one
would set this parameter to <a class="p3_prop"><i>100-150 200-250</i></a>.<br>
<br>
Primer3 favors product-size ranges to the left side of the parameter string.
Primer3 will return legal primers pairs in the first range
regardless the value of the objective function for pairs in subsequent ranges.
Only if there are an insufficient number of primers in the first
range will Primer3 return primers in a subsequent range.<br>
<br>
For those with primarily a computational background,
the PCR product size is the size (in base pairs)
of the DNA fragment that would be produced by the
PCR reaction on the given sequence template.  This
would, of course, include the primers themselves.</p>
<h3><a id="PRIMER_PRODUCT_OPT_SIZE">PRIMER_PRODUCT_OPT_SIZE (int; default 0)</a></h3>
<p>The optimum size for the PCR product.  0 indicates that there is
no optimum product size.  This parameter influences primer pair
selection only if <a href="#PRIMER_PAIR_WT_PRODUCT_SIZE_GT">PRIMER_PAIR_WT_PRODUCT_SIZE_GT</a> or
<a href="#PRIMER_PAIR_WT_PRODUCT_SIZE_LT">PRIMER_PAIR_WT_PRODUCT_SIZE_LT</a> is non-0.<br>
A non-0 value for this parameter will likely increase calculation time, so set this only if
a product size near a specific value is truly important.
</p>
<h3><a id="PRIMER_PAIR_WT_PRODUCT_SIZE_LT">PRIMER_PAIR_WT_PRODUCT_SIZE_LT (float; default 0.0)</a></h3>
<p>Penalty weight for products shorter than <a href="#PRIMER_PRODUCT_OPT_SIZE">PRIMER_PRODUCT_OPT_SIZE</a>.</p>
<h3><a id="PRIMER_PAIR_WT_PRODUCT_SIZE_GT">PRIMER_PAIR_WT_PRODUCT_SIZE_GT (float; default 0.0)</a></h3>
<p>Penalty weight for products longer than <a href="#PRIMER_PRODUCT_OPT_SIZE">PRIMER_PRODUCT_OPT_SIZE</a>.</p>
<h3><a id="PRIMER_MIN_SIZE">PRIMER_MIN_SIZE (int; default 18)</a></h3>
<p>Minimum acceptable length of a primer.  Must be greater than 0
and less than or equal to <a href="#PRIMER_MAX_SIZE">PRIMER_MAX_SIZE</a>.</p>
<h3><a id="PRIMER_INTERNAL_MIN_SIZE">PRIMER_INTERNAL_MIN_SIZE (int; default 18)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MIN_SIZE">PRIMER_MIN_SIZE</a> for the internal oligo.</p>
<h3><a id="PRIMER_OPT_SIZE">PRIMER_OPT_SIZE (int; default 20)</a></h3>
<p>Optimum length (in bases) of a primer. Primer3 will attempt to
pick primers close to this length.</p>
<h3><a id="PRIMER_INTERNAL_OPT_SIZE">PRIMER_INTERNAL_OPT_SIZE (int; default 20)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a> for the internal oligo.</p>
<h3><a id="PRIMER_MAX_SIZE">PRIMER_MAX_SIZE (int; default 27)</a></h3>
<p>Maximum acceptable length (in bases) of a primer.  Currently this
parameter cannot be larger than 35.  This limit is governed by
maximum oligo size for which Primer3's melting-temperature is
valid.</p>
<h3><a id="PRIMER_INTERNAL_MAX_SIZE">PRIMER_INTERNAL_MAX_SIZE (int; default 27)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MAX_SIZE">PRIMER_MAX_SIZE</a> for the internal oligo.</p>
<h3><a id="PRIMER_WT_SIZE_LT">PRIMER_WT_SIZE_LT (float; default 1.0)</a></h3>
<p>Penalty weight for primers shorter than <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a>.</p>
<h3><a id="PRIMER_INTERNAL_WT_SIZE_LT">PRIMER_INTERNAL_WT_SIZE_LT (float; default 1.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_SIZE_LT">PRIMER_WT_SIZE_LT</a> for the internal oligo.</p>
<h3><a id="PRIMER_WT_SIZE_GT">PRIMER_WT_SIZE_GT (float; default 1.0)</a></h3>
<p>Penalty weight for primers longer than <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a>.</p>
<h3><a id="PRIMER_INTERNAL_WT_SIZE_GT">PRIMER_INTERNAL_WT_SIZE_GT (float; default 1.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_INTERNAL_WT_SIZE_GT">PRIMER_INTERNAL_WT_SIZE_GT</a> for the internal oligo.</p>
<h3><a id="PRIMER_MIN_GC">PRIMER_MIN_GC (float; default 20.0)</a></h3>
<p>Minimum allowable percentage of Gs and Cs in any primer.</p>
<h3><a id="PRIMER_INTERNAL_MIN_GC">PRIMER_INTERNAL_MIN_GC (float; default 20.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MIN_GC">PRIMER_MIN_GC</a> for the internal oligo.</p>
<h3><a id="PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT (float; default 50.0)</a></h3>
<p>Optimum GC percent.  This parameter influences primer selection only if
<a href="#PRIMER_WT_GC_PERCENT_GT">PRIMER_WT_GC_PERCENT_GT</a> or <a href="#PRIMER_WT_GC_PERCENT_LT">PRIMER_WT_GC_PERCENT_LT</a> are non-0.</p>
<h3><a id="PRIMER_INTERNAL_OPT_GC_PERCENT">PRIMER_INTERNAL_OPT_GC_PERCENT (float; default 50.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a> for the internal oligo.</p>
<h3><a id="PRIMER_MAX_GC">PRIMER_MAX_GC (float; default 80.0)</a></h3>
<p>Maximum allowable percentage of Gs and Cs in any primer generated
by Primer.</p>
<h3><a id="PRIMER_INTERNAL_MAX_GC">PRIMER_INTERNAL_MAX_GC (float; default 80.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MAX_GC">PRIMER_MAX_GC</a> for the internal oligo.</p>
<h3><a id="PRIMER_WT_GC_PERCENT_LT">PRIMER_WT_GC_PERCENT_LT (float; default 0.0)</a></h3>
<p>Penalty weight for primers with GC percent lower than
<a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a>.</p>
<h3><a id="PRIMER_INTERNAL_WT_GC_PERCENT_LT">PRIMER_INTERNAL_WT_GC_PERCENT_LT (float; default 0.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_GC_PERCENT_LT">PRIMER_WT_GC_PERCENT_LT</a> for the internal oligo.</p>
<h3><a id="PRIMER_WT_GC_PERCENT_GT">PRIMER_WT_GC_PERCENT_GT (float; default 0.0)</a></h3>
<p>Penalty weight for primers with GC percent higher than
<a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a>.</p>
<h3><a id="PRIMER_INTERNAL_WT_GC_PERCENT_GT">PRIMER_INTERNAL_WT_GC_PERCENT_GT (float; default 0.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_GC_PERCENT_GT">PRIMER_WT_GC_PERCENT_GT</a> for the internal oligo.</p>
<h3><a id="PRIMER_GC_CLAMP">PRIMER_GC_CLAMP (int; default 0)</a></h3>
<p>Require the specified number of consecutive Gs and Cs at the 3'
end of both the left and right primer.  (This parameter has no
effect on the internal oligo if one is requested.)</p>
<h3><a id="PRIMER_MAX_END_GC">PRIMER_MAX_END_GC (int; default 5)</a></h3>
<p>The maximum number of Gs or Cs allowed in the last five 3'
bases of a left or right primer.</p>
<h3><a id="PRIMER_MIN_TM">PRIMER_MIN_TM (float; default 57.0)</a></h3>
<p>Minimum acceptable melting temperature (Celsius) for a primer
oligo.</p>
<h3><a id="PRIMER_INTERNAL_MIN_TM">PRIMER_INTERNAL_MIN_TM (float; default 57.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MIN_TM">PRIMER_MIN_TM</a> for the internal oligo.</p>
<h3><a id="PRIMER_OPT_TM">PRIMER_OPT_TM (float; default 60.0)</a></h3>
<p>Optimum melting temperature (Celsius) for a primer. Primer3
will try to pick primers with melting temperatures are close to
this temperature.  The oligo melting temperature formula used can
be specified by user. Please see <a href="#PRIMER_TM_FORMULA">PRIMER_TM_FORMULA</a> for more
information.</p>
<h3><a id="PRIMER_INTERNAL_OPT_TM">PRIMER_INTERNAL_OPT_TM (float; default 60.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a> for the internal oligo.</p>
<h3><a id="PRIMER_MAX_TM">PRIMER_MAX_TM (float; default 63.0)</a></h3>
<p>Maximum acceptable melting temperature (Celsius) for a primer
oligo.</p>
<h3><a id="PRIMER_INTERNAL_MAX_TM">PRIMER_INTERNAL_MAX_TM (float; default 63.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MAX_TM">PRIMER_MAX_TM</a> for the internal oligo.</p>
<h3><a id="PRIMER_PAIR_MAX_DIFF_TM">PRIMER_PAIR_MAX_DIFF_TM (float; default 100.0)</a></h3>
<p>Maximum acceptable (unsigned) difference between the melting
temperatures of the left and right primers.</p>
<h3><a id="PRIMER_WT_TM_LT">PRIMER_WT_TM_LT (float; default 1.0)</a></h3>
<p>Penalty weight for primers with Tm lower than <a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a>.</p>
<h3><a id="PRIMER_INTERNAL_WT_TM_LT">PRIMER_INTERNAL_WT_TM_LT (float; default 1.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_TM_LT">PRIMER_WT_TM_LT</a> for the internal oligo.</p>
<h3><a id="PRIMER_WT_TM_GT">PRIMER_WT_TM_GT (float; default 1.0)</a></h3>
<p>Penalty weight for primers with Tm over <a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a>.</p>
<h3><a id="PRIMER_INTERNAL_WT_TM_GT">PRIMER_INTERNAL_WT_TM_GT (float; default 1.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_TM_GT">PRIMER_WT_TM_GT</a> for the internal oligo.</p>
<h3><a id="PRIMER_PAIR_WT_DIFF_TM">PRIMER_PAIR_WT_DIFF_TM (float; default 0.0)</a></h3>
<p>Penalty weight for the TM difference between the left primer and
the right primer.</p>
<h3><a id="PRIMER_PRODUCT_MIN_TM">PRIMER_PRODUCT_MIN_TM (float; default -1000000.0)</a></h3>
<p>The minimum allowed melting temperature of the amplicon.  Please
see the documentation on <a href="#PRIMER_PRODUCT_MAX_TM">PRIMER_PRODUCT_MAX_TM</a> for details.</p>
<h3><a id="PRIMER_PRODUCT_OPT_TM">PRIMER_PRODUCT_OPT_TM (float; default 0.0)</a></h3>
<p>The optimum melting temperature for the PCR product. 0 indicates
that there is no optimum temperature.</p>
<h3><a id="PRIMER_PRODUCT_MAX_TM">PRIMER_PRODUCT_MAX_TM (float; default 1000000.0)</a></h3>
<p>The maximum allowed melting temperature of the amplicon.  Primer3
calculates product Tm calculated using the formula from Bolton
and McCarthy, PNAS 84:1390 (1962) as presented in Sambrook,
Fritsch and Maniatis, Molecular Cloning, p 11.46 (1989, CSHL
Press).</p>
<pre>
   Tm = 81.5 + 16.6(log10([Na+])) + .41*(%GC) - 600/length
</pre>
<p>Where [Na+] is the molar sodium concentration, (%GC) is the
percent of Gs and Cs in the sequence, and length is the length of
the sequence.<br>
<br>
A similar formula is used by the prime primer selection program
in GCG (<a href="http://www.gcg.com">http://www.gcg.com</a>), which instead uses 675.0 / length in
the last term (after F. Baldino, Jr, M.-F. Chesselet, and M.E.
Lewis, Methods in Enzymology 168:766 (1989) eqn (1) on page 766
without the mismatch and formamide terms).  The formulas here and
in Baldino et al. assume Na+ rather than K+.  According to
J.G. Wetmur, Critical Reviews in BioChem. and Mol. Bio. 26:227
(1991) 50 mM K+ should be equivalent in these formulae to .2 M
Na+.  Primer3 uses the same salt concentration value for
calculating both the primer melting temperature and the oligo
melting temperature.  If you are planning to use the PCR product
for hybridization later this behavior will not give you the Tm
under hybridization conditions.</p>
<h3><a id="PRIMER_PAIR_WT_PRODUCT_TM_LT">PRIMER_PAIR_WT_PRODUCT_TM_LT (float; default 0.0)</a></h3>
<p>Penalty weight for products with a Tm lower than
<a href="#PRIMER_PRODUCT_OPT_TM">PRIMER_PRODUCT_OPT_TM</a>.</p>
<h3><a id="PRIMER_PAIR_WT_PRODUCT_TM_GT">PRIMER_PAIR_WT_PRODUCT_TM_GT (float; default 0.0)</a></h3>
<p>Penalty weight for products with a Tm higher than
<a href="#PRIMER_PRODUCT_OPT_TM">PRIMER_PRODUCT_OPT_TM</a>.</p>
<h3><a id="PRIMER_TM_FORMULA">PRIMER_TM_FORMULA (int; default 1)</a></h3>
<p>Specifies details of melting temperature calculation.  (New in
v. 1.1.0, added by Maido Remm and Triinu Koressaar.)<br>
<br>
A value of 0 directs Primer3 to a backward compatible calculation
(in other words, the only calculation available in previous
version of Primer3).<br>
<br>
This backward compatible calculation uses the table of
thermodynamic parameters in the paper [Breslauer KJ, Frank R,
Bl&ouml;cker H and Marky LA (1986) "Predicting DNA duplex stability
from the base sequence" Proc Natl Acad Sci 83:4746-50
<a href="http://dx.doi.org/10.1073/pnas.83.11.3746">http://dx.doi.org/10.1073/pnas.83.11.3746</a>],
and the method in the paper [Rychlik W, Spencer WJ and Rhoads
RE (1990) "Optimization of the annealing temperature for DNA
amplification in vitro", Nucleic Acids Res 18:6409-12
<a href="http://dx.doi.org/10.1093/nar/18.21.6409">http://dx.doi.org/10.1093/nar/18.21.6409</a>].<br>
<br>
A value of 1 (*RECOMMENDED*) directs Primer3 to use the table of
thermodynamic values and the method for melting temperature
calculation suggested in the paper [SantaLucia JR (1998) "A unified
view of polymer, dumbbell and oligonucleotide DNA nearest-neighbor
thermodynamics", Proc Natl Acad Sci 95:1460-65
<a href="http://dx.doi.org/10.1073/pnas.95.4.1460">http://dx.doi.org/10.1073/pnas.95.4.1460</a>].<br>
<br>
Use tag <a href="#PRIMER_SALT_CORRECTIONS">PRIMER_SALT_CORRECTIONS</a>, to specify the salt correction
method for melting temperature calculation.<br>
<br>
Example of calculating the melting temperature of an oligo if
<a href="#PRIMER_TM_FORMULA">PRIMER_TM_FORMULA</a>=1 and <a href="#PRIMER_SALT_CORRECTIONS">PRIMER_SALT_CORRECTIONS</a>=1
recommended values):<br>
<br>
primer=CGTGACGTGACGGACT<br>
<br>
Using default salt and DNA concentrations we have</p>
<pre>
Tm = deltaH/(deltaS + R*ln(C/4))
</pre>
<p>where R is the gas constant (1.987 cal/K mol) and C is the DNA
concentration.</p>
<pre>
deltaH(predicted) =
  = dH(CG) + dH(GT) + dH(TG) + .. + dH(CT) +
     + dH(init.w.term.GC) + dH(init.w.term.AT) =
  = -10.6 + (-8.4) + (-8.5) + .. + (-7.8) + 0.1 + 2.3  =
  = -128.8 kcal/mol
</pre>
<p>where 'init.w.term GC' and 'init.w.term AT' are two
initiation parameters for duplex formation: 'initiation with
terminal GC' and 'initiation with terminal AT'</p>
<pre>
deltaS(predicted) =
  = dS(CG) + dS(GT) + dS(TG) + .. + dS(CT) +
    + dS(init.w.term.GC) + dS(init.w.term.AT) =
  = -27.2 + (-22.4) + (-22.7) + .. + (-21.0) + (-2.8) + 4.1 =
  = -345.2 cal/k*mol
deltaS(salt corrected) =
  = deltaS(predicted) + 0.368*15(NN pairs)*ln(0.05M monovalent cations) =
  = -361.736
Tm = -128.800/(-361.736+1.987*ln((5*10^(-8))/4)) =
   = 323.704 K
Tm(C) = 323.704 - 273.15 = 50.554 C
</pre>
<h3><a id="PRIMER_SALT_MONOVALENT">PRIMER_SALT_MONOVALENT (float; default 50.0)</a></h3>
<p>The millimolar (mM) concentration of monovalent salt cations (usually KCl) in the PCR.
Primer3 uses this argument to calculate oligo and primer melting
temperatures.
Use tag <a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a> and
<a href="#PRIMER_INTERNAL_SALT_DIVALENT">PRIMER_INTERNAL_SALT_DIVALENT</a> to specify the concentrations
of divalent cations (in which case you also should also set tag <a href="#PRIMER_DNTP_CONC">PRIMER_DNTP_CONC</a> to
a reasonable value).
</p>
<h3><a id="PRIMER_INTERNAL_SALT_MONOVALENT">PRIMER_INTERNAL_SALT_MONOVALENT (float; default 50.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_SALT_MONOVALENT">PRIMER_SALT_MONOVALENT</a> for the internal oligo.</p>
<h3><a id="PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT (float; default 1.5)</a></h3>
<p>The millimolar concentration of divalent salt cations (usually MgCl^(2+)) in
the PCR. (New in v. 1.1.0, added by Maido Remm and Triinu Koressaar)<br>
<br>
Primer3 converts concentration of divalent cations to concentration
of monovalent cations using formula suggested in the paper [Ahsen von N,
Wittwer CT, Schutz E (2001) "Oligonucleotide Melting Temperatures under PCR
Conditions: Nearest-Neighbor Corrections for Mg^(2+), Deoxynucleotide Triphosphate,
and Dimethyl Sulfoxide Concentrations with Comparison to Alternative Empirical
Formulas", Clinical Chemistry 47:1956-61 <a href="http://www.clinchem.org/cgi/content/full/47/11/1956">http://www.clinchem.org/cgi/content/full/47/11/1956</a>].</p>
<pre>
[Monovalent cations] = [Monovalent cations] + 120*(([divalent cations] - [dNTP])^0.5)
</pre>
<p>In addition,
if the specified concentration of dNTPs (<a href="#PRIMER_DNTP_CONC">PRIMER_DNTP_CONC</a>)
is larger than the concentration of divalent cations (<a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a>)
then the effect of the divalent cations is not considered. The
concentration of dNTPs is considered in the formula above
because of some magnesium is bound by the dNTP.
The adjusted concentration of monovalent cations is used in the
calculation of oligo/primer
melting temperature, PCR product melting temperature,
the stability of oligo dimers and
secondary structures (when <a href="#PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT">PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT</a> is 1),
and the stability of ectopic annealing of oligos to template
(when <a href="#PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT">PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT</a> is 1).
If <a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a> > 0.0, be sure to
set tag <a href="#PRIMER_DNTP_CONC">PRIMER_DNTP_CONC</a> to specify the concentration of dNTPs.
</p>
<h3><a id="PRIMER_INTERNAL_SALT_DIVALENT">PRIMER_INTERNAL_SALT_DIVALENT (float; default 0.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a> for the internal oligo.</p>
<h3><a id="PRIMER_DNTP_CONC">PRIMER_DNTP_CONC (float; default 0.6)</a></h3>
<p>The millimolar concentration of the sum of all
deoxyribonucleotide triphosphates. A reaction mix containing
0.2 mM ATP, 0.2 mM CTP, 0.2 mM GTP and 0.2 mM TTP would have a
PRIMER_DNTP_CONC=0.8.
This argument is considered for oligo and
primer melting temperatures, for
PCR product melting temperature, or for secondary
structure calculations only if
<a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a> is > 0.0.
See <a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a>.</p>
<h3><a id="PRIMER_INTERNAL_DNTP_CONC">PRIMER_INTERNAL_DNTP_CONC (float; default 0.0)</a></h3>
<p>Parameter for internal oligos analogous to <a href="#PRIMER_DNTP_CONC">PRIMER_DNTP_CONC</a>.</p>
<h3><a id="PRIMER_SALT_CORRECTIONS">PRIMER_SALT_CORRECTIONS (int; default 1)</a></h3>
<p>Specifies the salt correction formula for the melting temperature
calculation.  (New in v. 1.1.0, added by Maido Remm and Triinu
Koressaar)<br>
<br>
A value of 0 directs Primer3 to use the the salt correction
formula in the paper [Schildkraut, C, and Lifson, S (1965)
"Dependence of the melting temperature of DNA on salt
concentration", Biopolymers 3:195-208 (not available on-line)].
This was the formula used in older versions of Primer3.
<br>
<br>
A value of 1 (*RECOMMENDED*) directs Primer3 to use the salt correction
formula in the paper [SantaLucia JR (1998) "A unified view of polymer,
dumbbell and oligonucleotide DNA nearest-neighbor thermodynamics",
Proc Natl Acad Sci 95:1460-65
<a href="http://dx.doi.org/10.1073/pnas.95.4.1460">http://dx.doi.org/10.1073/pnas.95.4.1460</a>]
<br>
<br>
A value of 2 directs Primer3 to use the salt correction formula
in the paper [Owczarzy, R., Moreira, B.G., You, Y., Behlke, M.A., and
Walder, J.A. (2008). Predicting stability of DNA duplexes in solutions
containing magnesium and monovalent cations. Biochemistry 47, 5336-5353
<a href="http://dx.doi.org/10.1021/bi702363u">http://dx.doi.org/10.1021/bi702363u</a>] following recommendations in the paper
[Ahsen, v.N., Wittwer, C.T., and Sch&uuml;tz, E. (2010). Monovalent and divalent
salt correction algorithms for Tm prediction-recommendations for Primer3
usage. Brief Bioinform 12, 514 <a href="http://dx.doi.org/10.1093/bib/bbq081">http://dx.doi.org/10.1093/bib/bbq081</a>].
<br>
<br>
For all values of <a href="#PRIMER_SALT_CORRECTIONS">PRIMER_SALT_CORRECTIONS</a>,
Primer3 also considers the values of the
tags <a href="#PRIMER_SALT_DIVALENT">PRIMER_SALT_DIVALENT</a>,
<a href="#PRIMER_INTERNAL_SALT_DIVALENT">PRIMER_INTERNAL_SALT_DIVALENT</a>,
<a href="#PRIMER_DNTP_CONC">PRIMER_DNTP_CONC</a>, and
<a href="#PRIMER_INTERNAL_DNTP_CONC">PRIMER_INTERNAL_DNTP_CONC</a>.
</p>
<h3><a id="PRIMER_DNA_CONC">PRIMER_DNA_CONC (float; default 50.0)</a></h3>
<p>
<p>
A value to use as nanomolar (nM) concentration of each annealing oligo
over the course the PCR.
Primer3 uses this argument to esimate oligo melting
temperatures. This parameter corresponds to 'c' in equation (ii)
of the paper [SantaLucia (1998) A unified view of polymer, dumbbell,
and oligonucleotide DNA nearest-neighbor thermodynamics.
Proc Natl Acad Sci 95:1460-1465
<a href="http://www.pnas.org/content/95/4/1460.full.pdf+html">http://www.pnas.org/content/95/4/1460.full.pdf+html</a>],
where a suitable value (for a lower initial concentration of template)
is "empirically determined".<br>
<br>
The default (50nM) works well with the standard
protocol used at the Whitehead/MIT Center for Genome
Research--0.5 microliters of 20 micromolar concentration for each
primer in a 20 microliter reaction with 10 nanograms
template, 0.025 units/microliter Taq polymerase in 0.1 mM each
dNTP, 1.5mM MgCl2, 50mM KCl, 10mM Tris-HCL (pH 9.3) using 35
cycles with an annealing temperature of 56 degrees Celsius.<br>
<br>
The value of this parameter is less than the actual
concentration of oligos in the initial reaction mix because it is the
concentration of annealing oligos, which in turn depends on the
amount of template (including PCR product) in a given cycle.
This concentration increases a great deal during a PCR;
fortunately PCR seems quite robust for a variety of oligo melting
temperatures.
<br>
See ADVICE FOR PICKING PRIMERS.</p>
<h3><a id="PRIMER_INTERNAL_DNA_CONC">PRIMER_INTERNAL_DNA_CONC (float; default 50.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_DNA_CONC">PRIMER_DNA_CONC</a> for the internal oligo.</p>
<h3><a id="PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT">PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT (boolean; default 1)</a></h3>
<p>If the associated value = 1, then Primer3 will use
thermodynamic models to calculate the
the propensity of oligos to form hairpins and dimers.
</p>
<h3><a id="PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT">PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT (boolean; default 0)</a></h3>
<p>If the associated value = 1, then Primer3 will use
thermodynamic models to calculate the
the propensity of oligos to anneal to undesired sites in the template sequence.
</p>
<h3><a id="PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT (boolean; default 0)</a></h3>
<p>If the associated value = 1, then Primer3 will print out the
calculated secondary structures, for example:</p>
<h4>Dimers:</h4>
<pre>
t: 7.8  dG: -1724  dH: -91400  dS: -289
  5' ACGCAAAGCACGCTCC-CGATC 3'
      ||   |||  |||   ||
3' CTAGCCC-TCGCACGAAACGCA 5'
</pre>
<h4>Hairpins:</h4>
<pre>
t: 36.4  dG: 42  dH: -22500  dS: -73
5' CGCAAAGCACGCT&#x2510;
            ||  &#x2502;
        3' AGCCC&#x2518;
</pre>
<p>
The tags <a href="#PRIMER_LEFT_4_SELF_ANY_STUCT">PRIMER_LEFT_4_SELF_ANY_STUCT</a>, <a href="#PRIMER_LEFT_4_SELF_END_STUCT">PRIMER_LEFT_4_SELF_END_STUCT</a>,
<a href="#PRIMER_LEFT_4_HAIRPIN_STUCT">PRIMER_LEFT_4_HAIRPIN_STUCT</a> (these tags are also present for RIGHT and INTERNAL
primers), <a href="#PRIMER_PAIR_4_COMPL_ANY_STUCT">PRIMER_PAIR_4_COMPL_ANY_STUCT</a> and <a href="#PRIMER_PAIR_4_COMPL_END_STUCT">PRIMER_PAIR_4_COMPL_END_STUCT</a>
are only present if a secondary structure could be calculated.<br>
<br>
As the string has to fit on one line, the newlines are indicated by the two characters
'\' and 'n' and have to be replaced (regex: /\\n/\n/g). Hairpins use unicode characters
for the turn. For html they have to be replaced (regex: /U\+25(\d\d)/&amp;#x25$1;/g).
</p>
<h3><a id="PRIMER_THERMODYNAMIC_PARAMETERS_PATH">PRIMER_THERMODYNAMIC_PARAMETERS_PATH (string; default ./primer3_config)</a></h3>
<p>This tag specifies the path to the directory that contains all the
parameter files used by the thermodynamic approach. In Linux, there are two
default locations that are tested if this tag is not defined:
<i>./primer3_config/</i> and <i>/opt/primer3_config/</i>.
For Windows, there is only one default location: <i>.\primer3_config\</i>.</p>
<h3><a id="PRIMER_MAX_SELF_ANY">PRIMER_MAX_SELF_ANY (decimal, 9999.99; default 8.00)</a></h3>
<p>PRIMER_MAX_SELF_ANY describes the tendency of a primer to bind to
itself (interfering with target sequence binding). It will score
ANY binding occurring within the entire primer sequence.<br>
It is the maximum allowable local alignment score when testing
a single primer for (local) self-complementarity.  Local
self-complementarity is taken to predict the tendency of primers
to anneal to each other without necessarily causing self-priming
in the PCR.  The scoring system gives 1.00 for complementary
bases, -0.25 for a match of any base (or N) with an N, -1.00 for
a mismatch, and -2.00 for a gap. Only single-base-pair gaps are
allowed. For example, the alignment</p>
<pre>
   5' ATCGNA 3'
      || | |
   3' TA-CGT 5'
</pre>
is allowed (and yields a score of 1.75), but the alignment
<pre>
   5' ATCCGNA 3'
      ||  | |
   3' TA--CGT 5'
</pre>
<p>is not considered.  Scores are non-negative, and a score of 0.00
indicates that there is no reasonable local alignment between two
oligos.</p>
<h3><a id="PRIMER_MAX_SELF_ANY_TH">PRIMER_MAX_SELF_ANY_TH (decimal, 9999,99; default 47.00)</a></h3>
<p> The same as PRIMER_MAX_SELF_ANY but all calculations are based on
	  thermodynamical approach. The melting temperature of the most stable
	  structure is calculated. To calculate secondary structures nearest-neighbor
	  parameters for perfect matches, single internal mismatches, terminal
	  mismatches, dangling ends have been used. Also parameters for
increments for length dependence of
bulge and internal loops have been used. This parameter is calculated only
if PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT=1. The default value is 10 degrees
lower than the default value of PRIMER_MIN_TM. For example, the alignment
width length 15nt</p>
<pre>
  5' ATTAGATAGAGCATC 3'
  3' TAATCTATCTCGTAG 5'
</pre>
is allowed (and yields a melting temperature of 32.1493
width by default Primer3 parameters), but the alignment
<pre>
     T        C
  5'  GCGGCCGC GCGC 3'
  3'  CGCCGGCG CGCG 5'
     A        A
</pre>
<p>is not considered (Tm=57.0997 and the length of oligo is 14nt).
Thermodynamical parameters and methods for finding
the most stable structure are described in following papers:</p>
<ul><li>[SantaLucia JR (1998) "A unified view of polymer, dumbbell and
oligonucleotide DNA nearest-neighbor
thermodynamics", Proc Natl Acad Sci 95:1460-65
<a href="http://dx.doi.org/10.1073/pnas.95.4.1460">http://dx.doi.org/10.1073/pnas.95.4.1460</a>]
</li><li>
[SantaLucia JR and Hicks D (2004) "The thermodynamics of DNA structural
motifs", Annu Rev Biophys Biomol Struct 33:415-40
<a href="http://dx.doi.org/10.1146/annurev.biophys.32.110601.141800">http://dx.doi.org/10.1146/annurev.biophys.32.110601.141800</a>]
</li><li>
[Bommarito S, Peyret N and SantaLucia J Jr (2000) "Thermodynamic parameters
for DNA sequences with dangling ends", Nucleic Acids Res 28(9):1929-34
<a href="http://dx.doi.org/10.1093/nar/28.9.1929">http://dx.doi.org/10.1093/nar/28.9.1929</a>]
</li><li>
[Peyret N, Seneviratne PA, Allawi HT, SantaLucia J Jr. (1999)
"Nearest-neighbor thermodynamics and NMR of DNA sequences with internal A.A,
C.C, G.G, and T.T mismatches", Biochemistry 38(12):3468-77
<a href="http://dx.doi.org/10.1021/bi9825091">http://dx.doi.org/10.1021/bi9825091</a>]
</li><li>
[Allawi HT and SantaLucia J Jr. (1998) "Nearest-Neighbor Thermodynamics of
Internal A&#183;C Mismatches in DNA:  Sequence Dependence and pH Effects",
Biochemistry 37(26):9435-44
<a href="http://dx.doi.org/10.1021/bi9803729">http://dx.doi.org/10.1021/bi9803729</a>
</li><li>
[Allawi HT and SantaLucia J Jr. (1998) "Thermodynamics of internal C.T
mismatches in DNA." Nucleic Acids Res 26(11):2694-701<a href="http://dx.doi.org/10.1093/nar/26.11.2694">http://dx.doi.org/10.1093/nar/26.11.2694</a>]
</li><li>
[Allawi HT and SantaLucia J Jr. (1998) "Nearest neighbor thermodynamic
parameters for internal G.A mismatches in DNA." Biochemistry 37(8):2170-9
<a href="http://dx.doi.org/10.1021/bi9724873">http://dx.doi.org/10.1021/bi9724873</a>]
</li><li>
[Allawi HT and SantaLucia J Jr. (1997) "Thermodynamics and NMR of internal
G.T mismatches in DNA." Biochemistry 36(34):10581-94
<a href="http://dx.doi.org/10.1021/bi962590c">http://dx.doi.org/10.1021/bi962590c</a>]
</li><li>
[SantaLucia J Jr and Peyret N. (2001) "Method and system for predicting
nucleic acid hybridization thermodynamics and computer-readable storage
medium for use therein" World Intellectual Property Organization, WO 01/94611
<a href="http://www.wipo.int/pctdb/en/wo.jsp?wo=2001094611">http://www.wipo.int/pctdb/en/wo.jsp?wo=2001094611</a>]
</li></ul>
<p><br>
Predicting secondary structures can improve primer design by eliminating
sequences with high possibility to form alternative secondary structures.
	  </p>
<h3><a id="PRIMER_INTERNAL_MAX_SELF_ANY">PRIMER_INTERNAL_MAX_SELF_ANY (decimal, 9999.99; default 12.00)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MAX_SELF_ANY">PRIMER_MAX_SELF_ANY</a> for the internal oligo.</p>
<h3><a id="PRIMER_INTERNAL_MAX_SELF_ANY_TH">PRIMER_INTERNAL_MAX_SELF_ANY_TH (decimal, 9999.99; default 47.00)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MAX_SELF_ANY_TH">PRIMER_MAX_SELF_ANY_TH</a> for the internal oligo.</p>
<h3><a id="PRIMER_PAIR_MAX_COMPL_ANY">PRIMER_PAIR_MAX_COMPL_ANY (decimal, 9999.99; default 8.00)</a></h3>
<p>PRIMER_PAIR_MAX_COMPL_ANY describes the tendency of the left primer
to bind to the right primer. It is the maximum allowable local alignment score when testing for
complementarity between left and right primers. It is similar to <a href="#PRIMER_MAX_SELF_ANY">PRIMER_MAX_SELF_ANY</a>.</p>
<h3><a id="PRIMER_PAIR_MAX_COMPL_ANY_TH">PRIMER_PAIR_MAX_COMPL_ANY_TH (decimal, 9999.99; default 47.00)</a></h3>
<p>PRIMER_PAIR_MAX_COMPL_ANY_TH describes the tendency of the left primer
    to bind to the right primer. It is similar to <a href="#PRIMER_MAX_SELF_ANY_TH">PRIMER_MAX_SELF_ANY_TH</a>.</p>
<h3><a id="PRIMER_WT_SELF_ANY">PRIMER_WT_SELF_ANY (float; default 0.0)</a></h3>
<p>Penalty weight for the individual primer self binding value as in
<a href="#PRIMER_MAX_SELF_ANY">PRIMER_MAX_SELF_ANY</a>.</p>
<h3><a id="PRIMER_WT_SELF_ANY_TH">PRIMER_WT_SELF_ANY_TH (float; default 0.0)</a></h3>
<p>Penalty weight for the individual primer self binding value as in
     <a href="#PRIMER_MAX_SELF_ANY_TH">PRIMER_MAX_SELF_ANY_TH</a>.</p>
<h3><a id="PRIMER_INTERNAL_WT_SELF_ANY">PRIMER_INTERNAL_WT_SELF_ANY (float; default 0.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_SELF_ANY">PRIMER_WT_SELF_ANY</a> for the internal oligo.</p>
<h3><a id="PRIMER_INTERNAL_WT_SELF_ANY_TH">PRIMER_INTERNAL_WT_SELF_ANY_TH (float; default 0.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_SELF_ANY_TH">PRIMER_WT_SELF_ANY_TH</a> for the internal oligo.</p>
<h3><a id="PRIMER_PAIR_WT_COMPL_ANY">PRIMER_PAIR_WT_COMPL_ANY (float; default 0.0)</a></h3>
<p>Penalty weight for the binding value of the primer pair as in
<a href="#PRIMER_MAX_SELF_ANY">PRIMER_MAX_SELF_ANY</a>.</p>
<h3><a id="PRIMER_PAIR_WT_COMPL_ANY_TH">PRIMER_PAIR_WT_COMPL_ANY_TH (float; default 0.0)</a></h3>
<p>Penalty weight for the binding value of the primer pair as in <a href="#PRIMER_MAX_SELF_ANY_TH">PRIMER_MAX_SELF_ANY_TH</a>.</p>
<h3><a id="PRIMER_MAX_SELF_END">PRIMER_MAX_SELF_END (decimal, 9999.99; default 3.00)</a></h3>
<p>
PRIMER_MAX_SELF_END tries to bind the 3'-END to a identical primer
and scores the best binding it can find. This is critical for
primer quality because it allows primers use itself as a target
and amplify a short piece (forming a primer-dimer). These primers
are then unable to bind and amplify the target sequence.<br>
PRIMER_MAX_SELF_END is the maximum allowable 3'-anchored global
alignment score when testing a single primer for
self-complementarity. The 3'-anchored global alignment score
is taken to predict the likelihood of PCR-priming primer-dimers,
for example</p>
<pre>
   5' ATGCCCTAGCTTCCGGATG 3'
                ||| |||||
             3' AAGTCCTACATTTAGCCTAGT 5'
</pre>
or
<pre>
   5` AGGCTATGGGCCTCGCGA 3'
                  ||||||
               3' AGCGCTCCGGGTATCGGA 5'
</pre>
<p>The scoring system is as for the Maximum Complementarity
argument. In the examples above the scores are 7.00 and 6.00
respectively. Scores are non-negative, and a score of 0.00
indicates that there is no reasonable 3'-anchored global
alignment between two oligos.  In order to estimate 3'-anchored
global alignments for candidate primers, Primer3
assumes that the sequence from which to choose primers is
presented 5'->3'. It is nonsensical to provide a larger value
for this parameter than for the Maximum (local) Complementarity
parameter (<a href="#PRIMER_MAX_SELF_ANY">PRIMER_MAX_SELF_ANY</a>) because the score of a local
alignment will always be at least as great as the score of a
global alignment.</p>
<h3><a id="PRIMER_MAX_SELF_END_TH">PRIMER_MAX_SELF_END_TH (decimal 9999.99; default 47.00)</a></h3>
<p>Same as <a href="#PRIMER_MAX_SELF_END">PRIMER_MAX_SELF_END</a> but is based on
thermodynamical approach - the stability of structure is analyzed. The value
of tag is expressed as melting temperature. See
<a href="#PRIMER_MAX_SELF_ANY_TH">PRIMER_MAX_SELF_ANY_TH</a> for details.</p>
<h3><a id="PRIMER_INTERNAL_MAX_SELF_END">PRIMER_INTERNAL_MAX_SELF_END (decimal 9999.99; default 12.00)</a></h3>
<p><a href="#PRIMER_INTERNAL_MAX_SELF_END">PRIMER_INTERNAL_MAX_SELF_END</a> is meaningless when applied
to internal oligos used for hybridization-based detection, since
primer-dimer will not occur.  We recommend that
<a href="#PRIMER_INTERNAL_MAX_SELF_END">PRIMER_INTERNAL_MAX_SELF_END</a> be set at least as high as
<a href="#PRIMER_INTERNAL_MAX_SELF_ANY">PRIMER_INTERNAL_MAX_SELF_ANY</a>.</p>
<h3><a id="PRIMER_INTERNAL_MAX_SELF_END_TH">PRIMER_INTERNAL_MAX_SELF_END_TH (decimal 9999.99; default 47.00)</a></h3>
<p>Same as <a href="#PRIMER_INTERNAL_MAX_SELF_END">PRIMER_INTERNAL_MAX_SELF_END</a> but for
calculating the score (melting temperature of structure) thermodynamical
approach is used.</p>
<h3><a id="PRIMER_PAIR_MAX_COMPL_END">PRIMER_PAIR_MAX_COMPL_END (decimal, 9999.99; default 3.00)</a></h3>
<p>PRIMER_PAIR_MAX_COMPL_END tries to bind the 3'-END of the left primer
to the right primer and scores the best binding it can find.
It is similar to <a href="#PRIMER_MAX_SELF_END">PRIMER_MAX_SELF_END</a>.</p>
<h3><a id="PRIMER_PAIR_MAX_COMPL_END_TH">PRIMER_PAIR_MAX_COMPL_END_TH (decimal, 9999.99; default 47.00)</a></h3>
<p>Same as <a href="#PRIMER_PAIR_MAX_COMPL_END">PRIMER_PAIR_MAX_COMPL_END</a> but for calculating the
score (melting temperature of structure) thermodynamical approach is used.</p>
<h3><a id="PRIMER_WT_SELF_END">PRIMER_WT_SELF_END (float; default 0.0)</a></h3>
<p>Penalty weight for the individual primer self binding value as in
<a href="#PRIMER_MAX_SELF_END">PRIMER_MAX_SELF_END</a>.</p>
<h3><a id="PRIMER_WT_SELF_END_TH">PRIMER_WT_SELF_END_TH (float; default 0.0)</a></h3>
<p>Penalty weight for the individual primer self binding
value as in <a href="#PRIMER_MAX_SELF_END_TH">PRIMER_MAX_SELF_END_TH</a></p>
<h3><a id="PRIMER_INTERNAL_WT_SELF_END">PRIMER_INTERNAL_WT_SELF_END (float; default 0.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_SELF_END">PRIMER_WT_SELF_END</a> for the internal oligo.</p>
<h3><a id="PRIMER_INTERNAL_WT_SELF_END_TH">PRIMER_INTERNAL_WT_SELF_END_TH (float; default 0.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_SELF_END_TH">PRIMER_WT_SELF_END_TH</a> for the internal oligo.</p>
<h3><a id="PRIMER_PAIR_WT_COMPL_END">PRIMER_PAIR_WT_COMPL_END (float; default 0.0)</a></h3>
<p>Penalty weight for the binding value of the primer pair as in
<a href="#PRIMER_MAX_SELF_END">PRIMER_MAX_SELF_END</a>.</p>
<h3><a id="PRIMER_PAIR_WT_COMPL_END_TH">PRIMER_PAIR_WT_COMPL_END_TH (float; default 0.0)</a></h3>
<p>Penalty weight for the binding value of the primer pair as in <a href="#PRIMER_MAX_SELF_END_TH">PRIMER_MAX_SELF_END_TH</a>.</p>
<h3><a id="PRIMER_MAX_HAIRPIN_TH">PRIMER_MAX_HAIRPIN_TH (float; default 47.0)</a></h3>
<p>This is the most stable monomer structure of internal oligo
calculated by thermodynamic approach. The hairpin loops,
bulge loops, internal loops, internal single mismatches, dangling ends,
terminal mismatches have been considered. This parameter is calculated only
if PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT=1. The default value is 10 degrees
lower than the default value of PRIMER_MIN_TM. For example the structure:</p>
<pre>
      -///------\\\-
   5' ACGCTGTGCTGCGA 3'
</pre>
with melting temperature 53.7263 (calculated according to by default values
of Primer3) and
<pre>
      //////----\\\\\\
   5' CCGCAGTAAGCTGCGG 3'
</pre>
<p>
with melting temperature 71.0918 (calculated according to by default values
of Primer3)
For details about papers used for calculating hairpins see <a href="#PRIMER_MAX_SELF_ANY_TH">PRIMER_MAX_SELF_ANY_TH</a></p>
<h3><a id="PRIMER_INTERNAL_MAX_HAIRPIN_TH">PRIMER_INTERNAL_MAX_HAIRPIN_TH (float; default 47.0)</a></h3>
<p>The most stable monomer structure of internal oligo
calculated by thermodynamic approach. See <a href="#PRIMER_MAX_HAIRPIN_TH">PRIMER_MAX_HAIRPIN_TH</a> for details.</p>
<h3><a id="PRIMER_WT_HAIRPIN_TH">PRIMER_WT_HAIRPIN_TH (float; default 0.0)</a></h3>
<p>Penalty weight for the individual primer hairpin structure value as in <a href="#PRIMER_MAX_HAIRPIN_TH">PRIMER_MAX_HAIRPIN_TH</a>.</p>
<h3><a id="PRIMER_INTERNAL_WT_HAIRPIN_TH">PRIMER_INTERNAL_WT_HAIRPIN_TH (float; default 0.0)</a></h3>
<p>Penalty weight for the most stable
primer hairpin structure value as in <a href="#PRIMER_INTERNAL_MAX_HAIRPIN_TH">PRIMER_INTERNAL_MAX_HAIRPIN_TH</a>.</p>
<h3><a id="PRIMER_MAX_END_STABILITY">PRIMER_MAX_END_STABILITY (float, 999.9999; default 100.0)</a></h3>
<p>The maximum stability for the last five 3' bases of a left or
right primer.  Bigger numbers mean more stable 3' ends.  The
value is the maximum delta G (kcal/mol) for duplex disruption for
the five 3' bases as calculated using the nearest-neighbor
parameter values specified by the option of <a href="#PRIMER_TM_FORMULA">PRIMER_TM_FORMULA</a>
<br>
<br>
For example if the table of thermodynamic parameters suggested
by <a href="http://dx.doi.org/10.1073/pnas.95.4.1460">SantaLucia 1998, DOI:10.1073/pnas.95.4.1460</a> is used the deltaG
values for the most stable and for the most labile 5mer duplex
are 6.86 kcal/mol (GCGCG) and 0.86 kcal/mol (TATAT) respectively.<br>
<br>
If the table of thermodynamic parameters suggested by
<a href="http://dx.doi.org/10.1073/pnas.83.11.3746">Breslauer et al. 1986, 10.1073/pnas.83.11.3746</a> is used the deltaG
values for the most stable and for the most labile 5mer are
13.4 kcal/mol (GCGCG) and 4.6 kcal/mol (TATAC) respectively.
<h3><a id="PRIMER_WT_END_STABILITY">PRIMER_WT_END_STABILITY (float; default 0.0)</a></h3>
<p>Penalty factor for the calculated maximum stability for the
last five 3' bases of a left or right primer.</p>
<h3><a id="PRIMER_MAX_NS_ACCEPTED">PRIMER_MAX_NS_ACCEPTED (int; default 0)</a></h3>
<p>Maximum number of unknown bases (N) allowable in any primer.</p>
<h3><a id="PRIMER_INTERNAL_MAX_NS_ACCEPTED">PRIMER_INTERNAL_MAX_NS_ACCEPTED (int; default 0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MAX_NS_ACCEPTED">PRIMER_MAX_NS_ACCEPTED</a> for the internal oligo.</p>
<h3><a id="PRIMER_WT_NUM_NS">PRIMER_WT_NUM_NS (float; default 0.0)</a></h3>
<p>Penalty weight for the number of Ns in the primer.</p>
<h3><a id="PRIMER_INTERNAL_WT_NUM_NS">PRIMER_INTERNAL_WT_NUM_NS (float; default 0.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_NUM_NS">PRIMER_WT_NUM_NS</a> for the internal oligo.</p>
<h3><a id="PRIMER_MAX_POLY_X">PRIMER_MAX_POLY_X (int; default 5)</a></h3>
<p>The maximum allowable length of a mononucleotide repeat,
for example AAAAAA.</p>
<h3><a id="PRIMER_INTERNAL_MAX_POLY_X">PRIMER_INTERNAL_MAX_POLY_X (int; default 5)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MAX_POLY_X">PRIMER_MAX_POLY_X</a> for the internal oligo.</p>
<h3><a id="PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE">PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE (int; default -1)</a></h3>
<p>When returning multiple primer pairs,
the minimum number of base pairs between the 3' ends of
any two left primers.<br>
<br>
Primers with 3' ends at positions e.g.  30 and 31 in the template
sequence have a three-prime distance of 1.<br>
<br>
In addition to positive values, the values -1 and 0
are acceptable and have special interpretations:
<br>
<br>
-1 indicates that a given left primer can appear in
multiple primer pairs returned by Primer3.
This is the default behavior.
<br>
<br>
0 indicates that a left primer is acceptable if it
was not already used.
In other words, two left primers are allowed to
have the same 3' position provided their 5' positions differ.
<br>
<br>
For <i>n</i> &gt; 0:
A left primer is acceptable if:<br>
<br>
NOT(3' end of left primer closer than <i>n</i> to the 3' end of a previously
used left primer)<br></p>
<h3><a id="PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE">PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE (int; default -1)</a></h3>
<p>Analogous to <a href="#PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE">PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE</a>,
but for right primers.<br></p>
<h3><a id="PRIMER_MIN_THREE_PRIME_DISTANCE">PRIMER_MIN_THREE_PRIME_DISTANCE (int; default -1)</a></h3>
<p>A "convenience" tag that simultaneously
sets
<a href="#PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE">PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE</a>
and
<a href="#PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE">PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE</a>
<br>
<br>
For example</p>
<pre>
PRIMER_MIN_THREE_PRIME_DISTANCE=3
</pre>
is equivalent to
<pre>
PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE=3
PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE=3
</pre>
<p>
It is an error to specify both
PRIMER_MIN_THREE_PRIME_DISTANCE and either
<a href="#PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE">PRIMER_MIN_LEFT_THREE_PRIME_DISTANCE</a> or
<a href="#PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE">PRIMER_MIN_RIGHT_THREE_PRIME_DISTANCE</a>
in the same input record.</p>
<h3><a id="PRIMER_PICK_ANYWAY">PRIMER_PICK_ANYWAY (boolean; default 0)</a></h3>
<p>If true use primer provided in <a href="#SEQUENCE_PRIMER">SEQUENCE_PRIMER</a>, <a href="#SEQUENCE_PRIMER_REVCOMP">SEQUENCE_PRIMER_REVCOMP</a>,
or <a href="#SEQUENCE_INTERNAL_OLIGO">SEQUENCE_INTERNAL_OLIGO</a> even if it violates specific
constraints.</p>
<h3><a id="PRIMER_LOWERCASE_MASKING">PRIMER_LOWERCASE_MASKING (int; default 0)</a></h3>
<p>This option allows for intelligent design of primers in sequence in
which masked regions (for example repeat-masked regions) are
lower-cased.  (New in v. 1.1.0, added by Maido Remm and Triinu
Koressaar)<br>
<br>
A value of 1 directs Primer3 to reject primers overlapping
lowercase a base exactly at the 3' end.<br>
<br>
This property relies on the assumption that masked features
(e.g. repeats) can partly overlap primer, but they cannot overlap
the 3'-end of the primer.  In other words, lowercase bases at
other positions in the primer are accepted, assuming that the
masked features do not influence the primer performance if they
do not overlap the 3'-end of primer.</p>
<h3><a id="PRIMER_EXPLAIN_FLAG">PRIMER_EXPLAIN_FLAG (boolean; default 0)</a></h3>
<p>If this flag is 1 (non-0), produce <a href="#PRIMER_LEFT_EXPLAIN">PRIMER_LEFT_EXPLAIN</a>,
<a href="#PRIMER_RIGHT_EXPLAIN">PRIMER_RIGHT_EXPLAIN</a>, <a href="#PRIMER_INTERNAL_EXPLAIN">PRIMER_INTERNAL_EXPLAIN</a> and/or
<a href="#PRIMER_PAIR_EXPLAIN">PRIMER_PAIR_EXPLAIN</a> output
tags as appropriate.
These output tags are intended to provide information on the number of
oligos and primer pairs that Primer3 examined and counts of
the number discarded for various reasons.  If -format_output is
set similar information is produced in the user-oriented output.</p>
<h3><a id="PRIMER_LIBERAL_BASE">PRIMER_LIBERAL_BASE (boolean; default 0)</a></h3>
<p>This parameter provides a quick-and-dirty way to get Primer3 to
accept IUB / IUPAC codes for ambiguous bases (i.e. by changing
all unrecognized bases to N).  If you wish to include an
ambiguous base in an oligo, you must set <a href="#PRIMER_MAX_NS_ACCEPTED">PRIMER_MAX_NS_ACCEPTED</a> to a
1 (non-0) value.<br>
<br>
Perhaps '-' and '* ' should be squeezed out rather than changed
to 'N', but currently they simply get converted to N's.  The authors
invite user comments.</p>
<h3><a id="PRIMER_FIRST_BASE_INDEX">PRIMER_FIRST_BASE_INDEX (int; default 0)</a></h3>
<p>This parameter is the index of the first base in the input
sequence.  For input and output using 1-based indexing (such as
that used in GenBank and to which many users are accustomed) set
this parameter to 1.  For input and output using 0-based indexing
set this parameter to 0.  (This parameter also affects the
indexes in the contents of the files produced when the primer
file flag is set.)</p>
<h3><a id="PRIMER_MAX_TEMPLATE_MISPRIMING">PRIMER_MAX_TEMPLATE_MISPRIMING (decimal, 9999.99; default -1.00)</a></h3>
<p>The maximum allowed similarity to ectopic sites in the
template.  A negative value means do not check.  The scoring
system is the same as used for <a href="#PRIMER_MAX_LIBRARY_MISPRIMING">PRIMER_MAX_LIBRARY_MISPRIMING</a>, except
that an ambiguity code in the template is never treated as a
consensus (see <a href="#PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS</a>).</p>
<h3><a id="PRIMER_MAX_TEMPLATE_MISPRIMING_TH">PRIMER_MAX_TEMPLATE_MISPRIMING_TH (decimal, 9999.99; default -1.00)</a></h3>
<p>Similar to <a href="#PRIMER_MAX_TEMPLATE_MISPRIMING">PRIMER_MAX_TEMPLATE_MISPRIMING</a> but
assesses alternative binding sites in the template using thermodynamic models
(when PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT=1).
This parameter specifies the maximum allowed
melting temperature of an oligo (primer)
at an "ectopic" site within the template sequence; 47.0 would be
a reasonable choice if <a href="#PRIMER_MIN_TM">PRIMER_MIN_TM</a> is 57.0.
</p>
<h3><a id="PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING">PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING (decimal, 9999.99; default -1.00)</a></h3>
<p>The maximum allowed summed similarity of both primers to
ectopic sites in the template. A negative value means do not
check.  The scoring system is the same as used for
<a href="#PRIMER_PAIR_MAX_LIBRARY_MISPRIMING">PRIMER_PAIR_MAX_LIBRARY_MISPRIMING</a>, except that an ambiguity code
in the template is never treated as a consensus (see
<a href="#PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS</a>).  Primer3 does not
check the similarity of hybridization oligos (internal
oligos) to locations outside of the amplicon.</p>
<h3><a id="PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING_TH">PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING_TH (decimal, 9999.99; default -1.00)</a></h3>
<p>The maximum allowed summed melting temperatures
of both primers at ectopic sites within the template (with
the two primers in an orientation that would allow PCR
amplification.)
The melting temperatures are calculated
as for <a href="#PRIMER_MAX_TEMPLATE_MISPRIMING_TH">PRIMER_MAX_TEMPLATE_MISPRIMING_TH</a>.
<h3><a id="PRIMER_WT_TEMPLATE_MISPRIMING">PRIMER_WT_TEMPLATE_MISPRIMING (float; default 0.0)</a></h3>
<p>Penalty for a single primer binding to the template sequence.<br>
<br>
The use of this Tag is modified from Primer3 version 2.0 on:
The values used with the older versions have to be multiplied
by the factor 100 to have the same effect.</p>
<h3><a id="PRIMER_WT_TEMPLATE_MISPRIMING_TH">PRIMER_WT_TEMPLATE_MISPRIMING_TH (float; default 0.0)</a></h3>
<p>Penalty for a single primer binding to the template sequence (thermodynamic approach, when PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT=1).</p>
<h3><a id="PRIMER_PAIR_WT_TEMPLATE_MISPRIMING">PRIMER_PAIR_WT_TEMPLATE_MISPRIMING (float; default 0.0)</a></h3>
<p>Penalty for a primer pair binding to the template sequence.<br>
<br>
The use of this Tag is modified from Primer3 version 2.0 on:
The values used with the older versions have to be multiplied
by the factor 100 to have the same effect.</p>
<h3><a id="PRIMER_PAIR_WT_TEMPLATE_MISPRIMING_TH">PRIMER_PAIR_WT_TEMPLATE_MISPRIMING_TH (float; default 0.0)</a></h3>
<p>Penalty for a primer pair binding to the template sequence
(thermodynamic approach, when PRIMER_THERMODYNAMIC_TEMPLATE_ALIGNMENT=1).</p>
<h3><a id="PRIMER_MISPRIMING_LIBRARY">PRIMER_MISPRIMING_LIBRARY (string; default empty)</a></h3>
<p>The name of a file containing a nucleotide sequence library of
sequences to avoid amplifying (for example repetitive sequences, or
possibly the sequences of genes in a gene family that should
not be amplified.)  The file must be in (a slightly restricted)
FASTA format (W. B. Pearson and D.J. Lipman, PNAS 85:8 pp
2444-2448 [1988]); we briefly discuss the organization of this
file below.  If this parameter is specified then Primer3 locally
aligns each candidate primer against each library sequence and
rejects those primers for which the local alignment score times a
specified weight (see below) exceeds <a href="#PRIMER_MAX_LIBRARY_MISPRIMING">PRIMER_MAX_LIBRARY_MISPRIMING</a>.
(The maximum value of the weight is arbitrarily set to 100.0.)<br>
<br>
Each sequence entry in the FASTA-format file must begin with an
"id line" that starts with '>'.  The contents of the id line is
"slightly restricted" in that Primer3 parses everything after any
optional asterisk ('*') as a floating point number to use as the
weight mentioned above.  If the id line contains no asterisk then
the weight defaults to 1.0.  The alignment scoring system used is
the same as for calculating complementarity among oligos (e.g.
<a href="#PRIMER_MAX_SELF_ANY">PRIMER_MAX_SELF_ANY</a>), except for the handling of IUB/IUPAC ambiguity
codes (discussed below).<br>
<br>
The remainder of an entry contains the sequence as lines
following the id line up until a line starting with '>' or
the end of the file.  Whitespace and newlines are ignored.
Characters 'A', 'T', 'G', 'C', 'a', 't', 'g', 'c' and
IUB/IUPAC 'ambiguity' codes ('R, 'Y', 'K', 'M', 'S', 'W',
'N', including lower case) are retained. For technical
reasons the length of the sequence must be &gt;= 3. Of course,
sequences of length &lt; 10 or so are probably useless, but
will be accepted without complaint.<br>
<br>
WARNING: always set <a href="#PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS</a>=0
if any sequence in the library contains strings of 'N's:
NNNNNNNNNNNNNNNNNNNN.<br>
NOWWW<br>
There are no restrictions on line length.<br>
<br>
An empty value for this parameter indicates that no repeat
library should be used and "turns off" the use of a
previously specified library.<br>
<br>
Repbase (J. Jurka, A.F.A. Smit, C. Pethiyagoda, and
others, 1995-1996, <a href="ftp://ftp.ncbi.nih.gov/repository/repbase/">ftp://ftp.ncbi.nih.gov/repository/repbase/</a>)
is an excellent source of repeat sequences and pointers to the
literature. (The Repbase files need to be converted to Fasta
format before they can be used by Primer3.)<br>
<br>
See <a href="#providedMisprimingLibs">providedMisprimingLibs</a> for the sequence
libraries available on this server.</p>
<h3><a id="PRIMER_INTERNAL_MISHYB_LIBRARY">PRIMER_INTERNAL_MISHYB_LIBRARY (string; default empty)</a></h3>
<p>Similar to <a href="#PRIMER_MISPRIMING_LIBRARY">PRIMER_MISPRIMING_LIBRARY</a>, except that the event we
seek to avoid is hybridization of the internal oligo to sequences
in this library rather than priming from them.</p>
<h3><a id="PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS">PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS (boolean; default 0)</a></h3>
<p>If set to 1, treat ambiguity codes as if they were consensus
codes when matching oligos to mispriming or mishyb
libraries. For example, if this flag is set, then a C in an
oligo will be scored as a perfect match to an S in a library
sequence, as will a G in the oligo. More importantly,
though, any base in an oligo will be scored as a perfect
match to an N in the library.  This is very bad if the
library contains strings of Ns, as no oligo will be legal
(and it will take a long time to find this out). So unless
you know for sure that your library does not have runs of Ns
(or Xs), then set this flag to 0.</p>
<h3><a id="PRIMER_MAX_LIBRARY_MISPRIMING">PRIMER_MAX_LIBRARY_MISPRIMING (decimal, 9999.99; default 12.00)</a></h3>
<p>The maximum allowed weighted similarity with any sequence in
<a href="#PRIMER_MISPRIMING_LIBRARY">PRIMER_MISPRIMING_LIBRARY</a>.</p>
<h3><a id="PRIMER_INTERNAL_MAX_LIBRARY_MISHYB">PRIMER_INTERNAL_MAX_LIBRARY_MISHYB (decimal,9999.99; default 12.00)</a></h3>
<p>Similar to <a href="#PRIMER_MAX_LIBRARY_MISPRIMING">PRIMER_MAX_LIBRARY_MISPRIMING</a> except that this parameter applies
to the similarity of candidate internal oligos to the library
specified in <a href="#PRIMER_INTERNAL_MISHYB_LIBRARY">PRIMER_INTERNAL_MISHYB_LIBRARY</a>.</p>
<h3><a id="PRIMER_PAIR_MAX_LIBRARY_MISPRIMING">PRIMER_PAIR_MAX_LIBRARY_MISPRIMING (decimal, 9999.99; default 24.00)</a></h3>
<p>The maximum allowed sum of similarities of a primer pair
(one similarity for each primer) with any single sequence in
<a href="#PRIMER_MISPRIMING_LIBRARY">PRIMER_MISPRIMING_LIBRARY</a>.
Library sequence weights are not used in computing the sum
of similarities.</p>
<h3><a id="PRIMER_WT_LIBRARY_MISPRIMING">PRIMER_WT_LIBRARY_MISPRIMING (float; default 0.0)</a></h3>
<p>Penalty for a single primer binding to any single sequence
in <a href="#PRIMER_MISPRIMING_LIBRARY">PRIMER_MISPRIMING_LIBRARY</a>.</p>
<h3><a id="PRIMER_INTERNAL_WT_LIBRARY_MISHYB">PRIMER_INTERNAL_WT_LIBRARY_MISHYB (float; default 0.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_LIBRARY_MISPRIMING">PRIMER_WT_LIBRARY_MISPRIMING</a> for the internal oligo.</p>
<h3><a id="PRIMER_PAIR_WT_LIBRARY_MISPRIMING">PRIMER_PAIR_WT_LIBRARY_MISPRIMING (float; default 0.0)</a></h3>
<p>Penalty for a primer pair binding to any single sequence
in <a href="#PRIMER_MISPRIMING_LIBRARY">PRIMER_MISPRIMING_LIBRARY</a>.</p>
<h3><a id="PRIMER_MASK_TEMPLATE">PRIMER_MASK_TEMPLATE (boolean; default 0)</a></h3>
<p>This feature helps to prevent designing primers to template
regions that are repetitive. Primers with more binding
sites tend to have higher failure rates. The masking is
based on statistical model, which calculates the
probability of failure P<sub>f</sub> as follows:</p>
<pre>P<sub>f</sub>= e<sup>m</sup> / (1 + e<sup>m</sup>),</pre>
<p>where m = 0.1772 * K11 + 0.239 * K16 - 4.336 <br>
and K11 and K16 are frequencies of 11-mers and 16-mers
in given genome. The frequencies are stored in
species-specific k-mer list files. Users can build their
own k-mer lists for species of interest. GenomeTester4
software for making properly formatted k-mer lists can be
downloaded from GitHub:
<a href="https://github.com/bioinfo-ut/GenomeTester4">https://github.com/bioinfo-ut/GenomeTester4</a>.</p>
<h3><a id="PRIMER_MASK_FAILURE_RATE">PRIMER_MASK_FAILURE_RATE (float; default 0.1)</a></h3>
<p>Cutoff value of accepted failure rate for masking
algorithm. Higher value gives lower stringency, meaning
that fewer nucleotides in target sequence is masked.</p>
<h3><a id="PRIMER_WT_MASK_FAILURE_RATE">PRIMER_WT_MASK_FAILURE_RATE (float; default 0.0)</a></h3>
<p>Penalty weight for the primer failure rate.</p>
<h3><a id="PRIMER_MASK_5P_DIRECTION">PRIMER_MASK_5P_DIRECTION (int; default 1)</a></h3>
<p>The number of nucleotides masking algorithm should mask
from 5' direction.</p>
<h3><a id="PRIMER_MASK_3P_DIRECTION">PRIMER_MASK_3P_DIRECTION (int; default 0)</a></h3>
<p>The number of nucleotides masking algorithm should mask
from 3' direction.</p>
<h3><a id="PRIMER_MASK_KMERLIST_PATH">PRIMER_MASK_KMERLIST_PATH (string; default ../kmer_lists/)</a></h3>
<p>This tag specifies the path to the directory that
contains k-mer list files for masking algorithm. Required
for command-line execution. On web interface the species
is selected from drop-down menu</p>
<h3><a id="PRIMER_MASK_KMERLIST_PREFIX">PRIMER_MASK_KMERLIST_PREFIX (string; default homo_sapiens)</a></h3>
<p>This tag specifies the species whose k-mer lists are used for
pre-masking.</p>
<h3><a id="PRIMER_MIN_QUALITY">PRIMER_MIN_QUALITY (int; default 0)</a></h3>
<p>The minimum sequence quality (as specified by
<a href="#SEQUENCE_QUALITY">SEQUENCE_QUALITY</a>) allowed within a primer.</p>
<h3><a id="PRIMER_INTERNAL_MIN_QUALITY">PRIMER_INTERNAL_MIN_QUALITY (int; default 0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_MIN_QUALITY">PRIMER_MIN_QUALITY</a> for the internal oligo.</p>
<h3><a id="PRIMER_MIN_END_QUALITY">PRIMER_MIN_END_QUALITY (int; default 0)</a></h3>
<p>The minimum sequence quality (as specified by
<a href="#SEQUENCE_QUALITY">SEQUENCE_QUALITY</a>) allowed within the 5' pentamer of a
primer. Note that there is no PRIMER_INTERNAL_MIN_END_QUALITY.</p>
<h3><a id="PRIMER_QUALITY_RANGE_MIN">PRIMER_QUALITY_RANGE_MIN (int; default 0)</a></h3>
<p>The minimum legal sequence quality (used for error checking
of <a href="#PRIMER_MIN_QUALITY">PRIMER_MIN_QUALITY</a> and <a href="#PRIMER_MIN_END_QUALITY">PRIMER_MIN_END_QUALITY</a>).</p>
<h3><a id="PRIMER_QUALITY_RANGE_MAX">PRIMER_QUALITY_RANGE_MAX (int; default 100)</a></h3>
<p>The maximum legal sequence quality (used for error checking
of <a href="#PRIMER_MIN_QUALITY">PRIMER_MIN_QUALITY</a> and <a href="#PRIMER_MIN_END_QUALITY">PRIMER_MIN_END_QUALITY</a>).</p>
<h3><a id="PRIMER_WT_SEQ_QUAL">PRIMER_WT_SEQ_QUAL (float; default 0.0)</a></h3>
<p>Penalty weight for the sequence quality of the primer.</p>
<h3><a id="PRIMER_INTERNAL_WT_SEQ_QUAL">PRIMER_INTERNAL_WT_SEQ_QUAL (float; default 0.0)</a></h3>
<p>Equivalent parameter of <a href="#PRIMER_WT_SEQ_QUAL">PRIMER_WT_SEQ_QUAL</a> for the internal oligo.</p>
<h3><a id="PRIMER_PAIR_WT_PR_PENALTY">PRIMER_PAIR_WT_PR_PENALTY (float; default 1.0)</a></h3>
<p>Penalty factor for the sum of the left and the right primer added
to the pair penalty. Setting this value below 1.0 will increase
running time.<br>
<br>
As <a href="#PRIMER_PAIR_WT_PR_PENALTY">PRIMER_PAIR_WT_PR_PENALTY</a> or the per-primer
penalties it multiplies become lower with respect to various
pair penalties (for example <a href="#PRIMER_PAIR_WT_PRODUCT_SIZE_LT">PRIMER_PAIR_WT_PRODUCT_SIZE_LT</a>
<a href="#PRIMER_PAIR_WT_PRODUCT_SIZE_GT">PRIMER_PAIR_WT_PRODUCT_SIZE_GT</a>
<a href="#PRIMER_PAIR_WT_DIFF_TM">PRIMER_PAIR_WT_DIFF_TM</a>, etc.) the running time of the
search for primer pairs is likely to grow substantially. The
reason is that the search algorithm must calculate the penalty
for more primer pairs (as opposed to excluding them based on
the penalties of the individual oligos).
</p>
<h3><a id="PRIMER_PAIR_WT_IO_PENALTY">PRIMER_PAIR_WT_IO_PENALTY (float; default 0.0)</a></h3>
<p>Penalty factor for the internal oligo added to the pair penalty.</p>
<h3><a id="PRIMER_INSIDE_PENALTY">PRIMER_INSIDE_PENALTY (float; default -1.0)</a></h3>
<p>Non-default values are valid only for sequences with 0 or 1
target regions.  If the primer is part of a pair that spans a
target and overlaps the target, then multiply this value times
the number of nucleotide positions by which the primer overlaps
the (unique) target to get the 'position penalty'.  The effect of
this parameter is to allow Primer3 to include overlap with the
target as a term in the objective function.</p>
<h3><a id="PRIMER_OUTSIDE_PENALTY">PRIMER_OUTSIDE_PENALTY (float; default 0.0)</a></h3>
<p>Non-default values are valid only for sequences with 0 or 1
target regions.  If the primer is part of a pair that spans a
target and does not overlap the target, then multiply this value
times the number of nucleotide positions from the 3' end to the
(unique) target to get the 'position penalty'.  The effect of
this parameter is to allow Primer3 to include nearness to the
target as a term in the objective function.</p>
<h3><a id="PRIMER_WT_POS_PENALTY">PRIMER_WT_POS_PENALTY (float; default 1.0)</a></h3>
<p>Penalty for the primer which do not overlap the target.</p>
<h3><a id="PRIMER_SEQUENCING_LEAD">PRIMER_SEQUENCING_LEAD (int; default 50)</a></h3>
<p>Defines the space from the 3'end of the primer to the point
were the trace signals are readable. Value only used if
<a href="#PRIMER_TASK">PRIMER_TASK</a>=pick_sequencing_primers.</p>
<h3><a id="PRIMER_SEQUENCING_SPACING">PRIMER_SEQUENCING_SPACING (int; default 500)</a></h3>
<p>Defines the space from the 3'end of the primer to the 3'end of
the next primer on the same strand. Value only used if
<a href="#PRIMER_TASK">PRIMER_TASK</a>=pick_sequencing_primers.</p>
<h3><a id="PRIMER_SEQUENCING_INTERVAL">PRIMER_SEQUENCING_INTERVAL (int; default 250)</a></h3>
<p>Defines the space from the 3'end of the primer to the 3'end of
the next primer on the reverse strand. Value only used if
<a href="#PRIMER_TASK">PRIMER_TASK</a>=pick_sequencing_primers.</p>
<h3><a id="PRIMER_SEQUENCING_ACCURACY">PRIMER_SEQUENCING_ACCURACY (int; default 20)</a></h3>
<p>Defines the space from the calculated position of the 3'end to
both sides in which Primer3Plus picks the best primer. Value only used if
<a href="#PRIMER_TASK">PRIMER_TASK</a>=pick_sequencing_primers.</p>
<h3><a id="PRIMER_WT_END_QUAL">PRIMER_WT_END_QUAL (float; default 0.0)</a></h3>
<p> </p>
<h3><a id="PRIMER_INTERNAL_WT_END_QUAL">PRIMER_INTERNAL_WT_END_QUAL (float; default 0.0)</a></h3>
<p> </p>
<h2><a id="programTags">18. "PROGRAM" INPUT TAGS</a></h2>
<p>"Program" input tags start with P3_... describe the
parameters that deal with the behavior of the Primer3 program
itself.)</p>
<h3><a id="P3_FILE_ID">P3_FILE_ID (string; default empty)</a></h3>
<p>
This tag is only valid in Primer3 setting files. It should be used to
identify the purpose of the settings files it appears in.
It is always printed out on the output of Primer3.
See also also the command line flag -echo_settings_file,
which causes the entire settings file to be echoed in primer3_core's
output.
</p>
<h3><a id="P3_FILE_TYPE">P3_FILE_TYPE (string; default sequence)</a></h3>
<p>
This tag is only valid in the header of Primer3 setting files. It
should be used in the second line to identify the type files it
appears in. Valid options are sequence, settings or all. In
Primer3Plus it can control which parameters are loaded.
</p>
<h3><a id="P3_FILE_FLAG">P3_FILE_FLAG (boolean; default 0)</a></h3>
<p>If the associated value = 1 (non-0), then Primer3 creates
one or more output
files for each input <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>.  File &lt;sequence_id&gt;.for lists all
acceptable left primers for &lt;sequence_id&gt;, and &lt;sequence_id&gt;.rev
lists all acceptable right primers for &lt;sequence_id&gt;, where
&lt;sequence_id&gt; is the value of the <a href="#SEQUENCE_ID">SEQUENCE_ID</a> tag (which
must be supplied).  If internal oligos are requested, Primer3
 produces a file &lt;sequence_id&gt;.int, which lists all
acceptable internal oligos.  Obviously, &lt;sequence_id&gt; needs
to be a string that will work as a file name.
<br>
<br>
See also the <code>pick_primer_list</code>
argument to <a href="#PRIMER_TASK">PRIMER_TASK</a>, which offers similar functionality on stdout.
</p>
<h3><a id="P3_COMMENT">P3_COMMENT (string; default empty)</a></h3>
<p>The value of this tag is ignored. It can be used to annotate input.</p>
<h2><a id="calculatePenalties">19. HOW PRIMER3 CALCULATES THE PENALTY VALUE</a></h2>
<p>
In essense, the penalty values define what is the best primer
pair.
The calculation of penalty values
takes into consideration penalty weights, which allow
one to fine-tune the selection of primers to specific needs.<br>
<br>
This section will explain the selection process of primers by
Primer3. In general the selection is a multi step process:<br>
<br>
In the first step, Primer3 evaluates every primer that
can be picked in the region of interest, possibly subject to
constraints due to target regions, product size ranges, and
so forth, that might preclude the use of primers in the
eventually selected primer pairs. In this pass the hard
limits are tested like <a href="#PRIMER_MAX_GC">PRIMER_MAX_GC</a> or <a href="#PRIMER_MIN_TM">PRIMER_MIN_TM</a>. Primers
with a GC lower than <a href="#PRIMER_MAX_GC">PRIMER_MAX_GC</a> or a Tm higher than
<a href="#PRIMER_MIN_TM">PRIMER_MIN_TM</a> are memorized, the primers which fail in one of
these tests are excluded. Primer3 can be forced to use primers
failing to pass this test by setting <a href="#PRIMER_PICK_ANYWAY">PRIMER_PICK_ANYWAY</a> to one
(only available for primers provided by the user).<br>
<br>
In the second step, Primer3 calculates a penalty for each
primer. This penalty is the only score by which Primer3
evaluates the primers It is also provided as output
<a href="#PRIMER_LEFT_4_PENALTY">PRIMER_LEFT_4_PENALTY</a>, <a href="#PRIMER_INTERNAL_4_PENALTY">PRIMER_INTERNAL_4_PENALTY</a> and
<a href="#PRIMER_RIGHT_4_PENALTY">PRIMER_RIGHT_4_PENALTY</a> (shown for the primer set 4). For each
primer, it is calculated like that:</p>
<h3>Left Primers:</h3>
<pre>
<a href="#PRIMER_LEFT_4_PENALTY">PRIMER_LEFT_4_PENALTY</a> =
   If <a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> &gt; <a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a> then this is added (+):
       + <a href="#PRIMER_WT_TM_GT">PRIMER_WT_TM_GT</a> * ( <a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - <a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a> )
   If <a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> &lt; <a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a> then this is added (+):
       + <a href="#PRIMER_WT_TM_LT">PRIMER_WT_TM_LT</a> * ( <a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a> - <a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> )
   If <a href="#PRIMER_LEFT_4_GC_PERCENT">PRIMER_LEFT_4_GC_PERCENT</a> &gt; <a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a> then
       this is added (+):
       + <a href="#PRIMER_WT_GC_PERCENT_GT">PRIMER_WT_GC_PERCENT_GT</a> *
            ( <a href="#PRIMER_LEFT_4_GC_PERCENT">PRIMER_LEFT_4_GC_PERCENT</a> - <a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a> )
   If <a href="#PRIMER_LEFT_4_GC_PERCENT">PRIMER_LEFT_4_GC_PERCENT</a> &lt; <a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a> then
       this is added (+):
       + <a href="#PRIMER_WT_GC_PERCENT_LT">PRIMER_WT_GC_PERCENT_LT</a> *
            ( <a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a> - <a href="#PRIMER_LEFT_4_GC_PERCENT">PRIMER_LEFT_4_GC_PERCENT</a> )
   If masking is used (<a href="#PRIMER_MASK_TEMPLATE">PRIMER_MASK_TEMPLATE</a>=1), then this is added (+):
       + <a href="#PRIMER_WT_MASK_FAILURE_RATE">PRIMER_WT_MASK_FAILURE_RATE</a> * PRIMER_LEFT_4_MASK_FAILURE_RATE
   The following section uses &lt;primer length&gt; as part of
   the term which is given as output in
   <a href="#PRIMER_LEFT_4">PRIMER_LEFT_4</a>=position,&lt;primer length&gt;
   If &lt;primer length&gt; &gt; <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a> then
       this is added (+):
       + <a href="#PRIMER_WT_SIZE_GT">PRIMER_WT_SIZE_GT</a> *
            ( &lt;primer length&gt; - <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a> )
   If &lt;primer length&gt; &lt; <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a> then
       this is added (+):
       + <a href="#PRIMER_WT_SIZE_LT">PRIMER_WT_SIZE_LT</a> *
            ( <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a> - &lt;primer length&gt; )
   If the primer does not overlap a target then
       this is added (+):
       + <a href="#PRIMER_WT_POS_PENALTY">PRIMER_WT_POS_PENALTY</a> * <a href="#PRIMER_LEFT_4_POSITION_PENALTY">PRIMER_LEFT_4_POSITION_PENALTY</a>
   These are allways added (+) to the penalty
   (if the thermodynamic approach is used then the part in <i>italic</i>
      is substituted with text below this calculation):
    <i>+ <a href="#PRIMER_WT_SELF_ANY">PRIMER_WT_SELF_ANY</a> * <a href="#PRIMER_LEFT_4_SELF_ANY">PRIMER_LEFT_4_SELF_ANY</a>
       + <a href="#PRIMER_WT_SELF_END">PRIMER_WT_SELF_END</a> * <a href="#PRIMER_LEFT_4_SELF_END">PRIMER_LEFT_4_SELF_END</a>
       + <a href="#PRIMER_WT_TEMPLATE_MISPRIMING">PRIMER_WT_TEMPLATE_MISPRIMING</a> *
            <a href="#PRIMER_LEFT_4_TEMPLATE_MISPRIMING">PRIMER_LEFT_4_TEMPLATE_MISPRIMING</a></i>
       + <a href="#PRIMER_WT_END_STABILITY">PRIMER_WT_END_STABILITY</a> * <a href="#PRIMER_LEFT_4_END_STABILITY">PRIMER_LEFT_4_END_STABILITY</a>
       + <a href="#PRIMER_WT_NUM_NS">PRIMER_WT_NUM_NS</a> * &lt;numbers of N in the selected primer&gt;
       + <a href="#PRIMER_WT_LIBRARY_MISPRIMING">PRIMER_WT_LIBRARY_MISPRIMING</a> * <a href="#PRIMER_LEFT_4_LIBRARY_MISPRIMING">PRIMER_LEFT_4_LIBRARY_MISPRIMING</a>
       + <a href="#PRIMER_WT_SEQ_QUAL">PRIMER_WT_SEQ_QUAL</a> *
            ( <a href="#PRIMER_QUALITY_RANGE_MAX">PRIMER_QUALITY_RANGE_MAX</a> -
              <a href="#PRIMER_LEFT_4_MIN_SEQ_QUALITY">PRIMER_LEFT_4_MIN_SEQ_QUALITY</a> )
      If the thermodynamic approach is used then the part of <i>italic</i> in
          the above calculation is replaced by this:
 <i>
     If ((<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5) &le; <a href="#PRIMER_LEFT_4_SELF_ANY_TH">PRIMER_LEFT_4_SELF_ANY_TH</a>) then is added (+):
      + <a href="#PRIMER_WT_SELF_ANY_TH">PRIMER_WT_SELF_ANY_TH</a> *
                  (<a href="#PRIMER_LEFT_4_SELF_ANY_TH">PRIMER_LEFT_4_SELF_ANY_TH</a> - (<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5 - 1))
    else if ((<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5) &gt; <a href="#PRIMER_LEFT_4_SELF_ANY_TH">PRIMER_LEFT_4_SELF_ANY_TH</a>) then is added (+):
      + <a href="#PRIMER_WT_SELF_ANY_TH">PRIMER_WT_SELF_ANY_TH</a> *
            (1/(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5 + 1 - <a href="#PRIMER_LEFT_4_SELF_ANY_TH">PRIMER_LEFT_4_SELF_ANY_TH</a>));
    If ((<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5) &le; <a href="#PRIMER_LEFT_4_SELF_END_TH">PRIMER_LEFT_4_SELF_END_TH</a>) then is added (+):
     + <a href="#PRIMER_WT_SELF_END_TH">PRIMER_WT_SELF_END_TH</a> *
               (<a href="#PRIMER_LEFT_4_SELF_END_TH">PRIMER_LEFT_4_SELF_END_TH</a> - (<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5 - 1))
    else if ((<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5) &gt; <a href="#PRIMER_LEFT_4_SELF_END_TH">PRIMER_LEFT_4_SELF_END_TH</a>) then is added (+):
     + <a href="#PRIMER_WT_SELF_END_TH">PRIMER_WT_SELF_END_TH</a> *
       (1/(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5 + 1 - <a href="#PRIMER_LEFT_4_SELF_ANY_TH">PRIMER_LEFT_4_SELF_ANY_TH</a>));
    If ((<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5) &le; <a href="#PRIMER_LEFT_4_TEMPLATE_MISPRIMING_TH">PRIMER_LEFT_4_TEMPLATE_MISPRIMING_TH</a>) then is added (+):
     + <a href="#PRIMER_WT_TEMPLATE_MISPRIMING_TH">PRIMER_WT_TEMPLATE_MISPRIMING_TH</a> *
       (<a href="#PRIMER_LEFT_4_TEMPLATE_MISPRIMING_TH">PRIMER_LEFT_4_TEMPLATE_MISPRIMING_TH</a> - (<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5 - 1))
    else if ((<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5) &gt; <a href="#PRIMER_LEFT_4_TEMPLATE_MISPRIMING_TH">PRIMER_LEFT_4_TEMPLATE_MISPRIMING_TH</a>) then is added (+):
     + <a href="#PRIMER_WT_TEMPLATE_MISPRIMING_TH">PRIMER_WT_TEMPLATE_MISPRIMING_TH</a> *
      (1/(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5 + 1 - <a href="#PRIMER_LEFT_4_TEMPLATE_MISPRIMING_TH">PRIMER_LEFT_4_TEMPLATE_MISPRIMING_TH</a>));
    If ((<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5) &le; <a href="#PRIMER_LEFT_4_HAIRPIN_TH">PRIMER_LEFT_4_HAIRPIN_TH</a>) then is added (+):
      + <a href="#PRIMER_WT_HAIRPIN_TH">PRIMER_WT_HAIRPIN_TH</a> *
             (<a href="#PRIMER_LEFT_4_HAIRPIN_TH">PRIMER_LEFT_4_HAIRPIN_TH</a> - (<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5 - 1))
    else if ((<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5) &gt; <a href="#PRIMER_LEFT_4_HAIRPIN_TH">PRIMER_LEFT_4_HAIRPIN_TH</a>) then is added (+):
      + <a href="#PRIMER_WT_HAIRPIN_TH">PRIMER_WT_HAIRPIN_TH</a> *
            (1/(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a> - 5 + 1 - <a href="#PRIMER_LEFT_4_HAIRPIN_TH">PRIMER_LEFT_4_HAIRPIN_TH</a>));
 </i>
</pre>
<h3>Right Primers (identical to Left Primers):</h3>
<pre>
<a href="#PRIMER_RIGHT_4_PENALTY">PRIMER_RIGHT_4_PENALTY</a> =
   If <a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> &gt; <a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a> then
       this is added (+):
       + <a href="#PRIMER_WT_TM_GT">PRIMER_WT_TM_GT</a> * ( <a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - <a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a> )
   If <a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> &lt; <a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a> then
       this is added (+):
       + <a href="#PRIMER_WT_TM_LT">PRIMER_WT_TM_LT</a> * ( <a href="#PRIMER_OPT_TM">PRIMER_OPT_TM</a> - <a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> )
   If <a href="#PRIMER_RIGHT_4_GC_PERCENT">PRIMER_RIGHT_4_GC_PERCENT</a> &gt; <a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a> then
       this is added (+):
       + <a href="#PRIMER_WT_GC_PERCENT_GT">PRIMER_WT_GC_PERCENT_GT</a> *
            ( <a href="#PRIMER_RIGHT_4_GC_PERCENT">PRIMER_RIGHT_4_GC_PERCENT</a> - <a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a> )
   If <a href="#PRIMER_RIGHT_4_GC_PERCENT">PRIMER_RIGHT_4_GC_PERCENT</a> &lt; <a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a> then
       this is added (+):
       + <a href="#PRIMER_WT_GC_PERCENT_LT">PRIMER_WT_GC_PERCENT_LT</a> *
            ( <a href="#PRIMER_OPT_GC_PERCENT">PRIMER_OPT_GC_PERCENT</a> - <a href="#PRIMER_RIGHT_4_GC_PERCENT">PRIMER_RIGHT_4_GC_PERCENT</a> )
   If masking is used (<a href="#PRIMER_MASK_TEMPLATE">PRIMER_MASK_TEMPLATE</a>=1), then this is added (+):
       + <a href="#PRIMER_WT_MASK_FAILURE_RATE">PRIMER_WT_MASK_FAILURE_RATE</a> * PRIMER_RIGHT_4_MASK_FAILURE_RATE
   The following section uses &lt;primer length&gt; as part of
   the term which is given as output in
   <a href="#PRIMER_RIGHT_4">PRIMER_RIGHT_4</a>=position,&lt;primer length&gt;
   If &lt;primer length&gt; &gt; <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a> then
       this is added (+):
       + <a href="#PRIMER_WT_SIZE_GT">PRIMER_WT_SIZE_GT</a> *
            ( &lt;primer length&gt; - <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a> )
   If &lt;primer length&gt; &lt; <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a> then
       this is added (+):
       + <a href="#PRIMER_WT_SIZE_LT">PRIMER_WT_SIZE_LT</a> *
            ( <a href="#PRIMER_OPT_SIZE">PRIMER_OPT_SIZE</a> - &lt;primer length&gt; )
   If the primer does not overlap a target then
       this is added (+):
       + <a href="#PRIMER_WT_POS_PENALTY">PRIMER_WT_POS_PENALTY</a> * <a href="#PRIMER_RIGHT_4_POSITION_PENALTY">PRIMER_RIGHT_4_POSITION_PENALTY</a>
   These are allways added (+) to the penalty
   (if the thermodynamic approach is used then the part in <i>italic</i>
      is substituted with text below this calculation):
   <i> + <a href="#PRIMER_WT_SELF_ANY">PRIMER_WT_SELF_ANY</a> * <a href="#PRIMER_RIGHT_4_SELF_ANY">PRIMER_RIGHT_4_SELF_ANY</a>
       + <a href="#PRIMER_WT_SELF_END">PRIMER_WT_SELF_END</a> * <a href="#PRIMER_RIGHT_4_SELF_END">PRIMER_RIGHT_4_SELF_END</a>
       + <a href="#PRIMER_WT_TEMPLATE_MISPRIMING">PRIMER_WT_TEMPLATE_MISPRIMING</a> *
            <a href="#PRIMER_RIGHT_4_TEMPLATE_MISPRIMING">PRIMER_RIGHT_4_TEMPLATE_MISPRIMING</a></i>
       + <a href="#PRIMER_WT_END_STABILITY">PRIMER_WT_END_STABILITY</a> * <a href="#PRIMER_RIGHT_4_END_STABILITY">PRIMER_RIGHT_4_END_STABILITY</a>
       + <a href="#PRIMER_WT_NUM_NS">PRIMER_WT_NUM_NS</a> * &lt;numbers of N in the selected primer&gt;
       + <a href="#PRIMER_WT_LIBRARY_MISPRIMING">PRIMER_WT_LIBRARY_MISPRIMING</a> * <a href="#PRIMER_RIGHT_4_LIBRARY_MISPRIMING">PRIMER_RIGHT_4_LIBRARY_MISPRIMING</a>
       + <a href="#PRIMER_WT_SEQ_QUAL">PRIMER_WT_SEQ_QUAL</a> *
            ( <a href="#PRIMER_QUALITY_RANGE_MAX">PRIMER_QUALITY_RANGE_MAX</a> -
              <a href="#PRIMER_RIGHT_4_MIN_SEQ_QUALITY">PRIMER_RIGHT_4_MIN_SEQ_QUALITY</a> )
     If the thermodynamic approach is used then the part of <i>italic</i> in
         the above calculation is replaced by this:
<i>
    If ((<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &le; <a href="#PRIMER_RIGHT_4_SELF_ANY_TH">PRIMER_RIGHT_4_SELF_ANY_TH</a>) then is added (+):
      + <a href="#PRIMER_WT_SELF_ANY_TH">PRIMER_WT_SELF_ANY_TH</a> *
                  (<a href="#PRIMER_RIGHT_4_SELF_ANY_TH">PRIMER_RIGHT_4_SELF_ANY_TH</a> - (<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 - 1))
    else if ((<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &gt; <a href="#PRIMER_RIGHT_4_SELF_ANY_TH">PRIMER_RIGHT_4_SELF_ANY_TH</a>) then is added (+):
      + <a href="#PRIMER_WT_SELF_ANY_TH">PRIMER_WT_SELF_ANY_TH</a> *
            (1/(<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 + 1 - <a href="#PRIMER_RIGHT_4_SELF_ANY_TH">PRIMER_RIGHT_4_SELF_ANY_TH</a>));
    If ((<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &le; <a href="#PRIMER_RIGHT_4_SELF_END_TH">PRIMER_RIGHT_4_SELF_END_TH</a>) then is added (+):
     + <a href="#PRIMER_WT_SELF_END_TH">PRIMER_WT_SELF_END_TH</a> *
               (<a href="#PRIMER_RIGHT_4_SELF_END_TH">PRIMER_RIGHT_4_SELF_END_TH</a> - (<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 - 1))
    else if ((<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &gt; <a href="#PRIMER_RIGHT_4_SELF_END_TH">PRIMER_RIGHT_4_SELF_END_TH</a>) then is added (+):
     + <a href="#PRIMER_WT_SELF_END_TH">PRIMER_WT_SELF_END_TH</a> *
               (1/(<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 + 1 - <a href="#PRIMER_RIGHT_4_SELF_ANY_TH">PRIMER_RIGHT_4_SELF_ANY_TH</a>));
    If ((<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &le; <a href="#PRIMER_RIGHT_4_TEMPLATE_MISPRIMING_TH">PRIMER_RIGHT_4_TEMPLATE_MISPRIMING_TH</a>) then is added (+):
      + <a href="#PRIMER_WT_TEMPLATE_MISPRIMING_TH">PRIMER_WT_TEMPLATE_MISPRIMING_TH</a> *
       (<a href="#PRIMER_RIGHT_4_TEMPLATE_MISPRIMING_TH">PRIMER_RIGHT_4_TEMPLATE_MISPRIMING_TH</a> - (<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 - 1))
    else if ((<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &gt; <a href="#PRIMER_RIGHT_4_TEMPLATE_MISPRIMING_TH">PRIMER_RIGHT_4_TEMPLATE_MISPRIMING_TH</a>) then is added (+):
      + <a href="#PRIMER_WT_TEMPLATE_MISPRIMING_TH">PRIMER_WT_TEMPLATE_MISPRIMING_TH</a> *
      (1/(<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 + 1 - <a href="#PRIMER_RIGHT_4_TEMPLATE_MISPRIMING_TH">PRIMER_RIGHT_4_TEMPLATE_MISPRIMING_TH</a>));
    If ((<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &le; <a href="#PRIMER_RIGHT_4_HAIRPIN_TH">PRIMER_RIGHT_4_HAIRPIN_TH</a>) then is added (+):
      + <a href="#PRIMER_WT_HAIRPIN_TH">PRIMER_WT_HAIRPIN_TH</a> *
             (<a href="#PRIMER_RIGHT_4_HAIRPIN_TH">PRIMER_RIGHT_4_HAIRPIN_TH</a> - (<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 - 1))
    else if ((<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &gt; <a href="#PRIMER_RIGHT_4_HAIRPIN_TH">PRIMER_RIGHT_4_HAIRPIN_TH</a>) then is added (+):
      + <a href="#PRIMER_WT_HAIRPIN_TH">PRIMER_WT_HAIRPIN_TH</a> *
            (1/(<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 + 1 - <a href="#PRIMER_RIGHT_4_HAIRPIN_TH">PRIMER_RIGHT_4_HAIRPIN_TH</a>));
</i>
</pre>
<h3>Internal Oligos:</h3>
<pre>
<a href="#PRIMER_INTERNAL_4_PENALTY">PRIMER_INTERNAL_4_PENALTY</a> =
   If <a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> &gt; <a href="#PRIMER_INTERNAL_OPT_TM">PRIMER_INTERNAL_OPT_TM</a> then
       this is added (+):
       + <a href="#PRIMER_INTERNAL_WT_TM_GT">PRIMER_INTERNAL_WT_TM_GT</a> *
            ( <a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - <a href="#PRIMER_INTERNAL_OPT_TM">PRIMER_INTERNAL_OPT_TM</a> )
   If <a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> &lt; <a href="#PRIMER_INTERNAL_OPT_TM">PRIMER_INTERNAL_OPT_TM</a> then
       this is added (+):
       + <a href="#PRIMER_INTERNAL_WT_TM_LT">PRIMER_INTERNAL_WT_TM_LT</a> *
            ( <a href="#PRIMER_INTERNAL_OPT_TM">PRIMER_INTERNAL_OPT_TM</a> - <a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> )
   If <a href="#PRIMER_INTERNAL_4_GC_PERCENT">PRIMER_INTERNAL_4_GC_PERCENT</a> &gt; <a href="#PRIMER_INTERNAL_OPT_GC_PERCENT">PRIMER_INTERNAL_OPT_GC_PERCENT</a>
       then this is added (+):
       + <a href="#PRIMER_INTERNAL_WT_GC_PERCENT_GT">PRIMER_INTERNAL_WT_GC_PERCENT_GT</a> *
            ( <a href="#PRIMER_INTERNAL_4_GC_PERCENT">PRIMER_INTERNAL_4_GC_PERCENT</a> -
              <a href="#PRIMER_INTERNAL_OPT_GC_PERCENT">PRIMER_INTERNAL_OPT_GC_PERCENT</a> )
   If <a href="#PRIMER_INTERNAL_4_GC_PERCENT">PRIMER_INTERNAL_4_GC_PERCENT</a> &lt; <a href="#PRIMER_INTERNAL_OPT_GC_PERCENT">PRIMER_INTERNAL_OPT_GC_PERCENT</a>
       then this is added (+):
       + <a href="#PRIMER_INTERNAL_WT_GC_PERCENT_LT">PRIMER_INTERNAL_WT_GC_PERCENT_LT</a> *
            ( <a href="#PRIMER_INTERNAL_OPT_GC_PERCENT">PRIMER_INTERNAL_OPT_GC_PERCENT</a> -
              <a href="#PRIMER_INTERNAL_4_GC_PERCENT">PRIMER_INTERNAL_4_GC_PERCENT</a> )
   The following section uses &lt;primer length&gt; as part of
   the term which is given as output in
   <a href="#PRIMER_INTERNAL_4">PRIMER_INTERNAL_4</a>=position,&lt;primer length&gt;
   If &lt;primer length&gt; &gt; <a href="#PRIMER_INTERNAL_OPT_SIZE">PRIMER_INTERNAL_OPT_SIZE</a> then
       this is added (+):
       + <a href="#PRIMER_INTERNAL_WT_SIZE_GT">PRIMER_INTERNAL_WT_SIZE_GT</a> *
            ( &lt;primer length&gt; - <a href="#PRIMER_INTERNAL_OPT_SIZE">PRIMER_INTERNAL_OPT_SIZE</a> )
   If &lt;primer length&gt; &lt; <a href="#PRIMER_INTERNAL_OPT_SIZE">PRIMER_INTERNAL_OPT_SIZE</a> then
       this is added (+):
       + <a href="#PRIMER_INTERNAL_WT_SIZE_LT">PRIMER_INTERNAL_WT_SIZE_LT</a> *
            ( <a href="#PRIMER_INTERNAL_OPT_SIZE">PRIMER_INTERNAL_OPT_SIZE</a> - &lt;primer length&gt; )
   These are always added (+) to the penalty:
   (if the thermodynamic approach is used then the part in <i>italic</i>
      is substituted with text below this calculation):
  <i>+ <a href="#PRIMER_INTERNAL_WT_SELF_ANY">PRIMER_INTERNAL_WT_SELF_ANY</a> * <a href="#PRIMER_INTERNAL_4_SELF_ANY">PRIMER_INTERNAL_4_SELF_ANY</a>
     + <a href="#PRIMER_INTERNAL_WT_SELF_END">PRIMER_INTERNAL_WT_SELF_END</a> * <a href="#PRIMER_INTERNAL_4_SELF_END">PRIMER_INTERNAL_4_SELF_END</a></i>
     + <a href="#PRIMER_INTERNAL_WT_NUM_NS">PRIMER_INTERNAL_WT_NUM_NS</a> *
          &lt;numbers of N in the selected primer&gt;
     + <a href="#PRIMER_INTERNAL_WT_LIBRARY_MISHYB">PRIMER_INTERNAL_WT_LIBRARY_MISHYB</a> *
          <a href="#PRIMER_INTERNAL_4_LIBRARY_MISHYB">PRIMER_INTERNAL_4_LIBRARY_MISHYB</a>
     + <a href="#PRIMER_INTERNAL_WT_SEQ_QUAL">PRIMER_INTERNAL_WT_SEQ_QUAL</a> *
          ( <a href="#PRIMER_QUALITY_RANGE_MAX">PRIMER_QUALITY_RANGE_MAX</a> -
            <a href="#PRIMER_INTERNAL_4_MIN_SEQ_QUALITY">PRIMER_INTERNAL_4_MIN_SEQ_QUALITY</a> )
   If the thermodynamic approach is used then the part of <i>italic</i> in
   the above calculation is replaced by this:
<i>
   If ((<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5) &le; <a href="#PRIMER_INTERNAL_4_SELF_ANY_TH">PRIMER_INTERNAL_4_SELF_ANY_TH</a>) then is added (+):
    + <a href="#PRIMER_INTERNAL_WT_SELF_ANY_TH">PRIMER_INTERNAL_WT_SELF_ANY_TH</a> *
        (<a href="#PRIMER_INTERNAL_4_SELF_ANY_TH">PRIMER_INTERNAL_4_SELF_ANY_TH</a> - (<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5 - 1))
    else if ((<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5) &gt; <a href="#PRIMER_INTERNAL_4_SELF_ANY_TH">PRIMER_INTERNAL_4_SELF_ANY_TH</a>) then is added (+):
    + <a href="#PRIMER_INTERNAL_WT_SELF_ANY_TH">PRIMER_INTERNAL_WT_SELF_ANY_TH</a> *
        (1/(<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5 + 1 - <a href="#PRIMER_INTERNAL_4_SELF_ANY_TH">PRIMER_INTERNAL_4_SELF_ANY_TH</a>));
    If ((<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5) &le; <a href="#PRIMER_INTERNAL_4_SELF_END_TH">PRIMER_INTERNAL_4_SELF_END_TH</a>) then is added (+):
     + <a href="#PRIMER_INTERNAL_WT_SELF_END_TH">PRIMER_INTERNAL_WT_SELF_END_TH</a> *
         (<a href="#PRIMER_INTERNAL_4_SELF_END_TH">PRIMER_INTERNAL_4_SELF_END_TH</a> - (<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5 - 1))
    else if ((<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5) &gt; <a href="#PRIMER_INTERNAL_4_SELF_END_TH">PRIMER_INTERNAL_4_SELF_END_TH</a>) then is added (+):
      + <a href="#PRIMER_INTERNAL_WT_SELF_END_TH">PRIMER_INTERNAL_WT_SELF_END_TH</a> *
          (1/(<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5 + 1 - <a href="#PRIMER_INTERNAL_4_SELF_ANY_TH">PRIMER_INTERNAL_4_SELF_ANY_TH</a>));
    If ((<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5) &le; <a href="#PRIMER_INTERNAL_4_HAIRPIN_TH">PRIMER_INTERNAL_4_HAIRPIN_TH</a>) then is added (+):
      + <a href="#PRIMER_INTERNAL_WT_HAIRPIN_TH">PRIMER_INTERNAL_WT_HAIRPIN_TH</a> *
        (<a href="#PRIMER_INTERNAL_4_HAIRPIN_TH">PRIMER_INTERNAL_4_HAIRPIN_TH</a> - (<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5 - 1))
    else if ((<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5) &gt; <a href="#PRIMER_INTERNAL_4_HAIRPIN_TH">PRIMER_INTERNAL_4_HAIRPIN_TH</a>) then is added (+):
      + <a href="#PRIMER_INTERNAL_WT_HAIRPIN_TH">PRIMER_INTERNAL_WT_HAIRPIN_TH</a> *
        (1/(<a href="#PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM</a> - 5 + 1 - <a href="#PRIMER_INTERNAL_4_HAIRPIN_TH">PRIMER_INTERNAL_4_HAIRPIN_TH</a>));
</i>
</pre>
<p>
The primers are then sorted by penalty and Primer3 tries to pick
the primers with the lowest penalty. For the <a href="#PRIMER_TASK">PRIMER_TASK</a>
<pre><i>pick_primer_list</i></pre> or <pre><i>pick_sequencing_primers</i></pre> the selection ends at
this point. If primer pairs have to be selected, a
<a href="#PRIMER_PAIR_4_PENALTY">PRIMER_PAIR_4_PENALTY</a> is calculated:
<pre>
<a href="#PRIMER_PAIR_4_PENALTY">PRIMER_PAIR_4_PENALTY</a> =
   To the pair penalty are at first the single primer penalties
   added (+):
       + <a href="#PRIMER_PAIR_WT_PR_PENALTY">PRIMER_PAIR_WT_PR_PENALTY</a> *
            ( <a href="#PRIMER_LEFT_4_PENALTY">PRIMER_LEFT_4_PENALTY</a> + <a href="#PRIMER_RIGHT_4_PENALTY">PRIMER_RIGHT_4_PENALTY</a> )
   If internal oligo is picked then this is added (+):
       + <a href="#PRIMER_PAIR_WT_IO_PENALTY">PRIMER_PAIR_WT_IO_PENALTY</a> * <a href="#PRIMER_INTERNAL_4_PENALTY">PRIMER_INTERNAL_4_PENALTY</a>
   If <a href="#PRIMER_PAIR_4_PRODUCT_TM">PRIMER_PAIR_4_PRODUCT_TM</a> &gt; <a href="#PRIMER_PRODUCT_OPT_TM">PRIMER_PRODUCT_OPT_TM</a> then
       this is added (+):
       + <a href="#PRIMER_PAIR_WT_PRODUCT_TM_GT">PRIMER_PAIR_WT_PRODUCT_TM_GT</a> *
            ( <a href="#PRIMER_PAIR_4_PRODUCT_TM">PRIMER_PAIR_4_PRODUCT_TM</a> - <a href="#PRIMER_PRODUCT_OPT_TM">PRIMER_PRODUCT_OPT_TM</a> )
   If <a href="#PRIMER_PAIR_4_PRODUCT_TM">PRIMER_PAIR_4_PRODUCT_TM</a> &lt; <a href="#PRIMER_PRODUCT_OPT_TM">PRIMER_PRODUCT_OPT_TM</a> then
       this is added (+):
       + <a href="#PRIMER_PAIR_WT_PRODUCT_TM_LT">PRIMER_PAIR_WT_PRODUCT_TM_LT</a> *
            ( <a href="#PRIMER_PRODUCT_OPT_TM">PRIMER_PRODUCT_OPT_TM</a> - <a href="#PRIMER_PAIR_4_PRODUCT_TM">PRIMER_PAIR_4_PRODUCT_TM</a> )
   If <a href="#PRIMER_PAIR_4_PRODUCT_SIZE">PRIMER_PAIR_4_PRODUCT_SIZE</a> &gt; <a href="#PRIMER_PRODUCT_OPT_SIZE">PRIMER_PRODUCT_OPT_SIZE</a> then
       this is added (+):
       + <a href="#PRIMER_PAIR_WT_PRODUCT_SIZE_GT">PRIMER_PAIR_WT_PRODUCT_SIZE_GT</a> *
            ( <a href="#PRIMER_PAIR_4_PRODUCT_SIZE">PRIMER_PAIR_4_PRODUCT_SIZE</a> - <a href="#PRIMER_PRODUCT_OPT_SIZE">PRIMER_PRODUCT_OPT_SIZE</a> )
   If <a href="#PRIMER_PAIR_4_PRODUCT_SIZE">PRIMER_PAIR_4_PRODUCT_SIZE</a> &lt; <a href="#PRIMER_PRODUCT_OPT_SIZE">PRIMER_PRODUCT_OPT_SIZE</a> then
       this is added (+):
       + <a href="#PRIMER_PAIR_WT_PRODUCT_SIZE_LT">PRIMER_PAIR_WT_PRODUCT_SIZE_LT</a> *
            ( <a href="#PRIMER_PRODUCT_OPT_SIZE">PRIMER_PRODUCT_OPT_SIZE</a> - <a href="#PRIMER_PAIR_4_PRODUCT_SIZE">PRIMER_PAIR_4_PRODUCT_SIZE</a> )
   These are allways added (+) to the penalty:
   (if the thermodynamic approach is used then the part in <i>italic</i>
       is substituted with text below this calculation):
     + <a href="#PRIMER_PAIR_WT_DIFF_TM">PRIMER_PAIR_WT_DIFF_TM</a> *
          &lt;difference in Tm between the left and the right primer&gt;
  <i>+ <a href="#PRIMER_PAIR_WT_COMPL_ANY">PRIMER_PAIR_WT_COMPL_ANY</a> * <a href="#PRIMER_PAIR_4_COMPL_ANY">PRIMER_PAIR_4_COMPL_ANY</a>
     + <a href="#PRIMER_PAIR_WT_COMPL_END">PRIMER_PAIR_WT_COMPL_END</a> * <a href="#PRIMER_PAIR_4_COMPL_END">PRIMER_PAIR_4_COMPL_END</a></i>
     + <a href="#PRIMER_PAIR_WT_LIBRARY_MISPRIMING">PRIMER_PAIR_WT_LIBRARY_MISPRIMING</a> * <a href="#PRIMER_PAIR_4_LIBRARY_MISPRIMING">PRIMER_PAIR_4_LIBRARY_MISPRIMING</a>
     + <a href="#PRIMER_PAIR_WT_TEMPLATE_MISPRIMING">PRIMER_PAIR_WT_TEMPLATE_MISPRIMING</a> *
          <a href="#PRIMER_PAIR_4_TEMPLATE_MISPRIMING">PRIMER_PAIR_4_TEMPLATE_MISPRIMING</a>
     If the thermodynamic approach is used then the part of <i>italic</i> in
     the above calculation is replaced by this:
<i>
   If ((min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a>) - 5) &le; <a href="#PRIMER_PAIR_4_COMPL_ANY_TH">PRIMER_PAIR_4_COMPL_ANY_TH</a>) then is added (+):
   + <a href="#PRIMER_PAIR_WT_COMPL_ANY_TH">PRIMER_PAIR_WT_COMPL_ANY_TH</a> *
       (<a href="#PRIMER_PAIR_4_COMPL_ANY_TH">PRIMER_PAIR_4_COMPL_ANY_TH</a> - (min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a>) - 5 - 1))
  else if ((min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a>) - 5) &gt; <a href="#PRIMER_PAIR_4_COMPL_ANY_TH">PRIMER_PAIR_4_COMPL_ANY_TH</a>) then is added (+):
   + <a href="#PRIMER_PAIR_WT_COMPL_ANY_TH">PRIMER_PAIR_WT_COMPL_ANY_TH</a> *
      (1/(min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 + 1 - <a href="#PRIMER_PAIR_4_COMPL_ANY_TH">PRIMER_PAIR_4_COMPL_ANY_TH</a>));
   If ((min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &le; <a href="#PRIMER_PAIR_4_COMPL_END_TH">PRIMER_PAIR_4_COMPL_END_TH</a>) then is added (+):
   + <a href="#PRIMER_PAIR_WT_COMPL_END_TH">PRIMER_PAIR_WT_COMPL_END_TH</a> *
       (<a href="#PRIMER_PAIR_4_COMPL_END_TH">PRIMER_PAIR_4_COMPL_END_TH</a> - (min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a>) - 5 - 1))
  else if ((min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &gt; <a href="#PRIMER_PAIR_4_COMPL_END_TH">PRIMER_PAIR_4_COMPL_END_TH</a>) then is added (+):
    + <a href="#PRIMER_PAIR_WT_COMPL_END_TH">PRIMER_PAIR_WT_COMPL_END_TH</a> *
         (1/(min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 + 1 - <a href="#PRIMER_PAIR_4_COMPL_ANY_TH">PRIMER_PAIR_4_COMPL_ANY_TH</a>));
  If ((min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &le; <a href="#PRIMER_PAIR_4_TEMPLATE_MISPRIMING_TH">PRIMER_PAIR_4_TEMPLATE_MISPRIMING_TH</a>) then is added (+):
   + <a href="#PRIMER_PAIR_WT_TEMPLATE_MISPRIMING_TH">PRIMER_PAIR_WT_TEMPLATE_MISPRIMING_TH</a> *
    (<a href="#PRIMER_PAIR_4_TEMPLATE_MISPRIMING_TH">PRIMER_PAIR_4_TEMPLATE_MISPRIMING_TH</a> - (min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 - 1))
  else if ((min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5) &gt; <a href="#PRIMER_PAIR_4_TEMPLATE_MISPRIMING_TH">PRIMER_PAIR_4_TEMPLATE_MISPRIMING_TH</a>) then is added (+):
   + <a href="#PRIMER_PAIR_WT_TEMPLATE_MISPRIMING_TH">PRIMER_PAIR_WT_TEMPLATE_MISPRIMING_TH</a> *
     (1/(min(<a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>,<a href="#PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM</a> - 5 + 1 - <a href="#PRIMER_PAIR_4_TEMPLATE_MISPRIMING_TH">PRIMER_PAIR_4_TEMPLATE_MISPRIMING_TH</a>));
</i>
</pre>
<p>Primer3 tries to select pairs with the lowest penalty which
still fulfill all necessary requirements like non-redundancy or
product size limits.</p>
<h2><a id="fileFormat">20. PRIMER3 SETTINGS FILE FORMAT</a></h2>
<p>Primer3 can read global settings from a text file at program
start up. This allows the user to save and exchange settings
tailored to special applications.<br>
<br>
Such a Primer3 settings file is a text file. The first three
lines of the file have to be as described below followed by tags
as they would be provided at standard input:</p>
<pre>
Primer3 File - http://primer3.org
P3_FILE_TYPE=settings
P3_FILE_ID=Description of the settings
SEQUENCE_TEMPLATE=ATG...
...
...
...
=
</pre>
<p>
In the first line "Primer3 File - http://primer3.org"
without tailing space. In the second line "P3_FILE_TYPE=settings".
Valid values for P3_FILE_TYPE are all_parameters, sequence and
settings. Up to now, only settings is supported. The third line
has to be empty. It is strongly advised to describe the settings
using the P3_FILE_ID tag. It will print the description of the
settings on the output. From the fourth line on regular
Boulder-IO can be used as it is used in regular input. It also
has to be terminated with a single "=". There can be only one
input per file.</p>
<h2><a id="outputTags">21. OUTPUT TAGS</a></h2>
<p>For each Boulder-IO record passed into Primer3 via stdin, exactly
one Boulder-IO record comes out of Primer3 on stdout.
If a settings file is provided and the option to echo the settings file is given
on the command line, then the contents of the settings file will also be
part of the output. Two additional tags are used to indicate where the
records of the settings file begin and end: P3_SETTINGS_FILE_USED specifies
the path to the settings file that was provided, P3_SETTINGS_FILE_END
does not have any value and it just indicates the end of the settings records.<br>
<br>
The output records contain everything that the input record contains, plus a
subset of the following tag/value pairs.  Unless noted by (*),
each tag appears for each primer pair returned.<br>
<br>
Tags are of the form PRIMER_{LEFT,RIGHT,INTERNAL,PAIR}_&lt;j&gt;_&lt;tag_name&gt;
where &lt;j&gt; is an integer from 0 to n, where n is at most the
value of <a href="#PRIMER_NUM_RETURN">PRIMER_NUM_RETURN</a>. In the documentation the output
number 4 is shown as for example: <a href="#PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM</a>.<br>
<br>
In the descriptions below, 'i,n' represents a start/length pair,
's' represents a string, x represents an arbitrary integer, and f
represents a float.</p>
<h3><a id="PRIMER_ERROR">PRIMER_ERROR=s (*)</a></h3>
<p>s describes user-correctable errors detected in the input
(separated by semicolons).  This tag is absent if there are no
errors.</p>
<h3><a id="PRIMER_WARNING">PRIMER_WARNING=s (*)</a></h3>
<p>s lists warnings generated by Primer3` (separated by semicolons);
this tag is absent if there are no warnings.</p>
<h3><a id="PRIMER_LEFT_NUM_RETURNED">PRIMER_LEFT_NUM_RETURNED=i</a></h3>
<h3><a id="PRIMER_RIGHT_NUM_RETURNED">PRIMER_RIGHT_NUM_RETURNED=i</a></h3>
<h3><a id="PRIMER_INTERNAL_NUM_RETURNED">PRIMER_INTERNAL_NUM_RETURNED=i</a></h3>
<h3><a id="PRIMER_PAIR_NUM_RETURNED">PRIMER_PAIR_NUM_RETURNED=i</a></h3>
<p>i is the number of primers
or primer pairs returned on standard output.
These tags are always generated under
IO version 4 if there are no internal errors and
if <a href="#PRIMER_ERROR">PRIMER_ERROR</a> is not present.<br>
<br>
If primer pairs were requested,
PRIMER_LEFT_NUM_RETURNED and
PRIMER_RIGHT_NUM_RETURNED will be equal to the
number of pairs returned, even if the
actual number of distinct left or right primers
was lower than the number of pairs.  If primer
pairs with internal oligos were requested,
PRIMER_INTERNAL_NUM_RETURNED will also be set
to the number of pairs returned.<br>
<br>
If only left or right primers or hybridization
(internal) oligos were requested, PRIMER_PAIR_NUM_RETURNED
will be 0 and only the relevant tag
will have a non-zero value.
For example, if only left primers were requested,
PRIMER_RIGHT_NUM_RETURNED, PRIMER_INTERNAL_NUM_RETURNED and
PRIMER_PAIR_NUM_RETURNED will be 0.<br>
<br>
Some tasks, such as pick_sequencing_primers or
pick_primer_list, return left and right primers
that are not parts of primer pairs. In this case
PRIMER_PAIR_NUM_RETURNED will be 0.
</p>
<h3><a id="PRIMER_LEFT_4_PROBLEMS">PRIMER_LEFT_4_PROBLEMS=s (*)</a></h3>
<h3><a id="PRIMER_INTERNAL_4_PROBLEMS">PRIMER_INTERNAL_4_PROBLEMS=s (*)</a></h3>
<h3><a id="PRIMER_RIGHT_4_PROBLEMS">PRIMER_RIGHT_4_PROBLEMS=s (*)</a></h3>
<p>s lists the problems (constraint violations) associated
with the corresponding primer oligo.</p>
<h3><a id="PRIMER_LEFT_EXPLAIN">PRIMER_LEFT_EXPLAIN=s (*)</a></h3>
<h3><a id="PRIMER_INTERNAL_EXPLAIN">PRIMER_INTERNAL_EXPLAIN=s (*)</a></h3>
<h3><a id="PRIMER_RIGHT_EXPLAIN">PRIMER_RIGHT_EXPLAIN=s (*)</a></h3>
<p>s is a (more or less) self-documenting string containing
statistics on the possibilities that Primer3 considered in
selecting a single oligo.  For example</p>
<pre>
PRIMER_LEFT_EXPLAIN=considered 62, too many Ns 53, ok 9
PRIMER_RIGHT_EXPLAIN=considered 62, too many Ns 53, ok 9
PRIMER_INTERNAL_OLIGO_EXPLAIN=considered 87, too many Ns 39, overlap excluded region 40, ok 8
</pre>
<p>
All the categories are exclusive, except the 'considered' category.
In some cases the ok count may be higher than the actual number
of ok oligos. This is because a primer can be considered as part of
pair before all of the primer's characteristics have been
computed and checked.  If a primer is never in a legal
pair or never in a pair with a fully evaluated penalty, then
this may occur.  This
situation never results in a primer pair that contains
an illegal primer.
</p>
<h3><a id="PRIMER_PAIR_EXPLAIN">PRIMER_PAIR_EXPLAIN=s (*)</a></h3>
<p>s is a self-documenting string containing statistics on picking a
primer pair (plus internal oligo if requested).  For example</p>
<pre>
PRIMER_PAIR_EXPLAIN=considered 81, unacceptable product size 49, no internal oligo 32, ok 0
</pre>
The purpose of this string is to provide information
in the case that not enough primer pairs are returned.  This
information can be used, for example, to decide
which constraints to relax. In some cases the
information in this string can also give insight into
the causes of long running time.
The counts in the string are only approximate,
because of several reasons:
<ul>
<li>When there are multiple <a href="#PRIMER_PRODUCT_SIZE_RANGE">PRIMER_PRODUCT_SIZE_RANGE</a>s,
a primer pair may not be ok in one size range but ok
in another.  Approximately, the counts for each statistic
are summed over all product size ranges.
</li>
<li>When <a href="#PRIMER_MIN_THREE_PRIME_DISTANCE">PRIMER_MIN_THREE_PRIME_DISTANCE</a>
is &gt; -1, the number of ok primer pairs
reported may be larger than the actual
number. The discrepancy is due to a primer
in the pair that would overlap a primer in a
'better' (lower penalty) pair.  (The search
algorithm does not record state
that would allow it to detect
when a pair that was formerly ok becomes
not ok as the result of
another primer being inserted
into the output list before it is.)
</li>
<li>In some instances,
Primer3 will examine a primer pair before it
discovers that one of the individual
primers in the pair violates specified
constraints.  In this case
<a href="#PRIMER_PAIR_EXPLAIN">PRIMER_PAIR_EXPLAIN</a> might have a non-0
number 'considered', even though one or more of
<a href="#PRIMER_LEFT_EXPLAIN">PRIMER_LEFT_EXPLAIN</a>, <a href="#PRIMER_RIGHT_EXPLAIN">PRIMER_RIGHT_EXPLAIN</a>, or
<a href="#PRIMER_INTERNAL_EXPLAIN">PRIMER_INTERNAL_EXPLAIN</a> has 'ok 0'.
</li>
</ul>
<h3><a id="PRIMER_LEFT_4">PRIMER_LEFT_4=i,n</a></h3>
<p>The selected left primer (the primer to the left in the input
sequence).  i is the 0-based index of the start base of the
primer, and n is t its length.</p>
<h3><a id="PRIMER_INTERNAL_4">PRIMER_INTERNAL_4=i,n</a></h3>
<p>The selected internal oligo. Primer3 outputs this tag if
<a href="#PRIMER_PICK_INTERNAL_OLIGO">PRIMER_PICK_INTERNAL_OLIGO</a> was non-0.  If Primer3 fails to pick a
middle oligo upon request, this tag will not be output.  i is the
0-based index of start base of the internal oligo, and n is its
length.</p>
<h3><a id="PRIMER_RIGHT_4">PRIMER_RIGHT_4=i,n</a></h3>
<p>The selected right primer (the primer to the right in the input
sequence).  i is the 0-based index of the last base of the
primer, and n is its length.</p>
<h3><a id="PRIMER_LEFT_4_SEQUENCE">PRIMER_LEFT_4_SEQUENCE=s</a></h3>
<h3><a id="PRIMER_INTERNAL_4_SEQUENCE">PRIMER_INTERNAL_4_SEQUENCE=s</a></h3>
<h3><a id="PRIMER_RIGHT_4_SEQUENCE">PRIMER_RIGHT_4_SEQUENCE=s</a></h3>
<p>The actual sequence of the oligo. The sequence of left primer and
internal oligo is presented 5' -> 3' on the same strand as the
input <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a> (which must be presented 5' -> 3').  The
sequence of the right primer is presented 5' -> 3' on the
opposite strand from the input <a href="#SEQUENCE_TEMPLATE">SEQUENCE_TEMPLATE</a>.</p>
<h3><a id="PRIMER_PAIR_4_PRODUCT_SIZE">PRIMER_PAIR_4_PRODUCT_SIZE=x</a></h3>
<p>x is the product size of the PCR product.</p>
<h3><a id="PRIMER_LEFT_4_PENALTY">PRIMER_LEFT_4_PENALTY=f</a></h3>
<h3><a id="PRIMER_INTERNAL_4_PENALTY">PRIMER_INTERNAL_4_PENALTY=f</a></h3>
<h3><a id="PRIMER_RIGHT_4_PENALTY">PRIMER_RIGHT_4_PENALTY=f</a></h3>
<p>The contribution of this individual primer or oligo to the
objective function.</p>
<h3><a id="PRIMER_PAIR_4_PENALTY">PRIMER_PAIR_4_PENALTY=f</a></h3>
<p>The value of the objective function for this pair (lower is better).</p>
<h3><a id="PRIMER_LEFT_4_TM">PRIMER_LEFT_4_TM=f</a></h3>
<h3><a id="PRIMER_INTERNAL_4_TM">PRIMER_INTERNAL_4_TM=f</a></h3>
<h3><a id="PRIMER_RIGHT_4_TM">PRIMER_RIGHT_4_TM=f</a></h3>
<p>The melting TM for the selected oligo.</p>
<h3><a id="PRIMER_PAIR_4_PRODUCT_TM">PRIMER_PAIR_4_PRODUCT_TM=f</a></h3>
<p>f is the melting temperature of the product. Calculated using
equation (iii) from the paper [Rychlik W, Spencer WJ and Rhoads
RE (1990) "Optimization of the annealing temperature for DNA
amplification in vitro", Nucleic Acids Res 18:6409-12
<a href="http://dx.doi.org/10.1093/nar/18.21.6409">http://dx.doi.org/10.1093/nar/18.21.6409</a>].
Printed only if a non-default value of <a href="#PRIMER_PRODUCT_MAX_TM">PRIMER_PRODUCT_MAX_TM</a> or
<a href="#PRIMER_PRODUCT_MIN_TM">PRIMER_PRODUCT_MIN_TM</a> is specified.</p>
<h3><a id="PRIMER_PAIR_4_PRODUCT_TM_OLIGO_TM_DIFF">PRIMER_PAIR_4_PRODUCT_TM_OLIGO_TM_DIFF=f</a></h3>
<p>f is the difference between the melting temperature of the
product and the melting temperature of the less stable primer.
Printed only if <a href="#PRIMER_PRODUCT_MAX_TM">PRIMER_PRODUCT_MAX_TM</a> or <a href="#PRIMER_PRODUCT_MIN_TM">PRIMER_PRODUCT_MIN_TM</a> is
specified.</p>
<h3><a id="PRIMER_PAIR_4_T_OPT_A">PRIMER_PAIR_4_T_OPT_A=f</a></h3>
<p>f is T sub a super OPT from equation (i) in [Rychlik W, Spencer
WJ and Rhoads RE (1990) "Optimization of the annealing
temperature for DNA amplification in vitro", Nucleic Acids Res
18:6409-12. <a href="http://dx.doi.org/10.1093/nar/18.21.6409">http://dx.doi.org/10.1093/nar/18.21.6409</a>].
Printed only if <a href="#PRIMER_PRODUCT_MAX_TM">PRIMER_PRODUCT_MAX_TM</a> or <a href="#PRIMER_PRODUCT_MIN_TM">PRIMER_PRODUCT_MIN_TM</a> is
specified.</p>
<h3><a id="PRIMER_LEFT_4_GC_PERCENT">PRIMER_LEFT_4_GC_PERCENT=f</a></h3>
<h3><a id="PRIMER_INTERNAL_4_GC_PERCENT">PRIMER_INTERNAL_4_GC_PERCENT=f</a></h3>
<h3><a id="PRIMER_RIGHT_4_GC_PERCENT">PRIMER_RIGHT_4_GC_PERCENT=f</a></h3>
<p>The percent GC for the selected oligo (denominator is the number
of non-ambiguous bases).</p>
<h3><a id="PRIMER_LEFT_4_SELF_ANY">PRIMER_LEFT_4_SELF_ANY=f</a></h3>
<h3><a id="PRIMER_INTERNAL_4_SELF_ANY">PRIMER_INTERNAL_4_SELF_ANY=f</a></h3>
<h3><a id="PRIMER_RIGHT_4_SELF_ANY">PRIMER_RIGHT_4_SELF_ANY=f</a></h3>
<p>The calculated value for the tendency of a primer to bind to
itself (interfering with target sequence binding). It will score
ANY binding occurring within the entire primer sequence. For
details see <a href="#PRIMER_MAX_SELF_ANY">PRIMER_MAX_SELF_ANY</a>.<br>
The self-complementarity measures for the selected oligo.</p>
<h3><a id="PRIMER_LEFT_4_SELF_ANY_TH">PRIMER_LEFT_4_SELF_ANY_TH=f</a></h3>
<h3><a id="PRIMER_INTERNAL_4_SELF_ANY_TH">PRIMER_INTERNAL_4_SELF_ANY_TH=f</a></h3>
<h3><a id="PRIMER_RIGHT_4_SELF_ANY_TH">PRIMER_RIGHT_4_SELF_ANY_TH=f</a></h3>
<p>The calculated value for the tendency of a primer to bind to
itself (interfering with target sequence binding). It will calculate the
melting temperature for ANY binding occurring within the entire primer
sequence. For details see <a href="#PRIMER_MAX_SELF_ANY_TH">PRIMER_MAX_SELF_ANY_TH</a>.
The self-complementarity measures for the selected oligo.</p>
<h3><a id="PRIMER_LEFT_4_SELF_ANY_STUCT">PRIMER_LEFT_4_SELF_ANY_STUCT=s</a></h3>
<h3><a id="PRIMER_INTERNAL_4_SELF_ANY_STUCT">PRIMER_INTERNAL_4_SELF_ANY_STUCT=s</a></h3>
<h3><a id="PRIMER_RIGHT_4_SELF_ANY_STUCT">PRIMER_RIGHT_4_SELF_ANY_STUCT=s</a></h3>
<p>A string representation of the calculated secondary
structure. The tag is only present if a secondary structure could
be calculated and <a href="#PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT</a>=1.<br>
See <a href="#PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT</a> for examples and
notes on the necessary reformating of the string.
</p>
<h3><a id="PRIMER_LEFT_4_SELF_END">PRIMER_LEFT_4_SELF_END=f</a></h3>
<h3><a id="PRIMER_INTERNAL_4_SELF_END">PRIMER_INTERNAL_4_SELF_END=f</a></h3>
<h3><a id="PRIMER_RIGHT_4_SELF_END">PRIMER_RIGHT_4_SELF_END=f</a></h3>
<p>The calculated value for the tendency of the 3'-END to bind to a
identical primer. This is critical for primer quality because it
allows primers use itself as a target and amplify a short piece
(forming a primer-dimer). These primer are then unable to bind
and amplify the target sequence. For details see <a href="#PRIMER_MAX_SELF_END">PRIMER_MAX_SELF_END</a>.<br>
The self-complementarity measures for the ends of selected oligo.</p>
<h3><a id="PRIMER_LEFT_4_SELF_END_TH">PRIMER_LEFT_4_SELF_END_TH=f</a></h3>
<h3><a id="PRIMER_INTERNAL_4_SELF_END_TH">PRIMER_INTERNAL_4_SELF_END_TH=f</a></h3>
<h3><a id="PRIMER_RIGHT_4_SELF_END_TH">PRIMER_RIGHT_4_SELF_END_TH=f</a></h3>
<p>The calculated value for the tendency of the 3'-END to bind
to a identical primer. This is critical for primer quality because it allows
primers use itself as a target and amplify a short piece (forming a
primer-dimer). These primer are then unable to bind and amplify the target
sequence. For details see <a href="#PRIMER_MAX_SELF_END_TH">PRIMER_MAX_SELF_END_TH</a>.
The self-complementarity measures for the ends of selected oligo.</p>
<h3><a id="PRIMER_LEFT_4_SELF_END_STUCT">PRIMER_LEFT_4_SELF_END_STUCT=s</a></h3>
<h3><a id="PRIMER_INTERNAL_4_SELF_END_STUCT">PRIMER_INTERNAL_4_SELF_END_STUCT=s</a></h3>
<h3><a id="PRIMER_RIGHT_4_SELF_END_STUCT">PRIMER_RIGHT_4_SELF_END_STUCT=s</a></h3>
<p>A string representation of the calculated secondary
structure. The tag is only present if a secondary structure could
be calculated and <a href="#PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT</a>=1.<br>
See <a href="#PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT</a> for examples and
notes on the necessary reformating of the string.
</p>
<h3><a id="PRIMER_LEFT_4_HAIRPIN_TH">PRIMER_LEFT_4_HAIRPIN_TH=f</a></h3>
<h3><a id="PRIMER_INTERNAL_4_HAIRPIN_TH">PRIMER_INTERNAL_4_HAIRPIN_TH=f</a></h3>
<h3><a id="PRIMER_RIGHT_4_HAIRPIN_TH">PRIMER_RIGHT_4_HAIRPIN_TH=f</a></h3>
<p>The calculated value of melting temperature of hairpin
structure of primer. For details see <a href="#PRIMER_MAX_HAIRPIN_TH">PRIMER_MAX_HAIRPIN_TH</a></p>
<h3><a id="PRIMER_LEFT_4_HAIRPIN_STUCT">PRIMER_LEFT_4_HAIRPIN_STUCT=s</a></h3>
<h3><a id="PRIMER_INTERNAL_4_HAIRPIN_STUCT">PRIMER_INTERNAL_4_HAIRPIN_STUCT=s</a></h3>
<h3><a id="PRIMER_RIGHT_4_HAIRPIN_STUCT">PRIMER_RIGHT_4_HAIRPIN_STUCT=s</a></h3>
<p>A string representation of the calculated secondary
structure. The tag is only present if a secondary structure could
be calculated and <a href="#PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT</a>=1.<br>
See <a href="#PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT</a> for examples and
notes on the necessary reformating of the string.
</p>
<h3><a id="PRIMER_PAIR_4_COMPL_ANY">PRIMER_PAIR_4_COMPL_ANY=f</a></h3>
<p>The calculated value for the tendency of a primer pair to bind to
each other (interfering with target sequence binding). It will
score ANY binding occurring within the entire primer sequence.
For details see <a href="#PRIMER_MAX_SELF_ANY">PRIMER_MAX_SELF_ANY</a>.<br>
The inter-pair complementarity measures over the complete primer
for selected left and right primer.</p>
<h3><a id="PRIMER_PAIR_4_COMPL_ANY_TH">PRIMER_PAIR_4_COMPL_ANY_TH=f</a></h3>
<p>The calculated value for the tendency of a primer pair to
bind to each other (interfering with target sequence binding). It will
calculate the melting temperature of ANY binding occurring within the entire
primer sequence. For details see <a href="#PRIMER_MAX_SELF_ANY_TH">PRIMER_MAX_SELF_ANY_TH</a>.
The inter-pair complementarity measures over the complete primer for
selected left and right primer.</p>
<h3><a id="PRIMER_PAIR_4_COMPL_ANY_STUCT">PRIMER_PAIR_4_COMPL_ANY_STUCT=s</a></h3>
<p>A string representation of the calculated secondary
structure. The tag is only present if a secondary structure could
be calculated and <a href="#PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT</a>=1.<br>
See <a href="#PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT</a> for examples and
notes on the necessary reformating of the string.
</p>
<h3><a id="PRIMER_PAIR_4_COMPL_END">PRIMER_PAIR_4_COMPL_END=f</a></h3>
<p>The calculated value for the tendency of the 3'-ENDs of a primer pair to bind to
each other. This is critical for primer quality because it
allows primers use itself as a target and amplify a short piece
(forming a primer-dimer). These primer are then unable to bind
and amplify the target sequence. For details see
<a href="#PRIMER_MAX_SELF_END">PRIMER_MAX_SELF_END</a>.<br>
The inter-pair complementarity measures for the ends of selected
left and right primer.</p>
<h3><a id="PRIMER_PAIR_4_COMPL_END_TH">PRIMER_PAIR_4_COMPL_END_TH=f</a></h3>
<p>The calculated value for the tendency of the 3'-ENDs of a
primer pair to bind to each other. This is critical for primer quality
because it allows primers use itself as a target and amplify a short piece
(forming a primer-dimer). These primer are then unable to bind and amplify
the target sequence. For details see <a href="#PRIMER_MAX_SELF_END_TH">PRIMER_MAX_SELF_END_TH</a>.
The inter-pair complementarity measures for the ends of selected left and
right primer.</p>
<h3><a id="PRIMER_PAIR_4_COMPL_END_STUCT">PRIMER_PAIR_4_COMPL_END_STUCT=s</a></h3>
<p>A string representation of the calculated secondary
structure. The tag is only present if a secondary structure could
be calculated and <a href="#PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT</a>=1.<br>
See <a href="#PRIMER_SECONDARY_STRUCTURE_ALIGNMENT">PRIMER_SECONDARY_STRUCTURE_ALIGNMENT</a> for examples and
notes on the necessary reformating of the string.
</p>
<h3><a id="PRIMER_LEFT_4_END_STABILITY">PRIMER_LEFT_4_END_STABILITY=f</a></h3>
<h3><a id="PRIMER_RIGHT_4_END_STABILITY">PRIMER_RIGHT_4_END_STABILITY=f</a></h3>
<p>f is the delta G of disruption of the five 3' bases of the
primer.</p>
<h3><a id="PRIMER_LEFT_4_TEMPLATE_MISPRIMING">PRIMER_LEFT_4_TEMPLATE_MISPRIMING=f</a></h3>
<h3><a id="PRIMER_RIGHT_4_TEMPLATE_MISPRIMING">PRIMER_RIGHT_4_TEMPLATE_MISPRIMING=f</a></h3>
<h3><a id="PRIMER_PAIR_4_TEMPLATE_MISPRIMING">PRIMER_PAIR_4_TEMPLATE_MISPRIMING=f</a></h3>
<p>Analogous to PRIMER_{LEFT,RIGHT,PAIR}_LIBRARY_MISPRIMING, except that
these output tags apply to mispriming within the template sequence.
This often arises, for example, in genes with repeated exons. For
backward compatibility, these tags only appear if the corresponding
input tags have defined values.</p>
<h3><a id="PRIMER_LEFT_4_TEMPLATE_MISPRIMING_TH">PRIMER_LEFT_4_TEMPLATE_MISPRIMING_TH=f</a></h3>
<h3><a id="PRIMER_RIGHT_4_TEMPLATE_MISPRIMING_TH">PRIMER_RIGHT_4_TEMPLATE_MISPRIMING_TH=f</a></h3>
<h3><a id="PRIMER_PAIR_4_TEMPLATE_MISPRIMING_TH">PRIMER_PAIR_4_TEMPLATE_MISPRIMING_TH=f</a></h3>
<p>These output tags apply to mispriming within the template sequence and the calculation
method is based on thermodynamical approach. This often arises, for example,
in genes with repeated exons.</p>
<h3><a id="PRIMER_LEFT_4_LIBRARY_MISPRIMING">PRIMER_LEFT_4_LIBRARY_MISPRIMING=f, s</a></h3>
<h3><a id="PRIMER_RIGHT_4_LIBRARY_MISPRIMING">PRIMER_RIGHT_4_LIBRARY_MISPRIMING=f, s</a></h3>
<h3><a id="PRIMER_PAIR_4_LIBRARY_MISPRIMING">PRIMER_PAIR_4_LIBRARY_MISPRIMING=f, s</a></h3>
<p>f is the maximum mispriming score for the right primer
against any sequence in the given <a href="#PRIMER_MISPRIMING_LIBRARY">PRIMER_MISPRIMING_LIBRARY</a>;
s is the id of corresponding library sequence.
<a href="#PRIMER_PAIR_MAX_LIBRARY_MISPRIMING">PRIMER_PAIR_MAX_LIBRARY_MISPRIMING</a> is the maximum sum of
mispriming scores in any single library sequence (perhaps a
more reasonable estimator of the likelihood of mispriming).</p>
<h3><a id="PRIMER_INTERNAL_4_LIBRARY_MISHYB">PRIMER_INTERNAL_4_LIBRARY_MISHYB=f, s</a></h3>
<p>f is the maximum mishybridization score for the right primer
against any sequence in the given
<a href="#PRIMER_INTERNAL_MISHYB_LIBRARY">PRIMER_INTERNAL_MISHYB_LIBRARY</a>; s is the id of
corresponding library sequence.</p>
<h3><a id="PRIMER_LEFT_4_MIN_SEQ_QUALITY">PRIMER_LEFT_4_MIN_SEQ_QUALITY=i</a></h3>
<h3><a id="PRIMER_INTERNAL_4_MIN_SEQ_QUALITY">PRIMER_INTERNAL_4_MIN_SEQ_QUALITY=i</a></h3>
<h3><a id="PRIMER_RIGHT_4_MIN_SEQ_QUALITY">PRIMER_RIGHT_4_MIN_SEQ_QUALITY=i</a></h3>
<p>i is the minimum _sequence_ quality within the primer
or oligo (not to be confused with the <a href="#PRIMER_PAIR_4_PENALTY">PRIMER_PAIR_4_PENALTY</a>
output tag, which is really the value of the objective
function.)</p>
<h3><a id="PRIMER_STOP_CODON_POSITION">PRIMER_STOP_CODON_POSITION=i</a></h3>
<p>i is the position of the first base of the stop codon,
if Primer3 found one, or -1 if Primer3 did not.  Printed
only if the input tag <a href="#SEQUENCE_START_CODON_POSITION">SEQUENCE_START_CODON_POSITION</a> with a
non-default value is supplied.</p>
<h3><a id="PRIMER_LEFT_4_POSITION_PENALTY">PRIMER_LEFT_4_POSITION_PENALTY=i</a></h3>
<h3><a id="PRIMER_RIGHT_4_POSITION_PENALTY">PRIMER_RIGHT_4_POSITION_PENALTY=i</a></h3>
<p>i is the penalty of the primer by its
position.</p>
<h2><a id="exampleOutput">22. EXAMPLE OUTPUT</a></h2>
<p>You should run it yourself.  Use the file 'example'
that came with this distribution
directory as input.</p>
<h2><a id="pickAdvice">23. ADVICE FOR PICKING PRIMERS</a></h2>
<p>We suggest consulting: Wojciech Rychlik (1993) "Selection of
Primers for Polymerase Chain Reaction" in BA White, Ed., "Methods
in Molecular Biology, Vol. 15: PCR Protocols: Current Methods and
Applications", pp 31-40, Humana Press, Totowa NJ.</p>
<h2><a id="cautions">24. CAUTIONS</a></h2>
<p>Some of the most important issues in primer picking can be
addressed only before using Primer3. These are sequence quality
(including making sure the sequence is not vector and not chimeric)
and avoiding repetitive elements.<br>
<br>
Techniques for avoiding problems include a thorough understanding
of possible vector contaminants and cloning artifacts coupled
with database searches using blast, Fasta, or other similarity
searching program to screen for vector contaminants and possible
repeats.  Repbase (J. Jurka, A.F.A. Smit, C. Pethiyagoda, and
others, 1995-1996, <a href="ftp://ftp.ncbi.nih.gov/repository/repbase/">ftp://ftp.ncbi.nih.gov/repository/repbase/</a>)
is an excellent source of repeat sequences and pointers to the
literature.  (The Repbase files need to be converted to Fasta format
before they can be used by Primer3.) Primer3 now allows you to screen
candidate oligos against a Mispriming Library (or a Mishyb Library in
the case of internal oligos).<br>
<br>
Sequence quality can be controlled by manual trace viewing and
quality clipping or automatic quality clipping programs.  Low-
quality bases should be changed to N's or can be made part of
Excluded Regions. The beginning of a sequencing read is often
problematic because of primer peaks, and the end of the read
often contains many low-quality or even meaningless called bases.
Therefore, when picking primers from single-pass sequence it is
often best to use the <a href="#SEQUENCE_INCLUDED_REGION">SEQUENCE_INCLUDED_REGION</a> parameter to
ensure that Primer3 chooses primers in the high quality region
of the read.<br>
<br>
In addition, Primer3 takes as input a <a href="#SEQUENCE_QUALITY">SEQUENCE_QUALITY</a> list for
use with those base calling programs (e.g. Phred)
that output this information.</p>
<h2><a id="findNoPrimers">25. WHAT TO DO IF PRIMER3 CANNOT FIND ANY PRIMERS?</a></h2>
<p>Try relaxing various parameters, including the
self-complementarity parameters and max and min oligo melting
temperatures.  For example, for very A-T-rich regions you might
have to increase maximum primer size or decrease minimum melting
temperature.  It is usually unwise to reduce the minimum primer
size if your template is complex (e.g. a mammalian genome), since
small primers are more likely to be non-specific.  Make sure that
there are adequate stretches of non-Ns in the regions in which
you wish to pick primers.  If necessary you can also allow an N
in your primer and use an oligo mixture containing all four bases
at that position.<br>
<br>
Try setting the <a href="#PRIMER_EXPLAIN_FLAG">PRIMER_EXPLAIN_FLAG</a> input tag.</p>
<h2><a id="earlierVersions">26. DIFFERENCES FROM EARLIER VERSIONS</a></h2>
<p>The section HOW TO MIGRATE TAGS TO
IO VERSION 4 describes th modified tags in detail.
See also the file release_notes.txt in this directory.
</p>
<h2><a id="exitStatusCodes">27. EXIT STATUS CODES</a></h2>
<pre> 0 on normal operation
-1 under the following conditions:
   illegal command-line arguments.
   unable to fflush stdout.
   unable to open (for writing and creating) a .for, .rev
     or .int file (probably due to a protection problem).
-2 on out-of-memory
-3 empty input
-4 error in a "Global" input tag (message in PRIMER_ERROR).
</pre>
<p>Primer3 calls abort() and dumps core (if possible) if a
programming error is detected by an assertion violation.<br>
<br>
SIGINT and SIGTERM are handled essentially as empty input, except
the signal received is returned as the exit status and printed to
stderr.<br>
<br>
In all of the error cases above Primer3 prints a message to stderr.</p>
<h2><a id="webInterface">28. PRIMER3 WWW INTERFACES</a></h2>
<p>There are two web interfaces available :<br>
The Bioinformatics workgroup at University of Tartu provides a basic
web-based interface to Primer3 named Primer3Web at
<a href="http://primer3.ut.ee/">http://primer3.ut.ee/</a><br>
<br>
A Primer3Plus web services is at
<a href="http://primer3plus.com">http://primer3plus.com</a><br>
<br>
<br>
Web interface code is available
on GitHub: <a href="https://github.com/primer3-org">https://github.com/primer3-org</a>.</p>
<h2><a id="acknowledgments">29. ACKNOWLEDGMENTS</a></h2>
<p>Initial development of Primer3 was funded by Howard Hughes Medical
Institute and by the National Institutes of Health, National Human
Genome Research Institute under grants R01-HG00257 (to David C. Page)
and P50-HG00098 (to Eric S. Lander),
but ongoing development and maintenance are not currently funded.
<br>
<br>
Primer3 was originally written by Helen J. Skaletsky (Howard Hughes
Medical Institute, Whitehead Institute) and Steve Rozen (Duke-NUS
Graduate Medical School Singapore, formerly at Whitehead Institute)
based on the design of earlier versions, notably Primer 0.5
(Steve Lincoln, Mark Daly, and Eric S. Lander).
The original web interface was designed by Richard Resnick.  Lincoln
Stein designed the Boulder-IO format in the days before XML and RDF, and
championed the idea of making Primer3 a software component, which
has been key to its wide utility.<br>
<br>
In addition, among others, Ernst Molitor, Carl Foeller, and James Bonfield
contributed to the early
design of Primer3. Brant Faircloth has helped with
ensuring that Primer3 runs on Windows and MacOS and with the
Primer3 web site.
Triinu Koressaar and Maido Remm modernized the melting
temperature calculations in 2008.
Triinu Koressaar added secondary structure,
primer-dimer, and template mispriming based on a thermodynamic
model in 2.2.0.
Ioana Cutcutache is responsible for most of the
remaining improvements
in 2.2.0, including performance enhancements, modern
command line arguments, and new input tags to
control primer location (with the "overlap junction"
tags initially implemented by Andreas Untergasser).
Jian Ye patiently provided new requirements.<br>
<br>
Harm Nijveen and Andreas Untergasser developed the webinterface
Primer3Plus in 2006-2009. Currently Primer3Plus is maintained by
Andreas Untergasser.
<br>
<br>
Primer3 is an open software development project hosted
on GitHub: <a href="https://github.com/primer3-org">https://github.com/primer3-org</a>.
</p>
</div>
</body>
</html>
